// @grapesjs/studio-sdk-plugins/dist/layoutSidebarButtons/index.es.js
var B = /* @__PURE__ */ ((e) => (e.free = "free", e.startup = "startup", e.business = "business", e.enterprise = "enterprise", e))(B || {}), T = /* @__PURE__ */ ((e) => (e.toastAdd = "studio:toastAdd", e.toastRemove = "studio:toastRemove", e.dialogOpen = "studio:dialogOpen", e.dialogClose = "studio:dialogClose", e.dialogExportCode = "studio:dialogExportCode", e.dialogImportCode = "studio:dialogImportCode", e.sidebarLeftSet = "studio:sidebarLeft:set", e.sidebarLeftGet = "studio:sidebarLeft:get", e.sidebarLeftToggle = "studio:sidebarLeft:toggle", e.sidebarRightSet = "studio:sidebarRight:set", e.sidebarRightGet = "studio:sidebarRight:get", e.sidebarRightToggle = "studio:sidebarRight:toggle", e.sidebarTopSet = "studio:sidebarTop:set", e.sidebarTopGet = "studio:sidebarTop:get", e.sidebarTopToggle = "studio:sidebarTop:toggle", e.sidebarBottomSet = "studio:sidebarBottom:set", e.sidebarBottomGet = "studio:sidebarBottom:get", e.sidebarBottomToggle = "studio:sidebarBottom:toggle", e.symbolAdd = "studio:symbolAdd", e.symbolDetach = "studio:symbolDetach", e.symbolOverride = "studio:symbolOverride", e.symbolPropagateStyles = "studio:propagateStyles", e.getPagesConfig = "studio:getPagesConfig", e.setPagesConfig = "studio:setPagesConfig", e.getPageSettings = "studio:getPageSettings", e.setPageSettings = "studio:setPageSettings", e.clearPage = "studio:clearPage", e.projectFiles = "studio:projectFiles", e.validateCode = "studio:validateCode", e.canvasReload = "studio:canvasReload", e.getBlocksPanel = "studio:getBlocksPanel", e.setBlocksPanel = "studio:setBlocksPanel", e.getStateContextMenu = "studio:getStateContextMenu", e.setStateContextMenu = "studio:setStateContextMenu", e.contextMenuComponent = "studio:contextMenuComponent", e.layoutAdd = "studio:layoutAdd", e.layoutRemove = "studio:layoutRemove", e.layoutToggle = "studio:layoutToggle", e.layoutUpdate = "studio:layoutUpdate", e.layoutGet = "studio:layoutGet", e.layoutConfigGet = "studio:layoutConfigGet", e.layoutConfigSet = "studio:layoutConfigSet", e.layoutComponentAdd = "studio:layoutComponentAdd", e.layoutComponentGet = "studio:layoutComponentGet", e.layoutComponentRemove = "studio:layoutComponentRemove", e.layoutComponentRender = "studio:layoutComponentRender", e.getStateTheme = "studio:getStateTheme", e.setStateTheme = "studio:setStateTheme", e.settings = "studio:settings", e.assetProviderGet = "studio:assetProviderGet", e.assetProviderAdd = "studio:assetProviderAdd", e.assetProviderRemove = "studio:assetProviderRemove", e.fontGet = "studio:fontGet", e.fontAdd = "studio:fontAdd", e.fontRemove = "studio:fontRemove", e.fontManagerOpen = "studio:fontManagerOpen", e.menuFontLoad = "studio:menuFontLoad", e.toggleStateDataSource = "studio:toggleStateDataSource", e.getStateDataSource = "studio:getStateDataSource", e.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", e.dataSourceSetImporter = "studio:dataSourceSetImporter", e.dataSourceSetExporter = "studio:dataSourceSetExporter", e.setDragAbsolute = "studio:setDragAbsolute", e))(T || {}), I = /* @__PURE__ */ ((e) => (e.layoutToggleId = "studio:layoutToggle:", e.toggleBlocksPanel = "studio:toggleBlocksPanel", e.pageSettingsUpdate = "studio:pageSettingsUpdate", e.toggleDataSourcesPreview = "studio:toggleDataSourcesPreview", e))(I || {});
const x = "app.grapesjs.com", O = "app-stage.grapesjs.com", _ = "app2.grapesjs.com", H = "app-stage2.grapesjs.com", D = [
  x,
  _,
  O,
  H,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], j = "license:check:end", W = () => typeof window < "u", S = ({
  isDev: e,
  isStage: a,
  isPlatform: o
}) => `${e ? "" : `https://${a ? O : x}`}/${o ? "platform-api" : "api"}`, m = () => {
  const e = W() && window.location.hostname;
  return !!e && (D.includes(e) || D.some((a) => e.endsWith(a)));
};
async function F({
  path: e,
  baseApiUrl: a,
  method: o = "GET",
  headers: c = {},
  params: i,
  body: h
}) {
  const r = `${a || S({ isDev: !1, isStage: !1 })}${e}`, d = {
    method: o,
    headers: {
      "Content-Type": "application/json",
      ...c
    }
  };
  h && (d.body = JSON.stringify(h));
  const b = i ? new URLSearchParams(i).toString() : "", u = b ? `?${b}` : "", t = await fetch(`${r}${u}`, d);
  if (!t.ok)
    throw new Error(`HTTP error! status: ${t.status}`);
  return t.json();
}
const G = {
  [B.free]: 0,
  [B.startup]: 10,
  [B.business]: 20,
  [B.enterprise]: 30
};
function V(e) {
  const a = e;
  return a.init = (o) => (c) => e(c, o), a;
}
const C = (e) => /* @__PURE__ */ V(e);
async function K({
  editor: e,
  plan: a,
  pluginName: o,
  licenseKey: c,
  onLicenseCheckResponse: i,
  cleanup: h
}) {
  let f = "", r;
  const d = (t) => {
    console.warn("Cleanup plugin:", o, "Reason:", t), h();
  }, b = (t = {}) => {
    var l;
    const { error: L, sdkLicense: n } = t, g = (l = t.plan) == null ? void 0 : l.category;
    if (!(n || t.license) || L)
      d(L || "Invalid license");
    else if (g) {
      const y = G[a], p = G[g];
      y > p && d({ pluginRequiredPlan: a, licensePlan: g });
    }
  };
  e.Commands.has(T.settings) && (r = e.runCommand(T.settings), f = (r == null ? void 0 : r.baseUrl) || "");
  const u = (t) => {
    i == null || i(t), t && b(t);
  };
  if (!r) {
    e.onReady(async () => {
      if (!m())
        if (c) {
          const t = await Z({ licenseKey: c, pluginName: o, baseApiUrl: f });
          u(t);
        } else
          d("The `licenseKey` option not provided");
    });
    return;
  }
  if (r.licensePlan || r.licenseError) {
    const t = q(r);
    u(t);
    return;
  }
  e.on(j, (t) => u(t));
}
const q = (e) => ({
  sdkLicense: e.license,
  license: e.license,
  error: e.licenseError,
  plan: e.licensePlan
});
async function Z(e) {
  const { licenseKey: a, pluginName: o, baseApiUrl: c } = e;
  try {
    return (await F({
      baseApiUrl: c,
      path: `/sdk/${a || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: o
      }
    })).result || {};
  } catch (i) {
    return console.error("Error during SDK license check:", i), !1;
  }
}
var v = /* @__PURE__ */ ((e) => (e.panelBlocks = "panelBlocks", e.panelPagesLayers = "panelPagesLayers", e.panelGlobalStyles = "panelGlobalStyles", e.panelSidebarTabs = "panelSidebarTabs", e.panelAssets = "panelAssets", e.aiChatPanel = "aiChatPanel", e.panelDataSources = "panelDataSources", e))(v || {});
const E = "sidebarButtonsTarget", w = (e, a) => {
  const { id: o, icon: c, label: i, tooltip: h, className: f, skipSelfClose: r, removeLayouts: d, layoutComponent: b, layoutCommand: u } = e, { pluginOpts: t, breakpointProps: L } = a, n = o;
  return {
    id: U(o),
    type: "button",
    icon: c,
    tooltip: i || h,
    className: f,
    editorEvents: {
      [`${I.layoutToggleId}${n}`]: ({ fromEvent: g, setState: s, editor: l }) => {
        s({ active: g.isOpen }), setTimeout(() => l.refresh({ tools: !0 }), 20);
      }
    },
    onClick: ({ editor: g, state: s }) => {
      var R;
      if (s.active && r) return;
      const l = (u == null ? void 0 : u.placer) || { type: "static", layoutId: E }, y = { ...l, skipCleanup: l.type === "static" }, p = d || Object.keys(v).filter((k) => k !== n);
      $(g, { removeLayouts: p, layout: { placer: y } });
      const A = {
        id: n,
        placer: l,
        header: (u == null ? void 0 : u.header) ?? { label: i, close: !r },
        layout: b,
        style: {
          width: 280,
          height: "100%",
          borderRightWidth: 1,
          ...u == null ? void 0 : u.style
        }
      }, P = ((R = t == null ? void 0 : t.sidebarLayoutCommand) == null ? void 0 : R.call(t, {
        ...L,
        layoutCommand: A,
        sidebarButtonProps: e
      })) || A;
      g.runCommand(T.layoutToggle, { ...P });
    }
  };
}, U = (e) => `button__${e}`, $ = (e, a = {}) => {
  (a.removeLayouts || Object.keys(v)).forEach((c) => e.runCommand(T.layoutRemove, { id: c, layout: a.layout }));
}, N = (e) => {
  const a = (e == null ? void 0 : e.breakpointTablet) ?? 1024, o = (e == null ? void 0 : e.breakpointMobile) ?? 768, c = {}, i = [
    {
      id: v.panelBlocks,
      icon: "plusBox",
      label: "Blocks",
      layoutComponent: { type: "panelBlocks" }
    },
    {
      id: v.panelPagesLayers,
      icon: "layers",
      label: "Pages & Layers",
      layoutComponent: { type: "panelPagesLayers" }
    },
    {
      id: v.panelGlobalStyles,
      icon: '<svg viewBox="0 0 24 24"> <path d="M20 14H6c-2.2 0-4 1.8-4 4s1.8 4 4 4h14a2 2 0 0 0 2-2v-4a2 2 0 0 0-2-2M6 20c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2m.3-8L13 5.3a2 2 0 0 1 2.8 0l2.8 2.8c.8.8.8 2 0 2.8l-.9 1.1H6.3M2 13.5V4c0-1.1.9-2 2-2h4a2 2 0 0 1 2 2v1.5l-8 8Z"/></svg>',
      label: "Global Styles",
      layoutComponent: { type: "panelGlobalStyles" }
    },
    {
      id: v.panelAssets,
      icon: '<svg viewBox="0 0 24 24"><path d="M22,16V4A2,2 0 0,0 20,2H8A2,2 0 0,0 6,4V16A2,2 0 0,0 8,18H20A2,2 0 0,0 22,16M11,12L13.03,14.71L16,11L20,16H8M2,6V20A2,2 0 0,0 4,22H18V20H4V6" /></svg>',
      label: "Assets",
      layoutComponent: {
        type: "panelAssets",
        content: { itemsPerRow: 2, header: { addUrl: !1 } },
        style: { padding: 7 },
        onSelect: ({ assetProps: n, editor: g }) => {
          var p;
          const s = g.getSelected(), l = { type: "image", src: n.src };
          let y;
          if (s != null && s.is("image"))
            return s.set("src", n.src);
          s && g.Components.canMove(s, l) ? y = s.append(l)[0] : s || (y = (p = g.getWrapper()) == null ? void 0 : p.append(l)[0]), y && g.select(y, { scroll: !0 });
        }
      }
    }
  ], h = [
    ...i,
    {
      id: v.panelSidebarTabs,
      layoutCommand: { header: !1 },
      icon: '<svg viewBox="0 0 24 24"><path d="M17.5 12a1.5 1.5 0 0 1-1.5-1.5A1.5 1.5 0 0 1 17.5 9a1.5 1.5 0 0 1 1.5 1.5 1.5 1.5 0 0 1-1.5 1.5m-3-4A1.5 1.5 0 0 1 13 6.5 1.5 1.5 0 0 1 14.5 5 1.5 1.5 0 0 1 16 6.5 1.5 1.5 0 0 1 14.5 8m-5 0A1.5 1.5 0 0 1 8 6.5 1.5 1.5 0 0 1 9.5 5 1.5 1.5 0 0 1 11 6.5 1.5 1.5 0 0 1 9.5 8m-3 4A1.5 1.5 0 0 1 5 10.5 1.5 1.5 0 0 1 6.5 9 1.5 1.5 0 0 1 8 10.5 1.5 1.5 0 0 1 6.5 12M12 3a9 9 0 0 0-9 9 9 9 0 0 0 9 9 1.5 1.5 0 0 0 1.5-1.5c0-.4-.2-.7-.4-1-.2-.3-.4-.6-.4-1a1.5 1.5 0 0 1 1.5-1.5H16a5 5 0 0 0 5-5c0-4.4-4-8-9-8Z"/></svg>',
      label: "Styles & Props",
      layoutComponent: { type: "panelSidebarTabs" }
    }
  ], f = {
    id: E,
    type: "column",
    className: "gs-cmp-layout-sidebar-target",
    style: { overflow: "hidden" }
  }, r = {
    type: "sidebarLeft",
    resizable: !1,
    style: { padding: "10px 5px", alignItems: "center", width: 45, gap: 10 },
    children: []
  }, d = {
    type: "canvasSidebarTop",
    sidebarTop: {
      leftContainer: {
        buttons: []
      }
    }
  }, b = (n) => ({
    breakpoint: n,
    breakpointTablet: a,
    breakpointMobile: o
  }), u = (n, g, s) => {
    const l = b(g), y = { pluginOpts: e, breakpointProps: l }, p = n.map((P) => {
      const R = w(P, y);
      return e != null && e.sidebarButton ? e.sidebarButton({
        id: P.id,
        buttonIds: v,
        buttonProps: R,
        sidebarButtonProps: P,
        ...l,
        createSidebarButton: (k) => w({ ...P, ...k }, y)
      }) : R;
    }).filter(Boolean);
    return (e != null && e.sidebarButtons ? e == null ? void 0 : e.sidebarButtons({
      buttonIds: v,
      sidebarButtons: p,
      ...l,
      createSidebarButton: (P) => w(
        {
          ...P,
          layoutCommand: {
            ...s == null ? void 0 : s.layoutCommand,
            ...P.layoutCommand
          }
        },
        y
      )
    }) : p).filter(Boolean);
  }, t = (n, g, s) => {
    var p;
    const l = b(s), y = { pluginOpts: e, breakpointProps: l };
    return ((p = e == null ? void 0 : e.rootLayout) == null ? void 0 : p.call(e, {
      sidebarButtons: g,
      rootLayout: n,
      layoutSidebarLeft: r,
      layoutSidebarTarget: f,
      ...l,
      createSidebarButton: (A) => w({ ...A }, y)
    })) ?? n;
  };
  if (a) {
    const n = u(h, a);
    c[a] = t(
      {
        type: "row",
        style: { height: "100%" },
        children: [
          {
            ...r,
            children: n
          },
          f,
          d
        ]
      },
      n,
      a
    );
  }
  if (o) {
    const n = {
      placer: { type: "absolute", position: "left" },
      style: { height: "calc(100% - 40px)" }
    }, g = h.map((l) => ({ ...l, layoutCommand: n })), s = u(g, o, { layoutCommand: n });
    c[o] = t(
      {
        type: "column",
        style: { height: "100%" },
        children: [
          { type: "sidebarTop", leftContainer: { buttons: [] } },
          { type: "canvas", grow: !0 },
          {
            type: "sidebarBottom",
            style: { padding: "0 5px", alignItems: "center", gap: 10, minHeight: 39 },
            children: s
          }
        ]
      },
      s,
      o
    );
  }
  const L = u(i, 0);
  return {
    default: t(
      {
        type: "row",
        style: { height: "100%" },
        children: [
          {
            ...r,
            children: L
          },
          f,
          d,
          { type: "sidebarRight" }
        ]
      },
      L,
      0
    ),
    responsive: c
  };
}, M = "layoutSidebarButtons", z = B.free, J = function(e, a = {}) {
  const { Commands: o } = e, c = o.events, i = {
    ...a
  };
  (() => {
    if (e.runCommand(T.layoutConfigGet) || i.skipLayoutConfig === !0) return;
    const r = N(i);
    e.runCommand(T.layoutConfigSet, { ...r });
  })(), o.add(`${M}:toggleButton`, (f, r, d) => {
    const b = document.getElementById(U(d == null ? void 0 : d.id));
    b == null || b.click();
  }), e.on(`${c.runCommand}core:preview`, () => {
    $(e);
  }), K({
    editor: e,
    licenseKey: i.licenseKey,
    plan: z,
    pluginName: M,
    cleanup: () => {
      e.runCommand(T.layoutConfigSet);
    }
  });
}, Q = C(J);
Q.createLayoutConfig = (e) => N(e);
export {
  Q as default
};
