// @grapesjs/studio-sdk-plugins/dist/youtubeAssetProvider/index.es.js
var g = /* @__PURE__ */ ((e) => (e.free = "free", e.startup = "startup", e.business = "business", e.enterprise = "enterprise", e))(g || {}), b = /* @__PURE__ */ ((e) => (e.toastAdd = "studio:toastAdd", e.toastRemove = "studio:toastRemove", e.dialogOpen = "studio:dialogOpen", e.dialogClose = "studio:dialogClose", e.dialogExportCode = "studio:dialogExportCode", e.dialogImportCode = "studio:dialogImportCode", e.sidebarLeftSet = "studio:sidebarLeft:set", e.sidebarLeftGet = "studio:sidebarLeft:get", e.sidebarLeftToggle = "studio:sidebarLeft:toggle", e.sidebarRightSet = "studio:sidebarRight:set", e.sidebarRightGet = "studio:sidebarRight:get", e.sidebarRightToggle = "studio:sidebarRight:toggle", e.sidebarTopSet = "studio:sidebarTop:set", e.sidebarTopGet = "studio:sidebarTop:get", e.sidebarTopToggle = "studio:sidebarTop:toggle", e.sidebarBottomSet = "studio:sidebarBottom:set", e.sidebarBottomGet = "studio:sidebarBottom:get", e.sidebarBottomToggle = "studio:sidebarBottom:toggle", e.symbolAdd = "studio:symbolAdd", e.symbolDetach = "studio:symbolDetach", e.symbolOverride = "studio:symbolOverride", e.symbolPropagateStyles = "studio:propagateStyles", e.getPagesConfig = "studio:getPagesConfig", e.setPagesConfig = "studio:setPagesConfig", e.getPageSettings = "studio:getPageSettings", e.setPageSettings = "studio:setPageSettings", e.clearPage = "studio:clearPage", e.projectFiles = "studio:projectFiles", e.validateCode = "studio:validateCode", e.canvasReload = "studio:canvasReload", e.getBlocksPanel = "studio:getBlocksPanel", e.setBlocksPanel = "studio:setBlocksPanel", e.getStateContextMenu = "studio:getStateContextMenu", e.setStateContextMenu = "studio:setStateContextMenu", e.contextMenuComponent = "studio:contextMenuComponent", e.layoutAdd = "studio:layoutAdd", e.layoutRemove = "studio:layoutRemove", e.layoutToggle = "studio:layoutToggle", e.layoutUpdate = "studio:layoutUpdate", e.layoutGet = "studio:layoutGet", e.layoutConfigGet = "studio:layoutConfigGet", e.layoutConfigSet = "studio:layoutConfigSet", e.layoutComponentAdd = "studio:layoutComponentAdd", e.layoutComponentGet = "studio:layoutComponentGet", e.layoutComponentRemove = "studio:layoutComponentRemove", e.layoutComponentRender = "studio:layoutComponentRender", e.getStateTheme = "studio:getStateTheme", e.setStateTheme = "studio:setStateTheme", e.settings = "studio:settings", e.assetProviderGet = "studio:assetProviderGet", e.assetProviderAdd = "studio:assetProviderAdd", e.assetProviderRemove = "studio:assetProviderRemove", e.fontGet = "studio:fontGet", e.fontAdd = "studio:fontAdd", e.fontRemove = "studio:fontRemove", e.fontManagerOpen = "studio:fontManagerOpen", e.menuFontLoad = "studio:menuFontLoad", e.toggleStateDataSource = "studio:toggleStateDataSource", e.getStateDataSource = "studio:getStateDataSource", e.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", e.dataSourceSetImporter = "studio:dataSourceSetImporter", e.dataSourceSetExporter = "studio:dataSourceSetExporter", e.setDragAbsolute = "studio:setDragAbsolute", e))(b || {});
const w = "app.grapesjs.com", R = "app-stage.grapesjs.com", x = "app2.grapesjs.com", G = "app-stage2.grapesjs.com", P = [
  w,
  x,
  R,
  G,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], U = "license:check:end", D = () => typeof window < "u", k = ({
  isDev: e,
  isStage: o,
  isPlatform: i
}) => `${e ? "" : `https://${o ? R : w}`}/${i ? "platform-api" : "api"}`, O = () => {
  const e = D() && window.location.hostname;
  return !!e && (P.includes(e) || P.some((o) => e.endsWith(o)));
};
async function $({
  path: e,
  baseApiUrl: o,
  method: i = "GET",
  headers: n = {},
  params: l,
  body: u
}) {
  const t = `${o || k({ isDev: !1, isStage: !1 })}${e}`, r = {
    method: i,
    headers: {
      "Content-Type": "application/json",
      ...n
    }
  };
  u && (r.body = JSON.stringify(u));
  const c = l ? new URLSearchParams(l).toString() : "", a = c ? `?${c}` : "", s = await fetch(`${t}${a}`, r);
  if (!s.ok)
    throw new Error(`HTTP error! status: ${s.status}`);
  return s.json();
}
const T = {
  [g.free]: 0,
  [g.startup]: 10,
  [g.business]: 20,
  [g.enterprise]: 30
};
function E(e) {
  const o = e;
  return o.init = (i) => (n) => e(n, i), o;
}
const M = (e) => /* @__PURE__ */ E(e);
async function j({
  editor: e,
  plan: o,
  pluginName: i,
  licenseKey: n,
  onLicenseCheckResponse: l,
  cleanup: u
}) {
  let p = "", t;
  const r = (s) => {
    console.warn("Cleanup plugin:", i, "Reason:", s), u();
  }, c = (s = {}) => {
    var v;
    const { error: h, sdkLicense: d } = s, y = (v = s.plan) == null ? void 0 : v.category;
    if (!(d || s.license) || h)
      r(h || "Invalid license");
    else if (y) {
      const I = T[o], L = T[y];
      I > L && r({ pluginRequiredPlan: o, licensePlan: y });
    }
  };
  e.Commands.has(b.settings) && (t = e.runCommand(b.settings), p = (t == null ? void 0 : t.baseUrl) || "");
  const a = (s) => {
    l == null || l(s), s && c(s);
  };
  if (!t) {
    e.onReady(async () => {
      if (!O())
        if (n) {
          const s = await N({ licenseKey: n, pluginName: i, baseApiUrl: p });
          a(s);
        } else
          r("The `licenseKey` option not provided");
    });
    return;
  }
  if (t.licensePlan || t.licenseError) {
    const s = B(t);
    a(s);
    return;
  }
  e.on(U, (s) => a(s));
}
const B = (e) => ({
  sdkLicense: e.license,
  license: e.license,
  error: e.licenseError,
  plan: e.licensePlan
});
async function N(e) {
  const { licenseKey: o, pluginName: i, baseApiUrl: n } = e;
  try {
    return (await $({
      baseApiUrl: n,
      path: `/sdk/${o || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: i
      }
    })).result || {};
  } catch (l) {
    return console.error("Error during SDK license check:", l), !1;
  }
}
const Y = (e) => {
  const o = {};
  for (const i of Object.keys(e)) {
    let n = e[i];
    n && (Array.isArray(n) ? n = n.map((l) => l.toString()).join(",") : n = n.toString(), o[i] = n);
  }
  return new URLSearchParams(o);
}, A = "youtube-video", W = (e) => {
  const { apiKey: o, searchParams: i, thumbnailQuality: n = "high" } = e, l = (t = {}) => {
    const c = "https://www.googleapis.com/youtube/v3/search", a = Y(t);
    return `${c}?${a}`;
  }, u = (t) => {
    var r, c, a;
    return ((r = t[n]) == null ? void 0 : r.url) || ((c = t.high) == null ? void 0 : c.url) || ((a = t.medium) == null ? void 0 : a.url) || t.default.url;
  }, p = (t) => {
    const r = u(t.snippet.thumbnails);
    return {
      id: t.id.videoId,
      src: `https://www.youtube.com/watch?v=${t.id.videoId}`,
      name: t.snippet.title,
      type: A,
      customData: { thumbnailUrl: r, videoItem: t }
    };
  };
  return {
    id: "youtube",
    types: [A],
    label: ({ editor: t }) => t.I18n.t("youtubeAssetProvider.providerLabel"),
    search: { reloadOnInput: !0, debounceMs: 1e3 },
    async onLoad({ searchValue: t, pageCustomData: r }) {
      var f;
      const c = {
        maxResults: 25,
        q: t,
        pageToken: r == null ? void 0 : r.token,
        type: ["video"],
        key: o
      }, a = i == null ? void 0 : i({ searchValue: t }), s = {
        ...c,
        ...a
      };
      s.part ? s.part.includes("snippet") || s.part.push("snippet") : s.part = ["snippet"];
      const d = await (await fetch(l(s))).json();
      return (f = d.items) != null && f.length ? {
        items: d.items.map(p),
        nextPageCustomData: { token: d.nextPageToken },
        isLastPage: !d.nextPageToken
      } : { items: [], isLastPage: !0 };
    },
    itemLayout: ({ assetProps: t, onSelect: r }) => ({
      type: "column",
      onClick: () => r(t),
      htmlAttrs: {
        title: t.name
      },
      style: {
        borderWidth: "1px",
        borderRadius: 8,
        overflow: "hidden",
        cursor: "pointer"
      },
      children: [
        {
          type: "custom",
          render: () => {
            var c;
            return `<img src="${(c = t.customData) == null ? void 0 : c.thumbnailUrl}" alt="${t.name}" style="width: 100%; height: 100px; object-fit: cover">`;
          }
        },
        {
          type: "text",
          content: t.name,
          style: {
            overflow: "hidden",
            textOverflow: "ellipsis",
            whiteSpace: "nowrap",
            paddingLeft: 8,
            paddingRight: 8
          }
        }
      ]
    })
  };
}, _ = {
  youtubeAssetProvider: {
    searchBtn: "Search on YouTube",
    modalTitle: "Select YouTube Video",
    providerLabel: "YouTube Videos"
  }
}, F = "youtubeAssetProvider", K = g.startup, q = function(e, o) {
  var c;
  const { skipVideoComponent: i, i18n: n = {} } = o, l = W(o);
  e.runCommand(b.assetProviderAdd, { provider: l });
  const u = "search-on-youtube", p = "video", t = (c = e.Components.getType(p)) == null ? void 0 : c.model.prototype, r = (a) => {
    e.Assets.open({
      modalTitle: e.I18n.t("youtubeAssetProvider.modalTitle"),
      providerId: l.id,
      typeId: l.types[0],
      providers: [],
      types: [],
      select: (s) => {
        e.Assets.close(), a.set({ videoId: s.attributes.id });
      },
      content: {
        header: {
          addUrl: !1,
          upload: !1
        }
      }
    });
  };
  !i && t && e.Components.addType(p, {
    model: {
      getYoutubeTraits() {
        const [a, ...s] = t.getYoutubeTraits.apply(this);
        return [
          a,
          {
            id: u,
            type: "button",
            label: e.I18n.t("youtubeAssetProvider.searchBtn"),
            command: () => r(this)
          },
          ...s
        ];
      }
    },
    view: {
      events: () => ({
        dblclick: "onActive"
      }),
      onActive() {
        const { model: a } = this, s = a.get("provider") || "";
        s != null && s.startsWith("yt") && r(a);
      }
    }
  }), e.I18n.addMessages({
    en: _,
    ...n
  }), j({
    editor: e,
    licenseKey: o.licenseKey,
    plan: K,
    pluginName: F,
    cleanup: () => {
      e.runCommand(b.assetProviderRemove, { id: l.id }), !i && t && e.Components.addType(p, {
        model: {
          getYoutubeTraits() {
            return t.getYoutubeTraits.apply(this).filter((a) => a.id !== u);
          }
        }
      });
    }
  });
}, V = M(q);
export {
  V as default
};
