// @grapesjs/studio-sdk-plugins/dist/swiperComponent/index.es.js
const te = "app.grapesjs.com", ae = "app-stage.grapesjs.com", le = "app2.grapesjs.com", se = "app-stage2.grapesjs.com", U = [
  te,
  le,
  ae,
  se,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], oe = "license:check:end", ie = () => typeof window < "u", ne = ({
  isDev: e,
  isStage: l,
  isPlatform: a
}) => `${e ? "" : `https://${l ? ae : te}`}/${a ? "platform-api" : "api"}`, re = () => {
  const e = ie() && window.location.hostname;
  return !!e && (U.includes(e) || U.some((l) => e.endsWith(l)));
}, ce = (e) => e.replace(/[A-Z]+(?![a-z])|[A-Z]/g, (l, a) => (a ? "-" : "") + l.toLowerCase()), de = (e) => typeof e < "u";
async function pe({
  path: e,
  baseApiUrl: l,
  method: a = "GET",
  headers: s = {},
  params: o,
  body: f
}) {
  const d = `${l || ne({ isDev: !1, isStage: !1 })}${e}`, p = {
    method: a,
    headers: {
      "Content-Type": "application/json",
      ...s
    }
  };
  f && (p.body = JSON.stringify(f));
  const b = o ? new URLSearchParams(o).toString() : "", g = b ? `?${b}` : "", i = await fetch(`${d}${g}`, p);
  if (!i.ok)
    throw new Error(`HTTP error! status: ${i.status}`);
  return i.json();
}
var x = /* @__PURE__ */ ((e) => (e.free = "free", e.startup = "startup", e.business = "business", e.enterprise = "enterprise", e))(x || {}), q = /* @__PURE__ */ ((e) => (e.toastAdd = "studio:toastAdd", e.toastRemove = "studio:toastRemove", e.dialogOpen = "studio:dialogOpen", e.dialogClose = "studio:dialogClose", e.dialogExportCode = "studio:dialogExportCode", e.dialogImportCode = "studio:dialogImportCode", e.sidebarLeftSet = "studio:sidebarLeft:set", e.sidebarLeftGet = "studio:sidebarLeft:get", e.sidebarLeftToggle = "studio:sidebarLeft:toggle", e.sidebarRightSet = "studio:sidebarRight:set", e.sidebarRightGet = "studio:sidebarRight:get", e.sidebarRightToggle = "studio:sidebarRight:toggle", e.sidebarTopSet = "studio:sidebarTop:set", e.sidebarTopGet = "studio:sidebarTop:get", e.sidebarTopToggle = "studio:sidebarTop:toggle", e.sidebarBottomSet = "studio:sidebarBottom:set", e.sidebarBottomGet = "studio:sidebarBottom:get", e.sidebarBottomToggle = "studio:sidebarBottom:toggle", e.symbolAdd = "studio:symbolAdd", e.symbolDetach = "studio:symbolDetach", e.symbolOverride = "studio:symbolOverride", e.symbolPropagateStyles = "studio:propagateStyles", e.getPagesConfig = "studio:getPagesConfig", e.setPagesConfig = "studio:setPagesConfig", e.getPageSettings = "studio:getPageSettings", e.setPageSettings = "studio:setPageSettings", e.clearPage = "studio:clearPage", e.projectFiles = "studio:projectFiles", e.validateCode = "studio:validateCode", e.canvasReload = "studio:canvasReload", e.getBlocksPanel = "studio:getBlocksPanel", e.setBlocksPanel = "studio:setBlocksPanel", e.getStateContextMenu = "studio:getStateContextMenu", e.setStateContextMenu = "studio:setStateContextMenu", e.contextMenuComponent = "studio:contextMenuComponent", e.layoutAdd = "studio:layoutAdd", e.layoutRemove = "studio:layoutRemove", e.layoutToggle = "studio:layoutToggle", e.layoutUpdate = "studio:layoutUpdate", e.layoutGet = "studio:layoutGet", e.layoutConfigGet = "studio:layoutConfigGet", e.layoutConfigSet = "studio:layoutConfigSet", e.layoutComponentAdd = "studio:layoutComponentAdd", e.layoutComponentGet = "studio:layoutComponentGet", e.layoutComponentRemove = "studio:layoutComponentRemove", e.layoutComponentRender = "studio:layoutComponentRender", e.getStateTheme = "studio:getStateTheme", e.setStateTheme = "studio:setStateTheme", e.settings = "studio:settings", e.assetProviderGet = "studio:assetProviderGet", e.assetProviderAdd = "studio:assetProviderAdd", e.assetProviderRemove = "studio:assetProviderRemove", e.fontGet = "studio:fontGet", e.fontAdd = "studio:fontAdd", e.fontRemove = "studio:fontRemove", e.fontManagerOpen = "studio:fontManagerOpen", e.menuFontLoad = "studio:menuFontLoad", e.toggleStateDataSource = "studio:toggleStateDataSource", e.getStateDataSource = "studio:getStateDataSource", e.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", e.dataSourceSetImporter = "studio:dataSourceSetImporter", e.dataSourceSetExporter = "studio:dataSourceSetExporter", e.setDragAbsolute = "studio:setDragAbsolute", e))(q || {});
const j = {
  [x.free]: 0,
  [x.startup]: 10,
  [x.business]: 20,
  [x.enterprise]: 30
};
function ue(e) {
  const l = e;
  return l.init = (a) => (s) => e(s, a), l;
}
const be = (e) => /* @__PURE__ */ ue(e);
async function fe({
  editor: e,
  plan: l,
  pluginName: a,
  licenseKey: s,
  onLicenseCheckResponse: o,
  cleanup: f
}) {
  let c = "", d;
  const p = (i) => {
    console.warn("Cleanup plugin:", a, "Reason:", i), f();
  }, b = (i = {}) => {
    var _;
    const { error: t, sdkLicense: n } = i, w = (_ = i.plan) == null ? void 0 : _.category;
    if (!(n || i.license) || t)
      p(t || "Invalid license");
    else if (w) {
      const B = j[l], $ = j[w];
      B > $ && p({ pluginRequiredPlan: l, licensePlan: w });
    }
  };
  e.Commands.has(q.settings) && (d = e.runCommand(q.settings), c = (d == null ? void 0 : d.baseUrl) || "");
  const g = (i) => {
    o == null || o(i), i && b(i);
  };
  if (!d) {
    e.onReady(async () => {
      if (!re())
        if (s) {
          const i = await he({ licenseKey: s, pluginName: a, baseApiUrl: c });
          g(i);
        } else
          p("The `licenseKey` option not provided");
    });
    return;
  }
  if (d.licensePlan || d.licenseError) {
    const i = ge(d);
    g(i);
    return;
  }
  e.on(oe, (i) => g(i));
}
const ge = (e) => ({
  sdkLicense: e.license,
  license: e.license,
  error: e.licenseError,
  plan: e.licensePlan
});
async function he(e) {
  const { licenseKey: l, pluginName: a, baseApiUrl: s } = e;
  try {
    return (await pe({
      baseApiUrl: s,
      path: `/sdk/${l || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: a
      }
    })).result || {};
  } catch (o) {
    return console.error("Error during SDK license check:", o), !1;
  }
}
const E = (e, l) => {
  var a;
  return !!((a = e == null ? void 0 : e.classList) != null && a.contains(l));
}, ve = (e, l = {}) => {
  var f;
  const a = {}, s = [], o = [];
  for (const c in e) {
    const { value: d, ...p } = e[c];
    s.push({ id: c, name: c, changeProp: !0, category: l.category, ...p }), a[c] = d, !de(d) && o.push(c);
    const { subTraits: b = {} } = p;
    for (const g in p.subTraits)
      (f = b[g]) == null || f.forEach((i) => {
        const t = i.id || i.name;
        a[t] = i.value;
      });
  }
  return {
    propKeys: Object.keys(a),
    propsNoValues: o,
    props: a,
    traits: s
  };
}, ye = (e, l) => {
  l.forEach((a) => {
    if (a.subTraits) {
      const s = `${a.id || a.name}`, o = a.changeProp ? `change:${s}` : `change:attributes:${s}`;
      e.on(o, () => Z(e, s)), Z(e, s);
    }
  });
}, Z = (e, l) => {
  const a = e.getTrait(l), s = a.get("category"), o = a.get("subTraits") || {}, c = (Object.values(o).flat() || []).map((b) => b.id || b.name);
  e.removeTrait(c);
  const d = o[a.getValue()], p = e.getTraitIndex(l) + 1;
  d && e.addTrait(
    d.map((b) => ({ category: s, ...b })),
    { at: p }
  );
}, we = (e, l) => {
  l.forEach((a) => {
    const s = ce(a);
    s !== a && e.has(s) && e.set({ [a]: e.get(s), [s]: void 0 });
  });
}, v = "swiper", O = `${v}-slide`, L = `${v}-wrapper`, A = `${v}-nav-prev`, I = `${v}-nav-next`, k = `${v}-pagination`, M = `${v}-scrollbar`, u = {
  slider: `<svg viewBox="0 0 24 24">
  <path d="M22 7.6c0-1-.5-1.6-1.3-1.6H3.4C2.5 6 2 6.7 2 7.6v9.8c0 1 .5 1.6 1.3 1.6h17.4c.8 0 1.3-.6 1.3-1.6V7.6zM21 18H3V7h18v11z" fill-rule="nonzero"/>
  <path d="M4 12.5L6 14v-3zM20 12.5L18 14v-3z"/>
</svg>`,
  effect: `<svg viewBox="0 0 24 24">
  <path d="M7.5 5.6 5 7l1.4-2.5L5 2l2.5 1.4L10 2 8.6 4.5 10 7 7.5 5.6m12 9.8L22 14l-1.4 2.5L22 19l-2.5-1.4L17 19l1.4-2.5L17 14l2.5 1.4M22 2l-1.4 2.5L22 7l-2.5-1.4L17 7l1.4-2.5L17 2l2.5 1.4L22 2m-8.7 10.8 2.5-2.5-2.1-2-2.5 2.4 2.1 2m1-5.4 2.4 2.3c.4.4.4 1 0 1.4L5 22.7a1 1 0 0 1-1.4 0l-2.3-2.3a1 1 0 0 1 0-1.4L13 7.3a1 1 0 0 1 1.4 0Z"/>
</svg>`,
  autoplay: `<svg viewBox="0 0 24 24">
  <path d="M5 6.3A1 1 0 0 1 5 5a1 1 0 0 1 1.3 0l6.8 5.4.3.3a2 2 0 0 1-2.8 2.8l-.3-.3L5 6.3M12 20a8 8 0 0 0 8-8 8 8 0 0 0-2.3-5.7L19 5a10 10 0 0 1 3 7 10 10 0 0 1-10 10A10 10 0 0 1 2 12h2a8 8 0 0 0 8 8m0-19a2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2 2 2 0 0 1 2-2Z"/>
</svg>`,
  pagination: `<svg viewBox="0 0 24 24">
  <path d="M16 12a2 2 0 0 1 2-2 2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2m-6 0a2 2 0 0 1 2-2 2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2m-6 0a2 2 0 0 1 2-2 2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2Z"/>
</svg>`,
  scrollbar: `<svg viewBox="0 0 24 24">
  <path d="M7.2 9.2 5.8 7.8 1.5 12l4.3 4.2 1.4-1.4L4.4 12zm9.6 5.6 1.4 1.4 4.3-4.2-4.3-4.2-1.4 1.4 2.8 2.8z"/>
  <path fill-rule="evenodd" d="M15 12a3 3 0 1 1-6 0 3 3 0 0 1 6 0m-2 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0" clip-rule="evenodd"/>
</svg>`,
  parallax: `<svg viewBox="0 0 24 24">
  <path d="M12.6 23.3h-.2v.5h.2v-.5m.3-.2-.2.1v.5l.2.1v-.6m-.8 0v.6h.2v-.5z"/>
  <path fill="currentColor" d="M18 9.4a2.5 2.5 0 0 1 3 2.3v5.7a3 3 0 0 1-2.2 2.9h-.2L13 21.4a2.5 2.5 0 0 1-3-2.3v-5.7a3 3 0 0 1 2.2-2.9h.2zM14.5 6a2.5 2.5 0 0 1 3 2.3v.4l-5.3 1A4 4 0 0 0 9 13.4v4.8a2.5 2.5 0 0 1-2.5-2.4v-5.7a3 3 0 0 1 2.2-2.9H9zM11 2.6a2.5 2.5 0 0 1 3 2.3v.3l-5.3 1A4 4 0 0 0 5.5 10v4.8A2.5 2.5 0 0 1 3 12.3V6.6a3 3 0 0 1 2.2-2.9h.2z"/>
</svg>`,
  responsive: `<svg viewBox="0 0 24 24">
  <path d="M17 21q-.4 0-.7-.3T16 20V10q0-.8-.6-1.4T14 8h-3.5q-.4 0-.7-.3T9.5 7V4q0-.4.3-.7t.7-.3H20q.4 0 .7.3t.3.7v16q0 .4-.3.7t-.7.3zm-6.5 0q-.4 0-.7-.3t-.3-.7v-9q0-.4.3-.7t.7-.3H13q.4 0 .7.3t.3.7v9q0 .4-.3.7t-.7.3zM4 21q-.4 0-.7-.3T3 20v-9q0-.4.3-.7T4 10h2.5q.4 0 .7.3t.3.7v9q0 .4-.3.7t-.7.3z"/>
</svg>`,
  extra: `<svg viewBox="0 0 24 24">
  <path fill-rule="evenodd" d="M5.4 3h13.2A2.4 2.4 0 0 1 21 5.4v13.2a2.4 2.4 0 0 1-2.4 2.4H5.4A2.4 2.4 0 0 1 3 18.6V5.4A2.4 2.4 0 0 1 5.4 3M5 12a2 2 0 1 1 4 0 2 2 0 0 1-4 0m5 0a2 2 0 1 1 4 0 2 2 0 0 1-4 0m7-2a2 2 0 1 0 0 4 2 2 0 0 0 0-4" clip-rule="evenodd"/>
</svg>`,
  reload: `<svg viewBox="0 0 24 24">
  <path d="M2 12c0 5 4 9 9 9 2.4 0 4.7-1 6.4-2.6l-1.5-1.5A6.7 6.7 0 0 1 11 19a7 7 0 1 1 7-7h-3l4 4h.1l3.9-4h-3a9 9 0 0 0-18 0Z"/>
</svg>`,
  prev: '<svg viewBox="0 0 24 24"><path d="M15.4 16.6 10.8 12l4.6-4.6L14 6l-6 6 6 6 1.4-1.4Z"/></svg>',
  next: '<svg viewBox="0 0 24 24"><path d="m8.6 16.6 4.6-4.6-4.6-4.6L10 6l6 6-6 6-1.4-1.4Z"/></svg>',
  add: `<svg viewBox="0 0 24 24">
  <path d="M12 20a8 8 0 1 1 0-16 8 8 0 0 1 0 16m0-18A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10A10 10 0 0 0 12 2m1 5h-2v4H7v2h4v4h2v-4h4v-2h-4V7Z"/>
</svg>`
}, K = { id: "swiper-responsive", label: "Responsive", icon: u.responsive }, Y = { id: "swiper-extra", label: "Extra", icon: u.extra }, me = {
  actions: {
    type: "radio",
    label: !1,
    labelToTitle: !0,
    options: [
      { id: "r", label: "Reload slider", icon: u.reload },
      { id: "p", label: "Previous slide", icon: u.prev },
      { id: "n", label: "Next slide", icon: u.next },
      { id: "a", label: "Add new slide", icon: u.add }
    ],
    setValue: ({ component: e, value: l }) => {
      var s;
      const a = (s = e.getView()) == null ? void 0 : s.getSwiper();
      switch (l) {
        case "r":
          e.reloadSlider();
          break;
        case "p":
          a == null || a.slidePrev();
          break;
        case "n":
          a == null || a.slideNext();
          break;
        case "a":
          e.addSlide();
          break;
      }
    }
  },
  vertical: {
    type: "checkbox",
    label: "Vertical",
    value: !1
  },
  loop: {
    type: "checkbox",
    label: "Loop",
    value: !1
  },
  freeMode: {
    type: "checkbox",
    label: "Free mode",
    value: !1
  },
  autoHeight: {
    type: "checkbox",
    label: "Auto height",
    value: !1
  },
  navigation: {
    type: "checkbox",
    label: "Navigation",
    getValue: ({ component: e }) => !!e.findFirstType(A),
    setValue: ({ component: e, value: l }) => {
      var a, s;
      l ? e.append([{ type: A }, { type: I }]) : ((a = e.findFirstType(A)) == null || a.remove(), (s = e.findFirstType(I)) == null || s.remove()), e.set({ navigation: l });
    }
  },
  initialSlide: {
    type: "number",
    label: "Initial slide",
    min: 0,
    value: "0"
  },
  speed: {
    type: "number",
    label: "Speed",
    tip: "Transition duration (in ms)",
    step: 500,
    min: 0,
    value: 300
  },
  effect: {
    category: { id: "swiper-effect", label: "Effects", icon: u.effect },
    type: "select",
    label: "Effect Type",
    value: "slide",
    options: [
      { id: "slide", label: "Slide" },
      { id: "fade", label: "Fade" },
      { id: "cube", label: "Cube" },
      { id: "coverflow", label: "Coverflow" },
      { id: "flip", label: "Flip" },
      { id: "cards", label: "Cards" }
    ],
    subTraits: {
      slide: [
        {
          type: "number",
          name: "slidesPerView",
          label: "Slides per view",
          changeProp: !0,
          min: 1,
          value: 1
        },
        {
          type: "number",
          name: "slidesPerGroup",
          label: "Slides per group",
          tip: "Numbers of slides for group sliding",
          changeProp: !0,
          min: 1,
          value: 1
        },
        {
          type: "number",
          name: "spaceBetween",
          label: "Space between",
          tip: "Distance between slides in px",
          changeProp: !0,
          min: 0,
          value: 0
        },
        {
          type: "number",
          name: "slidesOffsetBefore",
          label: "Slides offset before",
          tip: "Add additional slide offset in the beginning of the container (in px, before all slides)",
          changeProp: !0,
          min: 0,
          value: 0
        },
        {
          type: "number",
          name: "slidesOffsetAfter",
          label: "Slides offset after",
          tip: "Add additional slide offset in the end of the container (in px, after all slides)",
          changeProp: !0,
          min: 0,
          value: 0
        },
        {
          type: "checkbox",
          name: "centeredSlides",
          label: "Centered slides",
          tip: "If enabled, the active slide will be centered",
          changeProp: !0,
          value: !1
        }
      ],
      fade: [
        {
          type: "checkbox",
          name: "crossFade",
          tip: "Enables slides cross fade",
          label: "Cross fade",
          changeProp: !0,
          value: !1
        }
      ],
      cube: [
        {
          type: "checkbox",
          name: "cubeEffectShadow",
          label: "Shadow",
          changeProp: !0,
          value: !0
        },
        {
          type: "checkbox",
          name: "cubeEffectSlideShadows",
          label: "Slide shadows",
          changeProp: !0,
          value: !0
        },
        {
          type: "number",
          name: "cubeEffectShadowOffset",
          label: "Shadow offset",
          changeProp: !0,
          value: 20
        },
        {
          type: "number",
          name: "cubeEffectShadowScale",
          label: "Shadow scale",
          changeProp: !0,
          value: 0.94
        }
      ],
      coverflow: [
        {
          type: "number",
          name: "coverflowEffectDepth",
          tip: "Depth offset in px (slides translate in Z axis)",
          label: "Depth",
          changeProp: !0,
          min: 0,
          value: 100
        },
        {
          type: "number",
          name: "coverflowEffectModifier",
          tip: "Effect multiplier",
          label: "Modifier",
          changeProp: !0,
          step: 0.1,
          min: 0,
          value: 1
        },
        {
          type: "number",
          name: "coverflowEffectRotate",
          tip: "Slide rotate in degrees",
          label: "Rotate",
          changeProp: !0,
          min: 0,
          value: 50
        },
        {
          type: "number",
          name: "coverflowEffectScale",
          tip: "Slide scale effect",
          label: "Scale",
          changeProp: !0,
          min: 0,
          step: 0.1,
          value: 1
        },
        {
          type: "number",
          name: "coverflowEffectStretch",
          tip: "Stretch space between slides (in px)",
          label: "Stretch",
          changeProp: !0,
          min: 0,
          value: 0
        },
        {
          type: "checkbox",
          name: "coverflowEffectSlideShadows",
          tip: "Enables slides shadows",
          label: "Slide shadow",
          changeProp: !0,
          value: !0
        }
      ],
      flip: [
        {
          type: "checkbox",
          name: "flipEffectLimitRotation",
          tip: "Limit edge slides rotation",
          label: "Limit rotation",
          changeProp: !0,
          value: !0
        },
        {
          type: "checkbox",
          name: "flipEffectSlideShadows",
          tip: "Enables slides shadows",
          label: "Slide shadows",
          changeProp: !0,
          value: !0
        }
      ],
      cards: [
        {
          type: "number",
          name: "cardsEffectPerSlideOffset",
          tip: "Offset distance per slide (in px)",
          label: "Slide offset",
          changeProp: !0,
          min: 0,
          value: 8
        },
        {
          type: "number",
          name: "cardsEffectPerSlideRotate",
          tip: "Rotate angle per slide (in degrees)",
          label: "Slide rotate",
          changeProp: !0,
          min: 0,
          value: 2
        },
        {
          type: "checkbox",
          name: "cardsEffectRotate",
          tip: "Enables cards rotation",
          label: "Rotate",
          changeProp: !0,
          value: !0
        },
        {
          type: "checkbox",
          name: "cardsEffectSlideShadows",
          tip: "Enables slides shadows",
          label: "Slide shadows",
          changeProp: !0,
          value: !0
        }
      ]
    }
  },
  autoplay: {
    category: { id: "swiper-autoplay", label: "Autoplay", icon: u.autoplay },
    type: "checkbox",
    label: "Enable Autoplay",
    value: !1,
    subTraits: {
      true: [
        {
          type: "checkbox",
          name: "autoplayEditor",
          label: "Autoplay in editor",
          tip: "Allow autoplay to run while editing in the canvas. Note: this may make selecting and editing slides difficult as they will automatically change.",
          changeProp: !0,
          value: !1
        },
        {
          type: "number",
          name: "autoplayDelay",
          label: "Delay",
          changeProp: !0,
          value: 3e3
        },
        {
          type: "checkbox",
          name: "autoplayDisableOnInteraction",
          label: "Disable on interaction",
          changeProp: !0,
          value: !0
        },
        {
          type: "checkbox",
          name: "autoplayWaitForTransition",
          label: "Wait for transition",
          changeProp: !0,
          value: !0
        },
        {
          type: "checkbox",
          name: "autoplayPauseOnMouseEnter",
          label: "Pause on mouse enter",
          changeProp: !0,
          value: !1
        },
        {
          type: "checkbox",
          name: "autoplayReverseDirection",
          label: "Reverse direction",
          changeProp: !0,
          value: !1
        },
        {
          type: "checkbox",
          name: "autoplayStopOnLastSlide",
          label: "Stop on last slide",
          changeProp: !0,
          value: !1
        }
      ]
    }
  },
  pagination: {
    category: { id: "swiper-pagination", label: "Pagination", icon: u.pagination },
    type: "select",
    label: "Pagination Type",
    options: [
      { id: "none", label: "None" },
      { id: "bullets", label: "Bullets" },
      { id: "progressbar", label: "Progressbar" },
      { id: "fraction", label: "Fraction" }
    ],
    getValue: ({ component: e }) => e.findFirstType(k) ? e.get("pagination") || "bullets" : "none",
    setValue: ({ component: e, value: l }) => {
      const a = e.findFirstType(k);
      a == null || a.remove(), l !== "none" && e.append({ type: k }), e.set({ pagination: l });
    },
    subTraits: {
      bullets: [
        {
          type: "checkbox",
          name: "dynamicBullets",
          label: "Dynamic bullets",
          changeProp: !0,
          value: !1
        },
        {
          type: "checkbox",
          name: "clickableBullets",
          label: "Clickable bullets",
          changeProp: !0,
          value: !0
        }
      ],
      progressbar: [
        {
          type: "checkbox",
          name: "progressbarOpposite",
          label: "Progress opposite",
          changeProp: !0,
          value: !1
        }
      ]
    }
  },
  scrollbar: {
    category: { id: "swiper-scrollbar", label: "Scrollbar", icon: u.scrollbar },
    type: "checkbox",
    label: "Enable Scrollbar",
    getValue: ({ component: e }) => !!e.findFirstType(M),
    setValue: ({ component: e, value: l }) => {
      const a = e.findFirstType(M);
      a == null || a.remove(), l && e.append({ type: M }), e.set({ scrollbar: l });
    },
    subTraits: {
      false: [],
      true: [
        {
          type: "checkbox",
          name: "scrollbarDraggable",
          label: "Draggable",
          changeProp: !0,
          value: !1
        },
        {
          type: "checkbox",
          name: "scrollbarHide",
          label: "Hide",
          tip: "Hide scrollbar automatically after user interaction",
          changeProp: !0,
          value: !0
        }
      ]
    }
  },
  parallax: {
    category: { id: "swiper-parallax", label: "Parallax", icon: u.parallax },
    type: "checkbox",
    label: "Enable Parallax",
    value: !1
  },
  mobile: {
    category: K,
    type: "checkbox",
    label: "Enable mobile breakpoint",
    value: !1,
    subTraits: {
      true: [
        {
          type: "number",
          name: "mobileBreakpoint",
          label: "Mobile breakpoint",
          tip: "Slide values below will be applied when the screen width is greater than this value",
          changeProp: !0,
          min: 1,
          value: 460
        },
        {
          type: "number",
          name: "mobileSlidesPerView",
          label: "Slides per view",
          changeProp: !0,
          min: 1,
          value: 1
        },
        {
          type: "number",
          name: "mobileSlidesPerGroup",
          label: "Slides per group",
          changeProp: !0,
          min: 1,
          value: 1
        },
        {
          type: "number",
          name: "mobileSpaceBetween",
          label: "Space between",
          changeProp: !0,
          min: 0,
          value: 0
        }
      ]
    }
  },
  tablet: {
    category: K,
    type: "checkbox",
    label: "Enable tablet breakpoint",
    value: !1,
    subTraits: {
      true: [
        {
          type: "number",
          name: "tabletBreakpoint",
          label: "Tablet breakpoint",
          tip: "Slide values below will be applied when the screen width is greater than this value",
          changeProp: !0,
          min: 1,
          value: 991
        },
        {
          type: "number",
          name: "tabletSlidesPerView",
          label: "Slides per view",
          changeProp: !0,
          min: 1,
          value: 1
        },
        {
          type: "number",
          name: "tabletSlidesPerGroup",
          label: "Slides per group",
          changeProp: !0,
          min: 1,
          value: 1
        },
        {
          type: "number",
          name: "tabletSpaceBetween",
          label: "Space between",
          changeProp: !0,
          min: 0,
          value: 0
        }
      ]
    }
  },
  allowTouchMove: {
    category: Y,
    type: "checkbox",
    label: "Allow touch move",
    tip: "Allow touch interactions",
    value: !0
  },
  grabCursor: {
    category: Y,
    type: "checkbox",
    label: "Grab cursor",
    tip: 'If enabled, the user will see the "grab" cursor when hover the slider',
    value: !1
  }
}, X = "Slider", z = "swiper", R = `${z}-`, Se = (e, l) => {
  const { Components: a, Blocks: s, Commands: o } = e, { block: f } = l, c = ve(me), d = "core:preview", p = o.events, b = function(t) {
    const n = this, w = window, N = n.__onLoad, _ = n.__activeSlide, B = n.__inPreview, $ = w._isEditor && !B ? t.autoplayEditor : !0, { cdnScript: G, cdnStyle: F, slidesPerView: C } = t || {}, r = (m) => isNaN(m) ? void 0 : parseFloat(m), H = () => {
      const m = n.parentElement;
      m && (m.style.minWidth = "0");
      const P = {
        loop: t.loop,
        speed: t.speed,
        initialSlide: t.initialSlide,
        direction: t.vertical ? "vertical" : "horizontal",
        effect: t.effect,
        freeMode: t.freeMode,
        parallax: t.parallax,
        autoHeight: t.autoHeight,
        allowTouchMove: t.allowTouchMove,
        grabCursor: t.grabCursor,
        simulateTouch: !0,
        breakpoints: {},
        slidesPerView: isNaN(C) ? C : r(C),
        slidesPerGroup: r(t.slidesPerGroup),
        spaceBetween: r(t.spaceBetween),
        slidesOffsetBefore: r(t.slidesOffsetBefore),
        slidesOffsetAfter: r(t.slidesOffsetAfter),
        centeredSlides: t.centeredSlides,
        fadeEffect: { crossFade: t.crossFade },
        cubeEffect: {
          shadow: t.cubeEffectShadow,
          slideShadows: t.cubeEffectSlideShadows,
          shadowOffset: r(t.cubeEffectShadowOffset),
          shadowScale: r(t.cubeEffectShadowScale)
        },
        coverflowEffect: {
          depth: r(t.coverflowEffectDepth),
          modifier: r(t.coverflowEffectModifier),
          rotate: r(t.coverflowEffectRotate),
          scale: r(t.coverflowEffectScale),
          stretch: r(t.coverflowEffectStretch),
          slideShadows: t.coverflowEffectSlideShadows
        },
        flipEffect: {
          limitRotation: t.flipEffectLimitRotation,
          slideShadows: t.flipEffectSlideShadows
        },
        cardsEffect: {
          perSlideOffset: r(t.cardsEffectPerSlideOffset),
          perSlideRotate: r(t.cardsEffectPerSlideRotate),
          rotate: t.cardsEffectRotate,
          slideShadows: t.cardsEffectSlideShadows
        },
        navigation: {
          nextEl: n.querySelector(".swiper-button-next"),
          prevEl: n.querySelector(".swiper-button-prev")
        },
        pagination: {
          type: t.pagination || "bullets",
          dynamicBullets: t.dynamicBullets,
          clickable: t.clickableBullets,
          progressbarOpposite: t.progressbarOpposite,
          el: n.querySelector(".swiper-pagination")
        },
        scrollbar: {
          el: n.querySelector(".swiper-scrollbar"),
          draggable: t.scrollbarDraggable,
          hide: t.scrollbarHide
        },
        autoplay: t.autoplay && $ && {
          delay: t.autoplayDelay,
          disableOnInteraction: t.autoplayDisableOnInteraction,
          pauseOnMouseEnter: t.autoplayPauseOnMouseEnter,
          reverseDirection: t.autoplayReverseDirection,
          stopOnLastSlide: t.autoplayStopOnLastSlide,
          waitForTransition: t.autoplayWaitForTransition
        }
      };
      if (t.tablet) {
        const { tabletBreakpoint: T, tabletSlidesPerView: y } = t, h = r(T);
        P.breakpoints[h] = {
          slidesPerView: isNaN(y) ? y : r(y),
          slidesPerGroup: r(t.tabletSlidesPerGroup),
          spaceBetween: r(t.tabletSpaceBetween)
        };
      }
      if (t.mobile) {
        const { mobileBreakpoint: T, mobileSlidesPerView: y } = t, h = r(T);
        P.breakpoints[h] = {
          slidesPerView: isNaN(y) ? y : r(y),
          slidesPerGroup: r(t.mobileSlidesPerGroup),
          spaceBetween: r(t.mobileSpaceBetween)
        };
      }
      N && !B && (P.simulateTouch = !1), _ && (P.initialSlide = _);
      const W = new w.Swiper(n, P);
      N == null || N(W);
    };
    if (w.Swiper)
      H();
    else {
      const { head: m } = document, P = Array.isArray(G) ? [...G] : [G], W = Array.isArray(F) ? [...F] : [F], T = (h) => {
        if (h.length) {
          const S = document.createElement("link");
          S.href = h.shift(), S.rel = "stylesheet", m.prepend(S), T(h);
        }
      }, y = (h) => {
        if (h.length) {
          const S = document.createElement("script");
          S.src = h.shift(), S.onerror = S.onload = y.bind(null, h), m.appendChild(S);
        } else
          H();
      };
      T(W), y(P);
    }
  }, g = (t = 0) => ({
    type: O,
    style: { padding: "50px" },
    components: { type: "text", components: { type: "textnode", content: `Slide ${t}` } }
  }), i = a.getType("default").model;
  return a.addType(v, {
    block: f && {
      label: X,
      media: u.slider,
      category: "Extra",
      select: !0,
      ...l.block
    },
    isComponent: (t) => E(t, z),
    model: {
      defaults: () => ({
        name: X,
        icon: u.slider,
        classes: z,
        draggable: (t, n) => !n.is(v),
        components: [
          {
            type: L,
            components: [g(1), g(2), g(3)]
          },
          { type: k }
        ],
        script: b,
        cdnScript: l.cdnScript,
        cdnStyle: l.cdnStyle,
        "script-props": ["cdnScript", "cdnStyle", ...c.propKeys],
        ...c.props,
        traits: [...c.traits]
      }),
      init() {
        const { em: t } = this;
        ye(this, c.traits), we(this, c.propKeys), this.listenTo(t, `${p.runCommand}${d}`, this.onPreviewOn), this.listenTo(t, `${p.stopCommand}${d}`, this.onPreviewOff);
      },
      onPreviewOn() {
        this.__inPreview = !0, this.reloadSlider();
      },
      onPreviewOff() {
        this.__inPreview = !1, this.reloadSlider();
      },
      reloadSlider() {
        this.trigger("rerender");
      },
      addSlide() {
        const t = this.findFirstType(L);
        t == null || t.append(this.getNewSlideComponentDef());
      },
      getNewSlideComponentDef() {
        var n;
        const t = ((n = this.findFirstType(L)) == null ? void 0 : n.components().length) || 0;
        return {
          type: O,
          components: { type: "text", components: { type: "textnode", content: `Slide ${t + 1}` } }
        };
      },
      toJSON(...t) {
        const n = i.prototype.toJSON.apply(this, t);
        return c.propsNoValues.forEach((w) => delete n[w]), delete n["script-props"], n;
      }
    },
    view: {
      init() {
        this.listenTo(this.model, a.events.scriptUnmount, this.destroySwiper);
      },
      getSwiper() {
        return this.el.swiper;
      },
      destroySwiper() {
        var t;
        (t = this.getSwiper()) == null || t.destroy();
      },
      onRender() {
        const t = this.el;
        t.__onLoad = (n) => n.on("slideChange", () => this.__activeSlide = n.activeIndex), t.__activeSlide = this.__activeSlide, t.__inPreview = this.model.__inPreview;
      }
    }
  }), () => {
    a.removeType(v), s.remove(v);
  };
}, Pe = "Nav Prev", Ee = "Nav Next", D = `${R}button-prev`, V = `${R}button-next`, Te = (e) => {
  const { Components: l } = e, a = (s) => ({
    type: "svg",
    selectable: !1,
    hoverable: !1,
    attributes: { viewBox: "0 0 27 44" },
    components: {
      type: "svg-in",
      tagName: "path",
      attributes: { d: s, fill: "currentColor" }
    }
  });
  return l.addType(A, {
    isComponent: (s) => E(s, D),
    model: {
      defaults: {
        name: Pe,
        classes: D,
        copyable: !1,
        highlightable: !1,
        droppable: !1,
        draggable: !1,
        selectable: !1,
        hoverable: !1,
        layerable: !1,
        components: a("M0,22L22,0l2.1,2.1L4.2,22l19.9,19.9L22,44L0,22L0,22L0,22z"),
        styles: `
          .${D} {
            color: inherit;
          }
          .${D}:after {
            content: none;
          }
        `
      }
    }
  }), l.addType(I, {
    extend: A,
    isComponent: (s) => E(s, V),
    model: {
      defaults: {
        name: Ee,
        classes: V,
        components: a("M27,22L27,22L5,44l-2.1-2.1L22.8,22L2.9,2.1L5,0L27,22L27,22z"),
        styles: `
          .${V} {
            color: inherit;
          }
          .${V}:after {
            content: none;
          }
        `
      }
    }
  }), () => {
    l.removeType(A), l.removeType(I);
  };
}, xe = (e) => {
  const { Components: l } = e, a = `${R}pagination`;
  return l.addType(k, {
    isComponent: (s) => E(s, a),
    model: {
      defaults: {
        name: "Pagination",
        classes: a,
        copyable: !1,
        highlightable: !1,
        droppable: !1,
        draggable: !1,
        selectable: !1,
        hoverable: !1,
        layerable: !1,
        styles: `
          .${a}-bullet-active {
            background-color: currentColor;
          }
          .${a}-progressbar-fill {
            --swiper-pagination-color: currentColor;
          }
        `
      }
    }
  }), () => {
    l.removeType(k);
  };
}, Le = "Scrollbar", J = `${R}scrollbar`, Ae = (e) => {
  const { Components: l } = e;
  return l.addType(M, {
    isComponent: (a) => E(a, J),
    model: {
      defaults: {
        name: Le,
        classes: J,
        copyable: !1,
        highlightable: !1,
        removable: !1,
        selectable: !1,
        hoverable: !1,
        layerable: !1,
        droppable: !1,
        draggable: !1
      }
    }
  }), () => {
    l.removeType(M);
  };
}, ke = "Slide", Q = `${R}slide`, Re = (e) => {
  const { Components: l } = e;
  return l.addType(O, {
    isComponent: (a) => E(a, Q),
    model: {
      defaults: {
        name: ke,
        classes: Q,
        icon: u.slider,
        draggable: (a, s) => s.is(L)
      }
      // TODO VIEW slide to this slide when selected
    }
  }), () => {
    l.removeType(O);
  };
}, Ne = "Wrapper", ee = `${R}wrapper`, _e = (e) => {
  const { Components: l } = e, { events: a } = l;
  return l.addType(L, {
    isComponent: (s) => E(s, ee),
    model: {
      defaults: {
        name: Ne,
        icon: u.slider,
        classes: ee,
        copyable: !1,
        removable: !1,
        draggable: !1,
        highlightable: !1,
        droppable: (s) => s.is(O)
      },
      init() {
        const s = this.components();
        this.listenTo(s, "add remove", this.updateSlider);
      },
      updateSlider(s) {
        const o = this.closestType(v);
        if (o) {
          const f = o.getView();
          o.reloadSlider(), this.em.once(a.scriptMount, () => {
            setTimeout(() => {
              var c;
              return (c = f.getSwiper()) == null ? void 0 : c.slideTo(s.index());
            }, 100);
          });
        }
      }
    }
  }), () => {
    l.removeType(L);
  };
}, Me = "swiperComponent", Oe = x.startup, Be = function(e, l = {}) {
  const a = {
    block: {},
    cdnScript: "https://cdn.jsdelivr.net/npm/swiper@11/swiper-bundle.min.js",
    cdnStyle: "https://cdn.jsdelivr.net/npm/swiper@11/swiper-bundle.min.css",
    ...l
  }, s = [
    Se(e, a),
    Te(e),
    _e(e),
    Re(e),
    xe(e),
    Ae(e)
  ];
  fe({
    editor: e,
    plan: Oe,
    licenseKey: a.licenseKey,
    pluginName: Me,
    cleanup: () => {
      s.forEach((o) => o());
    }
  });
}, De = be(Be);
export {
  De as default
};
