// @grapesjs/studio-sdk-plugins/dist/shapeDividerComponent/index.es.js
const L = "app.grapesjs.com", H = "app-stage.grapesjs.com", k = "app2.grapesjs.com", G = "app-stage2.grapesjs.com", z = [
  L,
  k,
  H,
  G,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], E = "license:check:end", I = () => typeof window < "u", O = ({
  isDev: e,
  isStage: a,
  isPlatform: i
}) => `${e ? "" : `https://${a ? H : L}`}/${i ? "platform-api" : "api"}`, F = () => {
  const e = I() && window.location.hostname;
  return !!e && (z.includes(e) || z.some((a) => e.endsWith(a)));
};
async function N({
  path: e,
  baseApiUrl: a,
  method: i = "GET",
  headers: c = {},
  params: n,
  body: r
}) {
  const o = `${a || O({ isDev: !1, isStage: !1 })}${e}`, t = {
    method: i,
    headers: {
      "Content-Type": "application/json",
      ...c
    }
  };
  r && (t.body = JSON.stringify(r));
  const g = n ? new URLSearchParams(n).toString() : "", h = g ? `?${g}` : "", l = await fetch(`${o}${h}`, t);
  if (!l.ok)
    throw new Error(`HTTP error! status: ${l.status}`);
  return l.json();
}
var y = /* @__PURE__ */ ((e) => (e.free = "free", e.startup = "startup", e.business = "business", e.enterprise = "enterprise", e))(y || {}), V = /* @__PURE__ */ ((e) => (e.toastAdd = "studio:toastAdd", e.toastRemove = "studio:toastRemove", e.dialogOpen = "studio:dialogOpen", e.dialogClose = "studio:dialogClose", e.dialogExportCode = "studio:dialogExportCode", e.dialogImportCode = "studio:dialogImportCode", e.sidebarLeftSet = "studio:sidebarLeft:set", e.sidebarLeftGet = "studio:sidebarLeft:get", e.sidebarLeftToggle = "studio:sidebarLeft:toggle", e.sidebarRightSet = "studio:sidebarRight:set", e.sidebarRightGet = "studio:sidebarRight:get", e.sidebarRightToggle = "studio:sidebarRight:toggle", e.sidebarTopSet = "studio:sidebarTop:set", e.sidebarTopGet = "studio:sidebarTop:get", e.sidebarTopToggle = "studio:sidebarTop:toggle", e.sidebarBottomSet = "studio:sidebarBottom:set", e.sidebarBottomGet = "studio:sidebarBottom:get", e.sidebarBottomToggle = "studio:sidebarBottom:toggle", e.symbolAdd = "studio:symbolAdd", e.symbolDetach = "studio:symbolDetach", e.symbolOverride = "studio:symbolOverride", e.symbolPropagateStyles = "studio:propagateStyles", e.getPagesConfig = "studio:getPagesConfig", e.setPagesConfig = "studio:setPagesConfig", e.getPageSettings = "studio:getPageSettings", e.setPageSettings = "studio:setPageSettings", e.clearPage = "studio:clearPage", e.projectFiles = "studio:projectFiles", e.validateCode = "studio:validateCode", e.canvasReload = "studio:canvasReload", e.getBlocksPanel = "studio:getBlocksPanel", e.setBlocksPanel = "studio:setBlocksPanel", e.getStateContextMenu = "studio:getStateContextMenu", e.setStateContextMenu = "studio:setStateContextMenu", e.contextMenuComponent = "studio:contextMenuComponent", e.layoutAdd = "studio:layoutAdd", e.layoutRemove = "studio:layoutRemove", e.layoutToggle = "studio:layoutToggle", e.layoutUpdate = "studio:layoutUpdate", e.layoutGet = "studio:layoutGet", e.layoutConfigGet = "studio:layoutConfigGet", e.layoutConfigSet = "studio:layoutConfigSet", e.layoutComponentAdd = "studio:layoutComponentAdd", e.layoutComponentGet = "studio:layoutComponentGet", e.layoutComponentRemove = "studio:layoutComponentRemove", e.layoutComponentRender = "studio:layoutComponentRender", e.getStateTheme = "studio:getStateTheme", e.setStateTheme = "studio:setStateTheme", e.settings = "studio:settings", e.assetProviderGet = "studio:assetProviderGet", e.assetProviderAdd = "studio:assetProviderAdd", e.assetProviderRemove = "studio:assetProviderRemove", e.fontGet = "studio:fontGet", e.fontAdd = "studio:fontAdd", e.fontRemove = "studio:fontRemove", e.fontManagerOpen = "studio:fontManagerOpen", e.menuFontLoad = "studio:menuFontLoad", e.toggleStateDataSource = "studio:toggleStateDataSource", e.getStateDataSource = "studio:getStateDataSource", e.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", e.dataSourceSetImporter = "studio:dataSourceSetImporter", e.dataSourceSetExporter = "studio:dataSourceSetExporter", e.setDragAbsolute = "studio:setDragAbsolute", e))(V || {});
const T = {
  [y.free]: 0,
  [y.startup]: 10,
  [y.business]: 20,
  [y.enterprise]: 30
};
function U(e) {
  const a = e;
  return a.init = (i) => (c) => e(c, i), a;
}
const C = (e) => /* @__PURE__ */ U(e);
async function W({
  editor: e,
  plan: a,
  pluginName: i,
  licenseKey: c,
  onLicenseCheckResponse: n,
  cleanup: r
}) {
  let p = "", o;
  const t = (l) => {
    console.warn("Cleanup plugin:", i, "Reason:", l), r();
  }, g = (l = {}) => {
    var w;
    const { error: A, sdkLicense: b } = l, u = (w = l.plan) == null ? void 0 : w.category;
    if (!(b || l.license) || A)
      t(A || "Invalid license");
    else if (u) {
      const R = T[a], x = T[u];
      R > x && t({ pluginRequiredPlan: a, licensePlan: u });
    }
  };
  e.Commands.has(V.settings) && (o = e.runCommand(V.settings), p = (o == null ? void 0 : o.baseUrl) || "");
  const h = (l) => {
    n == null || n(l), l && g(l);
  };
  if (!o) {
    e.onReady(async () => {
      if (!F())
        if (c) {
          const l = await j({ licenseKey: c, pluginName: i, baseApiUrl: p });
          h(l);
        } else
          t("The `licenseKey` option not provided");
    });
    return;
  }
  if (o.licensePlan || o.licenseError) {
    const l = _(o);
    h(l);
    return;
  }
  e.on(E, (l) => h(l));
}
const _ = (e) => ({
  sdkLicense: e.license,
  license: e.license,
  error: e.licenseError,
  plan: e.licensePlan
});
async function j(e) {
  const { licenseKey: a, pluginName: i, baseApiUrl: c } = e;
  try {
    return (await N({
      baseApiUrl: c,
      path: `/sdk/${a || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: i
      }
    })).result || {};
  } catch (n) {
    return console.error("Error during SDK license check:", n), !1;
  }
}
const Y = (e, a) => {
  var i;
  return !!((i = e == null ? void 0 : e.classList) != null && i.contains(a));
}, P = (e) => (a) => {
  var i;
  return ((i = a.getAttribute) == null ? void 0 : i.call(a, $)) === e;
}, B = "gjs-plg-", $ = "data-type-role", K = "shapeDividerComponent", X = y.startup, Z = function(e, a = {}) {
  const { Components: i, Blocks: c, Selectors: n } = e, r = "shape-divider", p = `${r}-cnt`, o = `${B}${r}`, t = `${B}${p}`, g = [
    {
      id: "tilt",
      label: "Tilt",
      value: `<svg preserveAspectRatio="none" viewBox="0 0 240 24">
              <path d="M240 24V0L0 24h240z"/>
          </svg>`
    },
    {
      id: "tilt-split",
      label: "Tilt split",
      value: `<svg preserveAspectRatio="none" viewBox="0 0 240 24">
              <path fill-opacity=".3" d="M240 24V0L0 24h240z"/>
              <path d="M240 24V3.7L0 24h240z"/>
              <path fill-opacity=".3" d="M240 24V2L0 24h240z"/>
          </svg>`
    },
    {
      id: "triangle",
      label: "Trinagle",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 1200 120">
              <path d="M1200 0L0 0 598.97 114.72 1200 0z"/>
          </svg>`
    },
    {
      id: "triangle-inv",
      label: "Trinagle inverted",
      value: `<svg preserveAspectRatio="none" viewBox="0 0 1200 120">
              <path d="M598.97 114.72L0 0 0 120 1200 120 1200 0 598.97 114.72z"></path>
          </svg>`
    },
    {
      id: "triangle-asy",
      label: "Trinagle asymmetrical",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 1200 120">
              <path d="M1200 0L0 0 892.25 114.72 1200 0z"></path>
          </svg>`
    },
    {
      id: "book",
      label: "Book",
      value: `<svg preserveAspectRatio="none" viewBox="0 0 1200 120">
              <path d="M602.45,3.86h0S572.9,116.24,281.94,120H923C632,116.24,602.45,3.86,602.45,3.86Z"></path>
          </svg>`
    },
    {
      id: "book-inv",
      label: "Book inverted",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 1200 120">
              <path d="M1200,0H0V120H281.94C572.9,116.24,602.45,3.86,602.45,3.86h0S632,116.24,923,120h277Z" ></path>
          </svg>`
    },
    {
      id: "hills",
      label: "Hills",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 240 24">
              <path d="M0 0v14.2a49 49 0 0110.3-2.7c17.5-2.8 24.4 6 43.5 11.1 26.8 7.2 39.3-19.2 63.4-19.2 22.6 0 40.4 27 65 18.6 14.7-5 19-14 39.9-17.6 7-1.3 13-1.8 17.9-1V0"/>
          </svg>`
    },
    {
      id: "hills-split",
      label: "Hills split",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 240 24">
              <path d="M0 0v14.2a49 49 0 0110.3-2.7c17.5-2.8 24.4 6 43.5 11.1 26.8 7.2 39.3-19.2 63.4-19.2 22.6 0 40.4 27 65 18.6 14.7-5 19-14 39.9-17.6 7-1.3 13-1.8 17.9-1V0"/>
              <path fill-opacity=".3" d="M0 0v18.1a49 49 0 0115.9-4.9c17-2.3 23.8 5.2 42.5 9.7C84.5 29 96.6 6.3 120 6.3c22.1 0 39.5 23.3 63.6 16 14.2-4.3 18.5-12 38.8-15.2A62.3 62.3 0 01240 6V0"/>
              <path fill-opacity=".3" d="M0 0v13.4c3.8-.8 7.6-1.4 10.3-1.8 17.3-2.8 24.4 6 47.2 11.1 32 7.2 35-18.1 60.7-18.1 24.1 0 45.2 23 67.4 14.6 13.2-5 19-13.7 39.4-17.4 5.5-1 10.4-1.5 15-1.3V0"/>
          </svg>`
    },
    {
      id: "low-hills",
      label: "Low hills",
      value: `<svg preserveAspectRatio="none" viewBox="0 0 240 24">
              <path d="M240 24V0c-51.8 0-69.9 13.2-94.7 15.6-24.7 2.4-43.9-1.2-63.8-1-19.1 0-31.2 3.6-51.3 6.5A126 126 0 010 22V24h240z"/>
          </svg>`
    },
    {
      id: "low-hills-split",
      label: "Low hills split",
      value: `<svg preserveAspectRatio="none" viewBox="0 0 240 24">
              <path fill-opacity=".3" d="M240 24V0c-51.8 0-69.9 13.2-94.7 15.6-24.7 2.4-43.9-1.2-63.8-1-19.1 0-31.2 3.6-51.3 6.5A126 126 0 010 22V24h240z"/>
              <path fill-opacity=".3" d="M240 24V2.2c-51.8 0-69.9 12-94.7 14.2-24.7 2.1-43.9-1.1-63.8-1-19.1 0-31.2 3.3-51.3 6-14.6 1.8-25.9 1.2-30.2.8V24h240z"/>
              <path d="M240 24V3.7c-51.8 0-69.9 11.7-94.7 14-24.7 2.4-43.9-3.2-63.8-3.1-19.1 0-31.2 3.6-51.3 6.5a128 128 0 01-30.2 1V24h240z"/>
          </svg>`
    },
    {
      id: "high-hills",
      label: "High hills",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 240 24">
              <path d="M240 0H0v23.5c4.3.5 15.5 1.2 30.2-.9 20.1-2.8 32.1-6.2 51.3-6.3 20 0 39 3.4 63.8 1 24.8-2.2 42.9-15 94.7-15"/>
          </svg>`
    },
    {
      id: "high-hills-split",
      label: "High hills split",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 240 24">
              <path fill-opacity=".3" d="M240 0H0v23.5c4.3.5 15.5 1.2 30.2-.9 20.1-2.8 32.1-6.2 51.3-6.3 20 0 39 3.4 63.8 1 24.8-2.2 42.9-15 94.7-15"/>
              <path fill-opacity=".3" d="M240 0H0v23.4c4.3.4 15.5 1.2 30.2-1 20.1-3.1 32.1-6.9 51.3-7 20 0 39 6 63.8 3.4C170 16.3 188.2 4 240 4"/>
              <path d="M0 23.3c4.3.5 15.5 1.3 30.2-1 20.1-3 32.1-6.8 51.3-7 20 0 39 3.8 63.8 1.2C170 14 188.2 0 240 0H0v23.3z"/>
          </svg>`
    },
    {
      id: "large-hill",
      label: "One hill",
      value: `<svg preserveAspectRatio="none" viewBox="0 0 240 24">
              <path d="M0 13.2V24h240V13.3A541.2 541.2 0 00119.8 0C69.2 0 27.6 6.9 0 13.2z"/>
          </svg>`
    },
    {
      id: "large-hill-split",
      label: "One hill split",
      value: `<svg preserveAspectRatio="none" viewBox="0 0 240 24">
              <path fill-opacity=".3" d="M0 13.2V24h240V13.3A541.2 541.2 0 00119.8 0C69.2 0 27.6 6.9 0 13.2z"/>
              <path fill-opacity=".3" d="M0 18.9V24h240v-5A437.4 437.4 0 00119.8 2.7C64.4 2.7 21.6 12.6 0 19z"/>
              <path d="M119.8 5.1C47.8 5.1 0 24 0 24h240S191.8 5.1 119.8 5.1z"/>
          </svg>`
    },
    {
      id: "pit",
      label: "Pit",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 240 24">
              <path d="M240 24V0H0v24S47.9 5.1 119.8 5.1C191.8 5.1 240 24 240 24z"/>
          </svg>`
    },
    {
      id: "pit-split",
      label: "Pit split",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 240 24">
              <path fill-opacity=".3" d="M240 24V0H0v24S47.9 5.1 119.8 5.1C191.8 5.1 240 24 240 24z"/>
              <path fill-opacity=".3" d="M119.8 2.7c55.6 0 98.5 10 120.2 16.2V0H0v18.9A434.6 434.6 0 01119.8 2.7z"/>
              <path d="M119.8.1A541 541 0 01240 13.4V0H0v13.3A538.9 538.9 0 01119.8.1z"/>
          </svg>`
    },
    {
      id: "waves",
      label: "Waves",
      value: `<svg preserveAspectRatio="none" viewBox="0 0 240 24">
              <path d="M240 0c-4.6-.2-9.6.4-15 1.4-20.4 3.7-26.3 12.3-39.4 17.3C163.3 27 144 4 119.9 4c-25.6 0-30.5 25.3-62.5 18-22.8-5.1-29.8-13.8-47-11-2.8.4-6.6 1-10.4 1.8V24h240V0z"/>
          </svg>`
    },
    {
      id: "waves-split",
      label: "Waves split",
      value: `<svg preserveAspectRatio="none" viewBox="0 0 240 24">
              <path fill-opacity=".3" d="M240 0c-4.6-.2-9.6.4-15 1.4-20.4 3.7-26.3 12.3-39.4 17.3C163.3 27 144 4 119.9 4c-25.6 0-30.5 25.3-62.5 18-22.8-5.1-29.8-13.8-47-11-2.8.4-6.6 1-10.4 1.8V24h240V0z"/>
              <path fill-opacity=".3" d="M222 3.9c-20.8 3.7-25.1 12.6-39.8 17.6C157.6 29.9 141.5 3 118.9 3 94.7 3 80.6 29.3 53.8 22c-19-5-26-13.8-43.5-11-3.7.6-7.2 1.5-10.3 2.7V24h240V2.9c-5-.8-10.9-.3-18 1z"/>
              <path d="M240 5.7c-4.8-.6-10.6-.2-17.5.9-20.3 3.2-24.6 10.9-38.9 15.2-24 7.3-39.7-16-61.8-16-23.5 0-37.3 22.7-63.5 16.5C39.7 18 33 10.3 16 12.7c-6.1.9-11.5 2.6-15.9 6.7V24h240V5.7z"/>
          </svg>`
    },
    {
      id: "windy-waves",
      label: "Windy waves",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 240 24">
              <path d="M0 0v18C21.5 7 49.6 2.2 64.8 6.4 82 11.2 75.4 25.2 91.7 23.9c14.2-1.1 25.8-12.3 40-12.3 10.4 0 7.8 6 18.8 6.9 11.5.8 21.6-5 37.2-9 19.9-5 42.1-5.3 52.3 0V0H0z"/>
          </svg>`
    },
    {
      id: "windy-waves-split",
      label: "Windy waves split",
      value: `<svg class="${t}-inv" preserveAspectRatio="none" viewBox="0 0 240 24">
              <path d="M0 0v14.9C21.6 4.7 57.5 0 71 4.3c16 4.8 9.6 19.3 26 18 14.5-1 25.5-13 40.5-12.9 11 0 10.1 6.6 21.7 7.5 12.1 1 17-7 36-11.4 17.6-4.2 33.4-6.1 44.8-5V0H0z"/>
              <path fill-opacity=".3" d="M0 16.7C19.2 5.9 55.8 1 70 5 86 10 78.4 24 94.1 22.8c14-1 27.8-12.7 42.4-12.7 10.7 0 9.1 6.5 20.5 7.4 11.9.9 16.8-6.8 34.8-11.2 19.8-4.8 36.9-5 48.1-1.7V0H0"/>
              <path fill-opacity=".3" d="M0 0v18C21.5 7 49.6 2.2 64.8 6.4 82 11.2 75.4 25.2 91.7 23.9c14.2-1.1 25.8-12.3 40-12.3 10.4 0 7.8 6 18.8 6.9 11.5.8 21.6-5 37.2-9 19.9-5 42.1-5.3 52.3 0V0H0z"/>
          </svg>`
    },
    {
      id: "zigzag",
      label: "Zigzag",
      value: `<svg class="${t}-inv" xmlns="http://www.w3.org/2000/svg" preserveAspectRatio="none" viewBox="0 0 1000 38">
              <path d="M75 38l25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25 25 25 25-25V.0H0V13l25 25 25-25z"/>
          </svg>`
    }
  ], h = {
    tl: !1,
    tc: !1,
    tr: !1,
    cl: !1,
    cr: !1,
    bl: !1,
    bc: !1,
    br: !1
  }, l = { ...h, bc: !0 }, A = { ...h, tc: !0 }, b = "shape", u = "flip-y", M = "flip-x", w = "Shape Divider", R = `<svg viewBox="0 0 22 22">
        <path fill="currentColor" d="M2 2v18h18V2H2zm0-2h18a2 2 0 012 2v18a2 2 0 01-2 2H2a2 2 0 01-2-2V2C0 .9.9 0 2 0zM1 13L2 0h18l1 9-20 4z"/>
      </svg>`, { block: x = {} } = a;
  e.on(n.events.add, (s) => {
    s.getFullName().startsWith(`.${o}`) && s.set({ private: !0 });
  }), i.addType(r, {
    isComponent: (s) => P(r)(s) || Y(s, o),
    block: x && {
      label: w,
      media: R,
      category: "Extra",
      select: !0,
      ...x
    },
    model: {
      defaults: {
        droppable: !1,
        highlightable: !1,
        classes: o,
        attributes: { [$]: r },
        components: { type: p },
        styles: `.${o} { position: relative; }`,
        delegate: {
          select: (s) => s.findFirstType(p),
          layer: (s) => s.findFirstType(p)
        }
      }
    },
    view: {
      onRender() {
        this.el.style.height = "1px";
      }
    }
  }), i.addType(p, {
    isComponent: P(p),
    model: {
      defaults: {
        droppable: !1,
        highlightable: !1,
        classes: t,
        emptyState: !0,
        icon: R,
        name: w,
        delegate: {
          remove: (s) => s.parent(),
          move: (s) => s.parent(),
          copy: (s) => s.parent()
        },
        attributes: { [$]: p },
        [b]: g[0].id,
        resizable: l,
        stylable: ["height", "width", "color"],
        styles: `
          .${t} {
              position: absolute;
              width: 100%;
              height: 100px;
              color: black;
              overflow: hidden;
          }
          .${t} > svg {
              height: 100%;
              width: 100%;
              transform: scaleY(-1);
          }
          .${t}--fl-v {
              transform: translateY(-100%);
          }
          .${t}--fl-v-h {
              transform: translateY(-100%);
          }
          .${t}--fl-v > svg {
              transform: scaleY(1);
          }
          .${t}--fl-h > svg {
              transform: scaleX(-1) scaleY(-1);
          }
          .${t}--fl-v-h > svg {
              transform: scaleY(1) scaleX(-1);
          }
          .${t} > svg > path {
              fill: currentColor;
          }
          .${t}-inv > path {
              transform: scale(-1, -1) translate(-100%, -100%);
          }
          `,
        traits: [
          {
            type: "select",
            name: b,
            label: "Shape",
            changeProp: !0,
            options: g.map(({ id: s, label: d, value: v }) => ({
              id: s,
              label: d,
              icon: { icon: v.replace("<svg ", '<svg style="width: 35px; height: 7px; fill: currentColor;"') }
            }))
          },
          {
            type: "checkbox",
            name: u,
            label: "Flip vertically",
            changeProp: !0
          },
          {
            type: "checkbox",
            name: M,
            label: "flip horizontally",
            changeProp: !0
          }
        ]
      },
      init() {
        this.on(`change:${b}`, this.onShapeChange), this.on(`change:${M} change:${u}`, this.onFlip), this.onShapeChange();
      },
      onShapeChange() {
        const s = this.get(b), d = g.find((f) => f.id === s), v = (d == null ? void 0 : d.value) || this.getInnerHTML().trim();
        this.components(""), v && this.set("content", v);
      },
      onFlip() {
        const s = this.get(u), d = this.get(M), v = ["fl-v", "fl-h", "fl-v-h"].map((D) => `${t}--${D}`);
        let f;
        s && d ? f = v[2] : d ? f = v[1] : s && (f = v[0]), this.removeClass(v), f && this.addClass(f), this.set("resizable", s ? A : l);
      }
    }
  }), W({
    editor: e,
    licenseKey: a.licenseKey,
    plan: X,
    pluginName: K,
    cleanup: () => {
      i.removeType(r), c.remove(r);
    }
  });
}, q = C(Z);
export {
  q as default
};
