// @grapesjs/studio-sdk-plugins/dist/linkImageComponent/index.es.js
const w = "app.grapesjs.com", D = "app-stage.grapesjs.com", G = "app2.grapesjs.com", k = "app-stage2.grapesjs.com", P = [
  w,
  G,
  D,
  k,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], I = "license:check:end", L = () => typeof window < "u", x = ({
  isDev: e,
  isStage: o,
  isPlatform: a
}) => `${e ? "" : `https://${o ? D : w}`}/${a ? "platform-api" : "api"}`, E = () => {
  const e = L() && window.location.hostname;
  return !!e && (P.includes(e) || P.some((o) => e.endsWith(o)));
};
async function M({
  path: e,
  baseApiUrl: o,
  method: a = "GET",
  headers: r = {},
  params: i,
  body: g
}) {
  const n = `${o || x({ isDev: !1, isStage: !1 })}${e}`, c = {
    method: a,
    headers: {
      "Content-Type": "application/json",
      ...r
    }
  };
  g && (c.body = JSON.stringify(g));
  const d = i ? new URLSearchParams(i).toString() : "", u = d ? `?${d}` : "", s = await fetch(`${n}${u}`, c);
  if (!s.ok)
    throw new Error(`HTTP error! status: ${s.status}`);
  return s.json();
}
var b = /* @__PURE__ */ ((e) => (e.free = "free", e.startup = "startup", e.business = "business", e.enterprise = "enterprise", e))(b || {}), T = /* @__PURE__ */ ((e) => (e.toastAdd = "studio:toastAdd", e.toastRemove = "studio:toastRemove", e.dialogOpen = "studio:dialogOpen", e.dialogClose = "studio:dialogClose", e.dialogExportCode = "studio:dialogExportCode", e.dialogImportCode = "studio:dialogImportCode", e.sidebarLeftSet = "studio:sidebarLeft:set", e.sidebarLeftGet = "studio:sidebarLeft:get", e.sidebarLeftToggle = "studio:sidebarLeft:toggle", e.sidebarRightSet = "studio:sidebarRight:set", e.sidebarRightGet = "studio:sidebarRight:get", e.sidebarRightToggle = "studio:sidebarRight:toggle", e.sidebarTopSet = "studio:sidebarTop:set", e.sidebarTopGet = "studio:sidebarTop:get", e.sidebarTopToggle = "studio:sidebarTop:toggle", e.sidebarBottomSet = "studio:sidebarBottom:set", e.sidebarBottomGet = "studio:sidebarBottom:get", e.sidebarBottomToggle = "studio:sidebarBottom:toggle", e.symbolAdd = "studio:symbolAdd", e.symbolDetach = "studio:symbolDetach", e.symbolOverride = "studio:symbolOverride", e.symbolPropagateStyles = "studio:propagateStyles", e.getPagesConfig = "studio:getPagesConfig", e.setPagesConfig = "studio:setPagesConfig", e.getPageSettings = "studio:getPageSettings", e.setPageSettings = "studio:setPageSettings", e.clearPage = "studio:clearPage", e.projectFiles = "studio:projectFiles", e.validateCode = "studio:validateCode", e.canvasReload = "studio:canvasReload", e.getBlocksPanel = "studio:getBlocksPanel", e.setBlocksPanel = "studio:setBlocksPanel", e.getStateContextMenu = "studio:getStateContextMenu", e.setStateContextMenu = "studio:setStateContextMenu", e.contextMenuComponent = "studio:contextMenuComponent", e.layoutAdd = "studio:layoutAdd", e.layoutRemove = "studio:layoutRemove", e.layoutToggle = "studio:layoutToggle", e.layoutUpdate = "studio:layoutUpdate", e.layoutGet = "studio:layoutGet", e.layoutConfigGet = "studio:layoutConfigGet", e.layoutConfigSet = "studio:layoutConfigSet", e.layoutComponentAdd = "studio:layoutComponentAdd", e.layoutComponentGet = "studio:layoutComponentGet", e.layoutComponentRemove = "studio:layoutComponentRemove", e.layoutComponentRender = "studio:layoutComponentRender", e.getStateTheme = "studio:getStateTheme", e.setStateTheme = "studio:setStateTheme", e.settings = "studio:settings", e.assetProviderGet = "studio:assetProviderGet", e.assetProviderAdd = "studio:assetProviderAdd", e.assetProviderRemove = "studio:assetProviderRemove", e.fontGet = "studio:fontGet", e.fontAdd = "studio:fontAdd", e.fontRemove = "studio:fontRemove", e.fontManagerOpen = "studio:fontManagerOpen", e.menuFontLoad = "studio:menuFontLoad", e.toggleStateDataSource = "studio:toggleStateDataSource", e.getStateDataSource = "studio:getStateDataSource", e.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", e.dataSourceSetImporter = "studio:dataSourceSetImporter", e.dataSourceSetExporter = "studio:dataSourceSetExporter", e.setDragAbsolute = "studio:setDragAbsolute", e))(T || {});
const A = {
  [b.free]: 0,
  [b.startup]: 10,
  [b.business]: 20,
  [b.enterprise]: 30
};
function N(e) {
  const o = e;
  return o.init = (a) => (r) => e(r, a), o;
}
const O = (e) => /* @__PURE__ */ N(e);
async function B({
  editor: e,
  plan: o,
  pluginName: a,
  licenseKey: r,
  onLicenseCheckResponse: i,
  cleanup: g
}) {
  let p = "", n;
  const c = (s) => {
    console.warn("Cleanup plugin:", a, "Reason:", s), g();
  }, d = (s = {}) => {
    var y;
    const { error: f, sdkLicense: t } = s, l = (y = s.plan) == null ? void 0 : y.category;
    if (!(t || s.license) || f)
      c(f || "Invalid license");
    else if (l) {
      const v = A[o], R = A[l];
      v > R && c({ pluginRequiredPlan: o, licensePlan: l });
    }
  };
  e.Commands.has(T.settings) && (n = e.runCommand(T.settings), p = (n == null ? void 0 : n.baseUrl) || "");
  const u = (s) => {
    i == null || i(s), s && d(s);
  };
  if (!n) {
    e.onReady(async () => {
      if (!E())
        if (r) {
          const s = await $({ licenseKey: r, pluginName: a, baseApiUrl: p });
          u(s);
        } else
          c("The `licenseKey` option not provided");
    });
    return;
  }
  if (n.licensePlan || n.licenseError) {
    const s = U(n);
    u(s);
    return;
  }
  e.on(I, (s) => u(s));
}
const U = (e) => ({
  sdkLicense: e.license,
  license: e.license,
  error: e.licenseError,
  plan: e.licensePlan
});
async function $(e) {
  const { licenseKey: o, pluginName: a, baseApiUrl: r } = e;
  try {
    return (await M({
      baseApiUrl: r,
      path: `/sdk/${o || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: a
      }
    })).result || {};
  } catch (i) {
    return console.error("Error during SDK license check:", i), !1;
  }
}
const j = "linkImageComponent", _ = b.startup, F = function(e, o = {}) {
  const { Components: a, Blocks: r } = e, i = "link-image", g = "Link Image", p = "link-image", { model: n, view: c } = e.Components.getType("image"), d = a.getType("link").model.getDefaults().traits, u = n.getDefaults().traits, s = d.filter((t) => t !== "title"), f = u.filter((t) => t.name !== "customAttributes").map((t) => typeof t == "string" ? { name: t } : t).map(
    (t) => ({
      ...t,
      getValue({ component: l }) {
        var h;
        return (h = l.components().at(0).getTrait(t.name)) == null ? void 0 : h.getValue();
      },
      setValue({ component: l, value: h, options: y }) {
        var v;
        (v = l.components().at(0).getTrait(t.name)) == null || v.setValue(h, y);
      }
    })
  );
  a.addType(i, {
    isComponent: (t) => {
      var l;
      return t.tagName === "A" && ((l = t.children) == null ? void 0 : l.length) === 1 && t.children[0].tagName === "IMG";
    },
    block: {
      label: g,
      media: '<svg viewBox="0 0 24 24"><path fill="currentColor" d="M2 6.3A2.3 2.3 0 0 1 4.3 4h15.4A2.3 2.3 0 0 1 22 6.3v7a4.7 4.7 0 0 0-2.8-.8h-3.4a4.8 4.8 0 0 0-4 7.5H4.3A2.3 2.3 0 0 1 2 17.7zm21 11a3.8 3.8 0 0 0-3.8-3.8.8.8 0 0 0 0 1.5h.2a2.3 2.3 0 0 1-.1 4.5H19a.8.8 0 0 0 .2 1.5h.2a3.8 3.8 0 0 0 3.5-3.8m-6.5-3a.8.8 0 0 0-.8-.7h-.1a3.8 3.8 0 0 0 .2 7.5.8.8 0 0 0 0-1.5h-.2a2.3 2.3 0 0 1 .2-4.5.8.8 0 0 0 .7-.7m3.5 3a.8.8 0 0 0-.8-.8h-3.6a.8.8 0 0 0 .2 1.5h3.6a.8.8 0 0 0 .6-.8"/></svg>'
    },
    extend: "link",
    extendFn: ["init"],
    model: {
      defaults: {
        name: g,
        classes: p,
        resizable: { ratioDefault: !0 },
        traits: [...f, ...s],
        droppable: !1,
        stylable: [
          "width",
          "min-width",
          "max-width",
          "height",
          "min-height",
          "max-height",
          "padding",
          "margin",
          "position",
          "border-radius",
          "border-width",
          "border-style",
          "border-color",
          "opacity",
          "box-shadow",
          "filter",
          "transform"
        ],
        components: { type: "image" },
        styles: `
          .${p} {
            display: inline-block;
            overflow: hidden;
          }
          .${p} > img {
            object-fit: cover;
            display: block;
            width: 100%;
            height: 100%;
          }
        `
      },
      init() {
        const t = this.components().at(0);
        t == null || t.set({ layerable: !1, locked: !0 });
      }
    },
    view: {
      onActive(t) {
        c.prototype.onActive.call(this, t);
      }
    }
  }), B({
    editor: e,
    licenseKey: o.licenseKey,
    plan: _,
    pluginName: j,
    cleanup: () => {
      a.removeType(i), r.remove(i);
    }
  });
}, W = O(F);
export {
  W as default
};
