// @grapesjs/studio-sdk-plugins/dist/dataSourceEjs/index.es.js
const I = "app.grapesjs.com", m = "app-stage.grapesjs.com", v = "app2.grapesjs.com", O = "app-stage2.grapesjs.com", $ = [
  I,
  v,
  m,
  O,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], w = "license:check:end", T = () => typeof window < "u", L = ({
  isDev: s,
  isStage: t,
  isPlatform: e
}) => `${s ? "" : `https://${t ? m : I}`}/${e ? "platform-api" : "api"}`, k = () => {
  const s = T() && window.location.hostname;
  return !!s && ($.includes(s) || $.some((t) => s.endsWith(t)));
};
async function A({
  path: s,
  baseApiUrl: t,
  method: e = "GET",
  headers: r = {},
  params: n,
  body: a
}) {
  const i = `${t || L({ isDev: !1, isStage: !1 })}${s}`, l = {
    method: e,
    headers: {
      "Content-Type": "application/json",
      ...r
    }
  };
  a && (l.body = JSON.stringify(a));
  const c = n ? new URLSearchParams(n).toString() : "", u = c ? `?${c}` : "", p = await fetch(`${i}${u}`, l);
  if (!p.ok)
    throw new Error(`HTTP error! status: ${p.status}`);
  return p.json();
}
var y = /* @__PURE__ */ ((s) => (s.free = "free", s.startup = "startup", s.business = "business", s.enterprise = "enterprise", s))(y || {}), b = /* @__PURE__ */ ((s) => (s.toastAdd = "studio:toastAdd", s.toastRemove = "studio:toastRemove", s.dialogOpen = "studio:dialogOpen", s.dialogClose = "studio:dialogClose", s.dialogExportCode = "studio:dialogExportCode", s.dialogImportCode = "studio:dialogImportCode", s.sidebarLeftSet = "studio:sidebarLeft:set", s.sidebarLeftGet = "studio:sidebarLeft:get", s.sidebarLeftToggle = "studio:sidebarLeft:toggle", s.sidebarRightSet = "studio:sidebarRight:set", s.sidebarRightGet = "studio:sidebarRight:get", s.sidebarRightToggle = "studio:sidebarRight:toggle", s.sidebarTopSet = "studio:sidebarTop:set", s.sidebarTopGet = "studio:sidebarTop:get", s.sidebarTopToggle = "studio:sidebarTop:toggle", s.sidebarBottomSet = "studio:sidebarBottom:set", s.sidebarBottomGet = "studio:sidebarBottom:get", s.sidebarBottomToggle = "studio:sidebarBottom:toggle", s.symbolAdd = "studio:symbolAdd", s.symbolDetach = "studio:symbolDetach", s.symbolOverride = "studio:symbolOverride", s.symbolPropagateStyles = "studio:propagateStyles", s.getPagesConfig = "studio:getPagesConfig", s.setPagesConfig = "studio:setPagesConfig", s.getPageSettings = "studio:getPageSettings", s.setPageSettings = "studio:setPageSettings", s.clearPage = "studio:clearPage", s.projectFiles = "studio:projectFiles", s.validateCode = "studio:validateCode", s.canvasReload = "studio:canvasReload", s.getBlocksPanel = "studio:getBlocksPanel", s.setBlocksPanel = "studio:setBlocksPanel", s.getStateContextMenu = "studio:getStateContextMenu", s.setStateContextMenu = "studio:setStateContextMenu", s.contextMenuComponent = "studio:contextMenuComponent", s.layoutAdd = "studio:layoutAdd", s.layoutRemove = "studio:layoutRemove", s.layoutToggle = "studio:layoutToggle", s.layoutUpdate = "studio:layoutUpdate", s.layoutGet = "studio:layoutGet", s.layoutConfigGet = "studio:layoutConfigGet", s.layoutConfigSet = "studio:layoutConfigSet", s.layoutComponentAdd = "studio:layoutComponentAdd", s.layoutComponentGet = "studio:layoutComponentGet", s.layoutComponentRemove = "studio:layoutComponentRemove", s.layoutComponentRender = "studio:layoutComponentRender", s.getStateTheme = "studio:getStateTheme", s.setStateTheme = "studio:setStateTheme", s.settings = "studio:settings", s.assetProviderGet = "studio:assetProviderGet", s.assetProviderAdd = "studio:assetProviderAdd", s.assetProviderRemove = "studio:assetProviderRemove", s.fontGet = "studio:fontGet", s.fontAdd = "studio:fontAdd", s.fontRemove = "studio:fontRemove", s.fontManagerOpen = "studio:fontManagerOpen", s.menuFontLoad = "studio:menuFontLoad", s.toggleStateDataSource = "studio:toggleStateDataSource", s.getStateDataSource = "studio:getStateDataSource", s.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", s.dataSourceSetImporter = "studio:dataSourceSetImporter", s.dataSourceSetExporter = "studio:dataSourceSetExporter", s.setDragAbsolute = "studio:setDragAbsolute", s))(b || {});
const S = {
  [y.free]: 0,
  [y.startup]: 10,
  [y.business]: 20,
  [y.enterprise]: 30
};
function B(s) {
  const t = s;
  return t.init = (e) => (r) => s(r, e), t;
}
const N = (s) => /* @__PURE__ */ B(s);
async function M({
  editor: s,
  plan: t,
  pluginName: e,
  licenseKey: r,
  onLicenseCheckResponse: n,
  cleanup: a
}) {
  let o = "", i;
  const l = (p) => {
    console.warn("Cleanup plugin:", e, "Reason:", p), a();
  }, c = (p = {}) => {
    var f;
    const { error: h, sdkLicense: E } = p, g = (f = p.plan) == null ? void 0 : f.category;
    if (!(E || p.license) || h)
      l(h || "Invalid license");
    else if (g) {
      const x = S[t], P = S[g];
      x > P && l({ pluginRequiredPlan: t, licensePlan: g });
    }
  };
  s.Commands.has(b.settings) && (i = s.runCommand(b.settings), o = (i == null ? void 0 : i.baseUrl) || "");
  const u = (p) => {
    n == null || n(p), p && c(p);
  };
  if (!i) {
    s.onReady(async () => {
      if (!k())
        if (r) {
          const p = await j({ licenseKey: r, pluginName: e, baseApiUrl: o });
          u(p);
        } else
          l("The `licenseKey` option not provided");
    });
    return;
  }
  if (i.licensePlan || i.licenseError) {
    const p = G(i);
    u(p);
    return;
  }
  s.on(w, (p) => u(p));
}
const G = (s) => ({
  sdkLicense: s.license,
  license: s.license,
  error: s.licenseError,
  plan: s.licensePlan
});
async function j(s) {
  const { licenseKey: t, pluginName: e, baseApiUrl: r } = s;
  try {
    return (await A({
      baseApiUrl: r,
      path: `/sdk/${t || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: e
      }
    })).result || {};
  } catch (n) {
    return console.error("Error during SDK license check:", n), !1;
  }
}
const D = (s) => {
  const t = s;
  return t.logicalOperator === "and" || t.logicalOperator === "or";
}, F = (s, t) => (s.config.optsHtml = {
  ...s.config.optsHtml,
  exporter: t
}, () => {
  s.config.optsHtml.exporter = void 0;
}), J = (s, t) => {
  const e = (n) => {
    const { input: a } = n, { contains: o } = t;
    (!o || a.includes(o)) && (n.input = t.importer.import(a));
  }, r = s.Parser.events.htmlBefore;
  return s.on(r, e), () => {
    s.off(r, e);
  };
};
class W {
  getVariableSyntax({ dataResolver: t }) {
    const { defaultValue: e } = t, r = this.getFullPath(t), n = e ? ` || \`${e}\`` : "";
    return `<%= ${r}${n} %>`;
  }
  getCollectionStartSyntax({ dataResolver: t }) {
    const { collectionId: e, dataSource: r, startIndex: n, endIndex: a } = t, o = (r == null ? void 0 : r.path) || e || "", i = this._sanitizeVariableName(e);
    let l = "";
    if (n !== void 0 || a !== void 0) {
      const c = n ?? 0, u = a !== void 0 ? a : "";
      l = `.slice(${c}${u !== "" ? `, ${u}` : ""})`;
    }
    return `<% ${o}${l}.forEach(function(currentItem, index, arr) {
const startIndex = parseInt(${n ?? 0}, 10);
const endIndex = startIndex + arr.length;
const ${i}={ currentItem, currentIndex: startIndex + index, remainingItems: endIndex - (startIndex + index)} %>`;
  }
  _sanitizeVariableName(t) {
    return t.replace(/[^a-zA-Z0-9_$]/g, "_");
  }
  getCollectionEndSyntax() {
    return "<% }) %>";
  }
  getConditionalStartSyntax({ dataResolver: t }) {
    return `<% if (${this._parseCondition(t.condition)}) { %>`;
  }
  getConditionElseSyntax() {
    return "<% } else { %>";
  }
  getConditionalEndSyntax() {
    return "<% } %>";
  }
  getFullPath({ collectionId: t, path: e, variableType: r }) {
    return (t ? [this._sanitizeVariableName(t), r, e].filter(Boolean).join(".") : e) || "";
  }
  _parseCondition(t) {
    if (!t) return "";
    if (D(t)) {
      const e = t.statements.map((a) => this._parseCondition(a)), r = _[t.logicalOperator];
      let n = e.length > 1 ? `(${e.join(`)${r}(`)})` : e[0];
      return t.logicalOperator === "xor" && (n = `!!(${n})`), n;
    } else {
      const e = this._parseValue(t.left), r = this._parseValue(t.right);
      switch (t.operator) {
        case "equals":
          return `${e} === ${r}`;
        case ">":
        case "<":
        case ">=":
        case "<=":
        case "=":
        case "!=":
          return `${e} ${t.operator} ${r}`;
        case "contains":
          return `${e}?.includes(${r})`;
        case "startsWith":
          return `${e}?.startsWith(${r})`;
        case "endsWith":
          return `${e}?.endsWith(${r})`;
        case "equalsIgnoreCase":
          return `${e}?.toLowerCase() === ${r}?.toLowerCase()`;
        case "trimEquals":
          return `${e}?.trim() === ${r}?.trim()`;
        case "isTruthy":
          return `!!${e}`;
        case "isFalsy":
          return `!${e}`;
        case "isDefined":
          return `${e} !== undefined`;
        case "isNull":
          return `${e} === null`;
        case "isUndefined":
          return `${e} === undefined`;
        case "isArray":
          return `Array.isArray(${e})`;
        case "isObject":
          return `typeof ${e} === 'object' && ${e} !== null && !Array.isArray(${e})`;
        case "isString":
          return `typeof ${e} === 'string'`;
        case "isNumber":
          return `typeof ${e} === 'number' && !isNaN(${e})`;
        case "isBoolean":
          return `typeof ${e} === 'boolean'`;
        default:
          return `${e} ${t.operator} ${r}`;
      }
    }
  }
  _parseValue(t) {
    return typeof t == "object" ? this.getFullPath(t) : typeof t == "string" ? `'${t.replace(/'/g, "\\'")}'` : String(t);
  }
}
const _ = {
  and: "&&",
  or: "||",
  xor: "^"
}, R = ["===", "==", "!==", "!=", "<=", ">=", "<", ">", "&&", "||"], U = {
  "===": "=",
  "==": "=",
  "!==": "!=",
  "!=": "!=",
  "<=": "<=",
  ">=": ">=",
  "<": "<",
  ">": ">",
  "&&": "&&",
  "||": "||"
};
class V {
  constructor() {
    this.messages = [], this.collectionStack = [];
  }
  import(t) {
    this.messages = [], this.collectionStack = [];
    try {
      const e = this.parseEjsTemplate(t);
      return this.astToGrapesJS(e);
    } catch (e) {
      return console.error("Error during import:", e), this.addMessage("error", `Error during import: ${e instanceof Error ? e.message : String(e)}`), "";
    }
  }
  parseEjsTemplate(t) {
    const e = [];
    let r = 0;
    for (; r < t.length; ) {
      const n = t.indexOf("<%", r);
      if (n === -1) {
        r < t.length && e.push(this.createAppendLiteral(t.slice(r)));
        break;
      }
      n > r && e.push(this.createAppendLiteral(t.slice(r, n)));
      const a = t.indexOf("%>", n);
      if (a === -1)
        throw new Error("Unclosed EJS tag");
      const o = t.slice(n + 2, a).trim();
      if (r = a + 2, o.startsWith("=") || o.startsWith("-")) {
        const i = o.slice(1).trim();
        e.push(this.createAppendExpression(i));
      } else {
        if (o.startsWith("#"))
          continue;
        if (o.startsWith("if")) {
          const i = this.parseIfStatement(o, t, r);
          e.push(i), r = i.endPos || r;
        } else if (o.startsWith("for")) {
          const i = this.parseForStatement(o, t, r);
          e.push(i), r = i.endPos || r;
        } else {
          const i = o.match(/([\w][\w.-]*[\w])\s*\.\s*forEach\s*\(\s*(\w+)\s*=>\s*\{?/);
          if (i) {
            const l = i[1], c = i[2], u = this.parseForEachStatement(l, c, t, r);
            e.push(u), r = u.endPos || r;
          } else
            this.addMessage("info", `Unsupported scriptlet content: ${o}`);
        }
      }
    }
    return {
      type: "Program",
      body: e
    };
  }
  createAppendLiteral(t) {
    return {
      type: "ExpressionStatement",
      expression: {
        type: "CallExpression",
        callee: { type: "Identifier", name: "__append" },
        arguments: [{ type: "Literal", value: t }]
      }
    };
  }
  createAppendExpression(t) {
    return {
      type: "ExpressionStatement",
      expression: {
        type: "CallExpression",
        callee: { type: "Identifier", name: "__append" },
        arguments: [this.parseJsExpression(t)]
      }
    };
  }
  parseIfStatement(t, e, r) {
    const n = t.match(/if\s*\((.*)\)\s*\{?/);
    if (!n)
      throw new Error("Invalid if statement syntax");
    const a = this.parseJsExpression(n[1]), { body: o, endPos: i } = this.parseTemplateBlock(e, r);
    let l = null, c = i;
    const u = e.indexOf("<%", c);
    if (u === c) {
      const p = e.indexOf("%>", u);
      if (p !== -1) {
        const h = e.slice(u + 2, p).trim(), E = p + 2 - u;
        if (h.startsWith("} else if")) {
          c = u + E;
          const g = h.substring(h.indexOf("if")).trim(), d = this.parseIfStatement(g, e, c);
          l = {
            type: "IfStatement",
            test: d.test,
            consequent: d.consequent,
            alternate: d.alternate
          }, c = d.endPos;
        } else if (h.startsWith("} else")) {
          c = u + E;
          const { body: g, endPos: d } = this.parseTemplateBlock(e, c);
          l = {
            type: "BlockStatement",
            body: g
          }, c = d;
          const f = e.indexOf("<%", c), x = e.indexOf("%>", f);
          f === c && e.slice(f + 2, x).trim() === "}" && (c = x + 2);
        } else h === "}" && (c = u + E);
      }
    }
    return {
      type: "IfStatement",
      test: a,
      consequent: {
        type: "BlockStatement",
        body: o
      },
      alternate: l,
      endPos: c
    };
  }
  parseForStatement(t, e, r) {
    const n = t.match(/for\s*\((.*)\)\s*\{?/);
    if (!n)
      throw new Error("Invalid for statement syntax");
    const a = n[1].split("of");
    if (a.length !== 2)
      throw new Error("Only for...of loops are supported");
    const o = a[0].trim().replace(/^let\s+|^const\s+|^var\s+/, ""), i = a[1].trim(), { body: l, endPos: c } = this.parseTemplateBlock(e, r);
    return {
      type: "ForStatement",
      init: {
        type: "VariableDeclaration",
        kind: "let",
        declarations: [
          {
            type: "VariableDeclarator",
            id: { type: "Identifier", name: o },
            init: this.parseJsExpression(i)
          }
        ]
      },
      test: null,
      update: null,
      body: {
        type: "BlockStatement",
        body: l
      },
      // Consume the closing '}' tag for the for loop
      endPos: (() => {
        const u = e.indexOf("<%", c), p = e.indexOf("%>", u);
        return u === c && e.slice(u + 2, p).trim() === "}" ? p + 2 : c;
      })()
    };
  }
  parseForEachStatement(t, e, r, n) {
    const { body: a, endPos: o } = this.parseTemplateBlock(r, n);
    return {
      type: "ForEachStatement",
      collection: this.parseJsExpression(t),
      itemIdentifier: { type: "Identifier", name: e },
      body: {
        type: "BlockStatement",
        body: a
      },
      endPos: (() => {
        const i = r.indexOf("<%", o), l = r.indexOf("%>", i);
        return i === o && r.slice(i + 2, l).trim() === "}" ? l + 2 : o;
      })()
    };
  }
  parseTemplateBlock(t, e) {
    const r = [];
    let n = e;
    for (; n < t.length; ) {
      const a = t.indexOf("<%", n);
      if (a === -1) {
        n < t.length && r.push(this.createAppendLiteral(t.slice(n))), n = t.length;
        break;
      }
      a > n && r.push(this.createAppendLiteral(t.slice(n, a)));
      const o = t.indexOf("%>", a);
      if (o === -1)
        throw new Error("Unclosed EJS tag within block");
      const i = t.slice(a + 2, o).trim();
      if (i === "}" || i.startsWith("} else")) {
        n = a;
        break;
      }
      if (n = o + 2, i.startsWith("=") || i.startsWith("-")) {
        const l = i.slice(1).trim();
        r.push(this.createAppendExpression(l));
      } else {
        if (i.startsWith("#"))
          continue;
        if (i.startsWith("if")) {
          const l = this.parseIfStatement(i, t, n);
          r.push(l), n = l.endPos || n;
        } else if (i.startsWith("for")) {
          const l = this.parseForStatement(i, t, n);
          r.push(l), n = l.endPos || n;
        } else {
          const l = i.match(/([\w][\w.-]*[\w])\s*\.\s*forEach\s*\(\s*(\w+)\s*=>\s*\{?/);
          if (l) {
            const c = l[1], u = l[2], p = this.parseForEachStatement(c, u, t, n);
            r.push(p), n = p.endPos || n;
          } else
            this.addMessage("info", `Unsupported scriptlet content within block: ${i}`);
        }
      }
    }
    return { body: r, endPos: n };
  }
  parseJsExpression(t) {
    if (/^(['"]).*\1$/.test(t))
      return {
        type: "Literal",
        value: t.slice(1, -1)
      };
    if (/^-?\d+$/.test(t))
      return {
        type: "Literal",
        value: parseInt(t, 10)
      };
    if (/^-?\d+\.\d+$/.test(t))
      return {
        type: "Literal",
        value: parseFloat(t)
      };
    if (t === "true" || t === "false")
      return {
        type: "Literal",
        value: t === "true"
      };
    for (const n of R) {
      const a = t.lastIndexOf(n);
      if (a > 0 && t[a - 1] !== "!" && t[a - 1] !== "=" && t[a - 1] !== "<" && t[a - 1] !== ">")
        return {
          type: n === "&&" || n === "||" ? "LogicalExpression" : "BinaryExpression",
          operator: U[n],
          left: this.parseJsExpression(t.slice(0, a).trim()),
          right: this.parseJsExpression(t.slice(a + n.length).trim())
        };
    }
    const e = t.indexOf("."), r = t.indexOf("[");
    if (e > 0 && (r === -1 || e < r) || r > 0 && (e === -1 || r < e)) {
      const n = e > 0 && (r === -1 || e < r) ? e : r;
      let a, o, i = !1;
      if (n === e)
        a = t.slice(0, e), o = { type: "Identifier", name: t.slice(e + 1) }, i = !1;
      else {
        a = t.slice(0, r);
        const l = t.indexOf("]", r);
        if (l === -1)
          throw new Error("Unclosed bracket in expression");
        const c = t.slice(r + 1, l);
        o = /^(['"]).*\1$/.test(c) || /^-?\d+(\.\d+)?$/.test(c) ? this.parseJsExpression(c) : { type: "Identifier", name: c }, i = !0;
      }
      return {
        type: "MemberExpression",
        object: this.parseJsExpression(a),
        property: o,
        computed: i
      };
    }
    return {
      type: "Identifier",
      name: t
    };
  }
  astToGrapesJS(t) {
    switch (t.type) {
      case "Program":
        return this.handleProgram(t);
      case "VariableDeclaration":
        return this.handleVariableDeclaration(t);
      case "ExpressionStatement":
        return this.handleExpressionStatement(t);
      case "IfStatement":
        return this.handleIfStatement(t);
      case "ForStatement":
        return this.handleForStatement(t);
      case "ForEachStatement":
        return this.handleForEachStatement(t);
      case "BlockStatement":
        return this.handleBlockStatement(t);
      case "Literal":
        return this.handleLiteral(t);
      case "Identifier":
        return this.handleIdentifier(t);
      case "MemberExpression":
        return this.handleMemberExpression(t);
      case "BinaryExpression":
        return this.handleBinaryExpression(t);
      case "LogicalExpression":
        return this.handleLogicalExpression(t);
      case "CallExpression":
        return this.handleCallExpression(t);
      default:
        return this.addMessage("warning", `Unsupported node type: ${t.type}`), "";
    }
  }
  handleProgram(t) {
    let e = "";
    for (const r of t.body)
      e += this.astToGrapesJS(r);
    return e;
  }
  handleVariableDeclaration(t) {
    return "";
  }
  handleExpressionStatement(t) {
    return t.expression.type === "CallExpression" && t.expression.callee.type === "Identifier" && t.expression.callee.name === "__append" && t.expression.arguments.length > 0 ? this.astToGrapesJS(t.expression.arguments[0]) : "";
  }
  handleIfStatement(t) {
    const e = this.getConditionExpression(t.test), r = this.astToGrapesJS(t.consequent);
    let n = "";
    return t.alternate && (n = this.astToGrapesJS(t.alternate)), `
      <data-condition data-gjs-data-resolver='${JSON.stringify({ condition: e })}'>
        <data-condition-true-content>${r}</data-condition-true-content>
        <data-condition-false-content>${n}</data-condition-false-content>
      </data-condition>
    `;
  }
  handleForStatement(t) {
    var i;
    if (((i = t.init) == null ? void 0 : i.type) !== "VariableDeclaration" || t.init.declarations.length === 0)
      return this.addMessage("warning", "Unsupported for loop init format"), "";
    const e = t.init.declarations[0];
    if (e.id.type !== "Identifier")
      return this.addMessage("warning", "Unsupported for loop variable declaration id"), "";
    let r = "";
    if (e.init)
      r = this.getExpressionString(e.init);
    else
      return this.addMessage("warning", "For loop collection is undefined"), "";
    const n = e.id.name, a = n;
    this.collectionStack.push(n);
    const o = this.astToGrapesJS(t.body);
    return this.collectionStack.pop(), `
      <data-collection data-gjs-data-resolver='${JSON.stringify({
      dataSource: { type: "data-variable", path: r },
      collectionId: a
    })}'>
        <data-collection-item>
          ${o}
        </data-collection-item>
      </data-collection>
    `;
  }
  handleForEachStatement(t) {
    const e = this.getExpressionString(t.collection), r = t.itemIdentifier.name, n = r;
    this.collectionStack.push(r);
    const a = this.astToGrapesJS(t.body);
    return this.collectionStack.pop(), `
      <data-collection data-gjs-data-resolver='${JSON.stringify({
      dataSource: { type: "data-variable", path: e },
      collectionId: n
    })}'>
        <data-collection-item>
          ${a}
        </data-collection-item>
      </data-collection>
    `;
  }
  handleBlockStatement(t) {
    let e = "";
    for (const r of t.body)
      e += this.astToGrapesJS(r);
    return e;
  }
  handleLiteral(t) {
    return String(t.value ?? "");
  }
  handleIdentifier(t) {
    let e = { path: t.name };
    if (this.collectionStack.length > 0) {
      const r = this.collectionStack[this.collectionStack.length - 1];
      t.name === r && (e = {
        collectionId: r,
        variableType: "currentItem"
      });
    }
    return `<data-variable data-gjs-data-resolver='${JSON.stringify(e)}'></data-variable>`;
  }
  handleMemberExpression(t, e = {}) {
    let n = { path: this.getExpressionString(t), ...e };
    if (this.collectionStack.length > 0) {
      const a = this.collectionStack[this.collectionStack.length - 1];
      let o = t;
      const i = [];
      for (; o.type === "MemberExpression"; )
        o.property.type === "Identifier" && !o.computed ? i.unshift(o.property.name) : o.property.type === "Literal" && o.computed ? i.unshift(String(o.property.value)) : i.unshift(this.getExpressionString(o.property)), o = o.object;
      if (o.type === "Identifier" && o.name === a) {
        const l = i.join(".");
        n = {
          ...e,
          collectionId: a,
          variableType: "currentItem",
          path: l
        };
      }
    }
    return `<data-variable data-gjs-data-resolver='${JSON.stringify(n)}'></data-variable>`;
  }
  handleBinaryExpression(t) {
    return `${this.getExpressionString(t.left)} ${t.operator} ${this.getExpressionString(t.right)}`;
  }
  handleLogicalExpression(t) {
    return t.operator === "||" && t.right.type === "Literal" ? this.handleMemberExpression(t.left, {
      defaultValue: t.right.value
    }) : `${this.getExpressionString(t.left)} ${t.operator} ${this.getExpressionString(t.right)}`;
  }
  handleCallExpression(t) {
    var e;
    if (t.callee.type === "Identifier" && t.callee.name === "include") {
      const r = ((e = t.arguments[0]) == null ? void 0 : e.type) === "Literal" ? String(t.arguments[0].value) : "";
      return this.addMessage("info", `Partial includes are not fully supported: ${r}`), "";
    }
    return "";
  }
  getConditionExpression(t) {
    if (t.type === "Identifier")
      return {
        logicalOperator: "and",
        statements: [
          {
            left: { type: "data-variable", path: t.name },
            operator: "isTruthy"
          }
        ]
      };
    if (t.type === "MemberExpression") {
      let e = this.getExpressionString(t), r = "isTruthy";
      return e.startsWith("!!") ? (r = "isTruthy", e = e.slice(2)) : e.startsWith("!") && (r = "isFalsy", e = e.slice(1)), {
        logicalOperator: "and",
        statements: [
          {
            left: { type: "data-variable", path: e },
            operator: r
          }
        ]
      };
    }
    if (t.type === "BinaryExpression" || t.type === "LogicalExpression") {
      const e = this.getConditionExpression(t.left), r = this.getConditionExpression(t.right);
      return `${e} ${t.operator} ${r}`;
    }
    return t.type === "Literal" ? typeof t.value == "string" ? `'${t.value}'` : String(t.value ?? "") : this.getExpressionString(t);
  }
  getExpressionString(t) {
    switch (t.type) {
      case "Identifier":
        return t.name;
      case "Literal":
        return String(t.value ?? "");
      case "MemberExpression": {
        const e = this.getExpressionString(t.object), r = this.getExpressionString(t.property);
        return t.computed ? `${e}[${r}]` : `${e}.${r}`;
      }
      case "BinaryExpression": {
        const e = this.getExpressionString(t.left), r = this.getExpressionString(t.right);
        return `${e} ${t.operator} ${r}`;
      }
      case "LogicalExpression": {
        const e = this.getExpressionString(t.left), r = this.getExpressionString(t.right);
        return `${e} ${t.operator} ${r}`;
      }
      default:
        return this.addMessage("warning", `Unsupported expression type for getExpressionString: ${t.type}`), "";
    }
  }
  addMessage(t, e) {
    this.messages.push({
      type: t,
      message: e
    });
  }
}
const q = "dataSourceEjs", H = y.startup, K = function(s, t = {}) {
  const e = F(s, new W()), r = J(s, {
    contains: "<%",
    importer: new V()
  });
  M({
    editor: s,
    licenseKey: t.licenseKey,
    plan: H,
    pluginName: q,
    cleanup: () => {
      e(), r();
    }
  });
}, z = N(K);
export {
  z as default
};
