// @grapesjs/studio-sdk-plugins/dist/animationComponent/index.umd.js
(function(b,I){typeof exports=="object"&&typeof module<"u"?module.exports=I():typeof define=="function"&&define.amd?define(I):(b=typeof globalThis<"u"?globalThis:b||self,b.StudioSdkPlugins_animationComponent=I())})(this,function(){"use strict";const b="app.grapesjs.com",I="app-stage.grapesjs.com",T=[b,"app2.grapesjs.com",I,"app-stage2.grapesjs.com","localhost","127.0.0.1",".local-credentialless.webcontainer.io",".local.webcontainer.io","-sandpack.codesandbox.io"],M="license:check:end",N=()=>typeof window<"u",F=({isDev:t,isStage:e,isPlatform:a})=>`${t?"":`https://${e?I:b}`}/${a?"platform-api":"api"}`,V=()=>{const t=N()&&window.location.hostname;return!!t&&(T.includes(t)||T.some(e=>t.endsWith(e)))};async function X({path:t,baseApiUrl:e,method:a="GET",headers:r={},params:i,body:o}){const s=`${e||F({isDev:!1,isStage:!1})}${t}`,p={method:a,headers:{"Content-Type":"application/json",...r}};o&&(p.body=JSON.stringify(o));const d=i?new URLSearchParams(i).toString():"",f=d?`?${d}`:"",n=await fetch(`${s}${f}`,p);if(!n.ok)throw new Error(`HTTP error! status: ${n.status}`);return n.json()}var v=(t=>(t.free="free",t.startup="startup",t.business="business",t.enterprise="enterprise",t))(v||{}),w=(t=>(t.toastAdd="studio:toastAdd",t.toastRemove="studio:toastRemove",t.dialogOpen="studio:dialogOpen",t.dialogClose="studio:dialogClose",t.dialogExportCode="studio:dialogExportCode",t.dialogImportCode="studio:dialogImportCode",t.sidebarLeftSet="studio:sidebarLeft:set",t.sidebarLeftGet="studio:sidebarLeft:get",t.sidebarLeftToggle="studio:sidebarLeft:toggle",t.sidebarRightSet="studio:sidebarRight:set",t.sidebarRightGet="studio:sidebarRight:get",t.sidebarRightToggle="studio:sidebarRight:toggle",t.sidebarTopSet="studio:sidebarTop:set",t.sidebarTopGet="studio:sidebarTop:get",t.sidebarTopToggle="studio:sidebarTop:toggle",t.sidebarBottomSet="studio:sidebarBottom:set",t.sidebarBottomGet="studio:sidebarBottom:get",t.sidebarBottomToggle="studio:sidebarBottom:toggle",t.symbolAdd="studio:symbolAdd",t.symbolDetach="studio:symbolDetach",t.symbolOverride="studio:symbolOverride",t.symbolPropagateStyles="studio:propagateStyles",t.getPagesConfig="studio:getPagesConfig",t.setPagesConfig="studio:setPagesConfig",t.getPageSettings="studio:getPageSettings",t.setPageSettings="studio:setPageSettings",t.clearPage="studio:clearPage",t.projectFiles="studio:projectFiles",t.validateCode="studio:validateCode",t.canvasReload="studio:canvasReload",t.getBlocksPanel="studio:getBlocksPanel",t.setBlocksPanel="studio:setBlocksPanel",t.getStateContextMenu="studio:getStateContextMenu",t.setStateContextMenu="studio:setStateContextMenu",t.contextMenuComponent="studio:contextMenuComponent",t.layoutAdd="studio:layoutAdd",t.layoutRemove="studio:layoutRemove",t.layoutToggle="studio:layoutToggle",t.layoutUpdate="studio:layoutUpdate",t.layoutGet="studio:layoutGet",t.layoutConfigGet="studio:layoutConfigGet",t.layoutConfigSet="studio:layoutConfigSet",t.layoutComponentAdd="studio:layoutComponentAdd",t.layoutComponentGet="studio:layoutComponentGet",t.layoutComponentRemove="studio:layoutComponentRemove",t.layoutComponentRender="studio:layoutComponentRender",t.getStateTheme="studio:getStateTheme",t.setStateTheme="studio:setStateTheme",t.settings="studio:settings",t.assetProviderGet="studio:assetProviderGet",t.assetProviderAdd="studio:assetProviderAdd",t.assetProviderRemove="studio:assetProviderRemove",t.fontGet="studio:fontGet",t.fontAdd="studio:fontAdd",t.fontRemove="studio:fontRemove",t.fontManagerOpen="studio:fontManagerOpen",t.menuFontLoad="studio:menuFontLoad",t.toggleStateDataSource="studio:toggleStateDataSource",t.getStateDataSource="studio:getStateDataSource",t.dataSourceSetGlobalData="studio:dataSourceSetGlobalData",t.dataSourceSetImporter="studio:dataSourceSetImporter",t.dataSourceSetExporter="studio:dataSourceSetExporter",t.setDragAbsolute="studio:setDragAbsolute",t))(w||{});const S={[v.free]:0,[v.startup]:10,[v.business]:20,[v.enterprise]:30};function j(t){const e=t;return e.init=a=>r=>t(r,a),e}const _=t=>j(t);async function $({editor:t,plan:e,pluginName:a,licenseKey:r,onLicenseCheckResponse:i,cleanup:o}){let l="",s;const p=n=>{console.warn("Cleanup plugin:",a,"Reason:",n),o()},d=(n={})=>{var E;const{error:c,sdkLicense:u}=n,h=(E=n.plan)==null?void 0:E.category;if(!(u||n.license)||c)p(c||"Invalid license");else if(h){const Z=S[e],at=S[h];Z>at&&p({pluginRequiredPlan:e,licensePlan:h})}};t.Commands.has(w.settings)&&(s=t.runCommand(w.settings),l=(s==null?void 0:s.baseUrl)||"");const f=n=>{i==null||i(n),n&&d(n)};if(!s){t.onReady(async()=>{if(!V())if(r){const n=await Y({licenseKey:r,pluginName:a,baseApiUrl:l});f(n)}else p("The `licenseKey` option not provided")});return}if(s.licensePlan||s.licenseError){const n=C(s);f(n);return}t.on(M,n=>f(n))}const C=t=>({sdkLicense:t.license,license:t.license,error:t.licenseError,plan:t.licensePlan});async function Y(t){const{licenseKey:e,pluginName:a,baseApiUrl:r}=t;try{return(await X({baseApiUrl:r,path:`/sdk/${e||"na"}`,method:"POST",params:{d:window.location.hostname,pn:a}})).result||{}}catch(i){return console.error("Error during SDK license check:",i),!1}}const W=(...t)=>e=>t.some(a=>e.is(a));var m=(t=>(t.name="animation-name",t.duration="animation-duration",t.delay="animation-delay",t.iterationCount="animation-iteration-count",t.timingFunction="animation-timing-function",t.fillMode="animation-fill-mode",t))(m||{}),x=(t=>(t.onScroll="--animation-on-scroll",t.threshold="--animation-threshold",t.repeat="--animation-repeat",t))(x||{});const g="animation",A="animation-group",L="animation:update",D="animation-group:update",U={category:{id:"animations",label:"Animations"},select:!0},P=t=>(t==null?void 0:t.getType())===g,R=t=>(t==null?void 0:t.getType())===A,G=t=>P(t)||R(t),k={isVisible:({component:t})=>P(t)},H={isVisible:({component:t})=>R(t)},K={isVisible:({component:t})=>P(t)&&R(t==null?void 0:t.parent())?!1:G(t)},O={isVisible:({component:t,sector:e})=>{var a;return P(t)&&R(t==null?void 0:t.parent())?!1:G(t)&&((a=e.getProperty(x.onScroll))==null?void 0:a.getValue())==="true"}},z={animation:'<svg viewBox="0 0 24 24"><path fill="currentColor" d="M16.5 14a6.5 6.5 0 1 0 0-13 6.5 6.5 0 0 0 0 13"/><path fill="currentColor" d="M17.14 15.98a8.5 8.5 0 0 1-9.12-9.12 6.5 6.5 0 1 0 9.12 9.12"/><path fill="currentColor" d="M12.64 20.48a8.5 8.5 0 0 1-9.12-9.12 6.5 6.5 0 1 0 9.12 9.12"/></svg>',animationGroup:'<svg viewBox="0 0 24 24"><path d="M4 2a2 2 0 0 0-2 2v10h2V4h10V2H4m4 4a2 2 0 0 0-2 2v10h2V8h10V6H8m4 4a2 2 0 0 0-2 2v8c0 1.1.9 2 2 2h8a2 2 0 0 0 2-2v-8a2 2 0 0 0-2-2h-8Z"/></svg>'},q=function(t){const e=this;let a=null;const r=()=>{a==null||a.disconnect(),a=null},i=()=>{r();const o=e.style;o.animationName="",requestAnimationFrame(()=>{const l=getComputedStyle(e),s=l.getPropertyValue("animation-name").trim(),p=l.getPropertyValue("--animation-on-scroll").trim()!=="false",d=parseFloat(l.getPropertyValue("--animation-threshold").trim())||.2,f=l.getPropertyValue("--animation-repeat").trim()==="true",n=()=>{o.animationName="none",e.offsetWidth,o.animationName=s,o.animationPlayState="running"};if(!s||s==="none"){o.animationName="none",o.animationPlayState="paused";return}if(!p){n();return}o.animationPlayState="paused",a=new IntersectionObserver(c=>{c.forEach(u=>{u.isIntersecting&&(n(),!f&&r())})},{threshold:d}),a.observe(e)})};e.addEventListener(t.animationUpdateEvent,i),i()},J=(t,e)=>{const{Components:a,Blocks:r,Canvas:i}=t,{block:o={},animationList:l=[]}=e,s=t.Components.events,p="Animation",d=Object.values(x),f=Object.values(m);return a.addType(g,{block:o?{label:p,media:z.animation,content:{type:g},...U,...o}:void 0,model:{defaults:{emptyState:!0,name:p,traits:[],script:q,animationUpdateEvent:L,"script-props":["animationUpdateEvent"],styles:l.map(n=>n.css).join(`
`)},init(){const{em:n}=this;this.addStyle({[m.name]:"fadeIn",[m.duration]:"1s",[m.fillMode]:"both",...e.animationStyle,...this.getStyle()}),this.listenTo(n,s.mount,c=>{var h;if(c!==this)return;((h=this.parent())==null?void 0:h.get("type"))!==A||d.forEach(y=>this.removeStyle(y))})}},view:{events:()=>({animationstart:"onAnimationStart",animationend:"onAnimationEnd"}),init(){const{model:n}=this;this.listenTo(n,s.styleUpdate,this.handleStyleChange)},onAnimationStart(){var c;if((c=this.cmpSpots)!=null&&c.length)return;const n=i.getSpots({component:this.model});this.cmpSpots=n,i.removeSpots(n),this.el.ownerDocument.body.style.overflowX="hidden"},onAnimationEnd(){const n=this.cmpSpots||[];i.spots.add(n),this.cmpSpots=[],this.el.ownerDocument.body.style.overflowX=""},triggerAnimation(){this.el.dispatchEvent(new CustomEvent(L))},handleStyleChange(n,c){Object.keys((c==null?void 0:c.style)||{}).some(y=>f.includes(y)||d.includes(y))&&this.triggerAnimation()},onRender(){this.triggerAnimation()}}}),()=>{r.remove(g),a.removeType(g)}},Q=function(t){const e=this,a="--animation-on-scroll",r="--animation-threshold",i="--animation-repeat";let o=null;const l=()=>{o==null||o.disconnect(),o=null},s=()=>{l();const p=getComputedStyle(e),d=p.getPropertyValue(a).trim()!=="false",f=parseFloat(p.getPropertyValue(r).trim())||.2,n=p.getPropertyValue(i).trim()==="true",c=()=>{Array.from(e.children).forEach(u=>{u instanceof HTMLElement&&typeof u.dispatchEvent=="function"&&u.dispatchEvent(new CustomEvent(D,{bubbles:!1}))})};if(!d){c();return}o=new IntersectionObserver(u=>{u.forEach(h=>{h.isIntersecting?(c(),!n&&l()):n&&Array.from(e.children).forEach(y=>{y instanceof HTMLElement&&getComputedStyle(y).getPropertyValue(a).trim()!=="false"&&(y.style.opacity="0",y.style.animationPlayState="paused")})})},{threshold:f}),o.observe(e)};e.addEventListener(t.animationGroupUpdateEvent,s),s()},tt=(t,e)=>{const{Components:a,Blocks:r}=t,{blockGroup:i={}}=e,o="Animation Group";return a.addType(A,{block:i?{label:o,media:z.animationGroup,content:{type:A,components:[{type:g},{type:g},{type:g}]},...U,...i}:void 0,model:{defaults:{name:o,emptyState:!0,traits:[],script:Q,animationGroupUpdateEvent:D,"script-props":["animationGroupUpdateEvent"],droppable:W(g)},init(){this.addStyle({...e.animationGroupStyle,...this.getStyle()});const l=this.components();this.listenTo(l,"add remove",this.updateChildrenStaggerStyles),this.updateChildrenStaggerStyles()},updateChildrenStaggerStyles(){this.components().forEach((l,s)=>{l.is(g)&&l.addStyle({"--stagger-index":s.toString(),[m.delay]:"calc(var(--stagger-index, 0) * var(--stagger-delay, 0.3s))"},{partial:!0})})}},view:{init(){const{model:l}=this;this.listenTo(l,"component:styleUpdate",this.handleStyleChange)},triggerAnimation(){this.el.dispatchEvent(new CustomEvent(D))},handleStyleChange(){this.triggerAnimation(),this.model.updateChildrenStaggerStyles()},onRender(){this.triggerAnimation()}}}),()=>{r.remove(A),a.removeType(A)}},et="animationComponent",nt=v.startup,B=[{id:"flash",name:"Flash",css:`@keyframes flash {
      0%, 50%, to {
          opacity:1
      }

      25%, 75% {
          opacity:0
      }
  }`},{id:"pulse",name:"Pulse",css:`@keyframes pulse {
      0% {
          transform:scaleX(1)
      }

      50% {
          transform:scale3d(1.05, 1.05, 1.05)
      }

      to {
          transform:scaleX(1)
      }
  }`},{id:"shake",name:"Shake",css:`@keyframes shake {
      0%, to {
          transform:translateZ(0)
      }

      10%, 30%, 50%, 70%, 90% {
          transform:translate3d(-10px, 0, 0)
      }

      20%, 40%, 60%, 80% {
          transform:translate3d(10px, 0, 0)
      }
  }`},{id:"tada",name:"Tada",css:`@keyframes tada {
      0% {
          transform:scaleX(1)
      }

      10%, 20% {
          transform:scale3d(.9, .9, .9) rotate(-3deg)
      }

      30%, 50%, 70%, 90% {
          transform:scale3d(1.1, 1.1, 1.1) rotate(3deg)
      }

      40%, 60%, 80% {
          transform:scale3d(1.1, 1.1, 1.1) rotate(-3deg)
      }

      to {
          transform:scaleX(1)
      }
  }`},{id:"heartBeat",name:"Heart Beat",css:`@keyframes heartBeat {
      0% {
          transform:scale(1)
      }

      14% {
          transform:scale(1.3)
      }

      28% {
          transform:scale(1)
      }

      42% {
          transform:scale(1.3)
      }

      70% {
          transform:scale(1)
      }
  }`},{id:"bounce",name:"Bounce",css:`@keyframes bounce {
      0%, 20%, 53%, 80%, to {
          animation-timing-function: cubic-bezier(.215, .61, .355, 1);
          transform:translateZ(0)
      }

      40%, 43% {
          animation-timing-function: cubic-bezier(.755, .05, .855, .06);
          transform:translate3d(0, -30px, 0)
      }

      70% {
          animation-timing-function: cubic-bezier(.755, .05, .855, .06);
          transform:translate3d(0, -15px, 0)
      }

      90% {
          transform:translate3d(0, -4px, 0)
      }
  }`},{id:"bounceIn",name:"Bounce In",css:`@keyframes bounceIn {
      0%, 20%, 40%, 60%, 80%, to {
          animation-timing-function:cubic-bezier(.215, .61, .355, 1)
      }

      0% {
          opacity: 0;
          transform:scale3d(.3, .3, .3)
      }

      20% {
          transform:scale3d(1.1, 1.1, 1.1)
      }

      40% {
          transform:scale3d(.9, .9, .9)
      }

      60% {
          opacity: 1;
          transform:scale3d(1.03, 1.03, 1.03)
      }

      80% {
          transform:scale3d(.97, .97, .97)
      }

      to {
          opacity: 1;
          transform:scaleX(1)
      }
  }`},{id:"bounceInDown",name:"Bounce In Down",css:`@keyframes bounceInDown {
      0%, 60%, 75%, 90%, to {
          animation-timing-function:cubic-bezier(.215, .61, .355, 1)
      }

      0% {
          opacity: 0;
          transform:translate3d(0, -100%, 0)
      }

      60% {
          opacity: 1;
          transform:translate3d(0, 25px, 0)
      }

      75% {
          transform:translate3d(0, -10px, 0)
      }

      90% {
          transform:translate3d(0, 5px, 0)
      }

      to {
          transform:translateZ(0)
      }
  }`},{id:"bounceInLeft",name:"Bounce In Left",css:`@keyframes bounceInLeft {
      0%, 1%, 60%, 75%, 90%, to {
          animation-timing-function:cubic-bezier(.215, .61, .355, .1)
      }

      0% {
          opacity: 0;
      }

      1% {
          transform:translate3d(-100%, 0, 0)
      }

      60% {
          opacity: 1;
          transform:translate3d(25px, 0, 0)
      }

      75% {
          transform:translate3d(-10px, 0, 0)
      }

      90% {
          transform:translate3d(5px, 0, 0)
      }

      to {
          transform:translateZ(0)
      }
  }`},{id:"bounceInRight",name:"Bounce In Right",css:`@keyframes bounceInRight {
      0%, 1%, 60%, 75%, 90%, to {
          animation-timing-function:cubic-bezier(.215, .61, .355, .1)
      }

      0% {
          opacity: 0;
      }

      1% {
          transform:translate3d(100%, 0, 0)
      }

      60% {
          opacity: 1;
          transform:translate3d(-25px, 0, 0)
      }

      75% {
          transform:translate3d(10px, 0, 0)
      }

      90% {
          transform:translate3d(-5px, 0, 0)
      }

      to {
          transform:translateZ(0)
      }
  }`},{id:"bounceInUp",name:"Bounce In Up",css:`@keyframes bounceInUp {
      0%, 1%, 60%, 75%, 90%, to {
          animation-timing-function:cubic-bezier(.215, .61, .355, .1)
      }

      0% {
          opacity: 0;
      }

      1% {
          transform:translate3d(0, 100%, 0)
      }

      60% {
          opacity: 1;
          transform:translate3d(0, -20px, 0)
      }

      75% {
          transform:translate3d(0, 10px, 0)
      }

      90% {
          transform:translate3d(0, -5px, 0)
      }

      to {
          transform:translateZ(0)
      }
  }`},{id:"fadeIn",name:"Fade In",css:`@keyframes fadeIn {
      0% {
          opacity:0
      }

      to {
          opacity:1
      }
  }`},{id:"fadeInDown",name:"Fade In Down",css:`@keyframes fadeInDown {
      0% {
          opacity: 0;
      }

      1% {
          transform:translate3d(0, -100%, 0)
      }

      to {
          opacity: 1;
          transform:translateZ(0)
      }
  }`},{id:"fadeInLeft",name:"Fade In Left",css:`@keyframes fadeInLeft {
      0% {
          opacity: 0;
      }

      1% {
          transform:translate3d(-100%, 0, 0)
      }

      to {
          opacity: 1;
          transform:translateZ(0)
      }
  }`},{id:"fadeInRight",name:"Fade In Right",css:`@keyframes fadeInRight {
      0% {
          opacity: 0;
      }

      1% {
          transform: translate3d(90%, 0, 0)
      }

      to {
          opacity: 1;
          transform:translateZ(0)
      }
  }`},{id:"fadeInUp",name:"Fade In Up",css:`@keyframes fadeInUp {
      0% {
          opacity: 0;
      }

      1% {
          transform:translate3d(0, 100%, 0)
      }

      to {
          opacity: 1;
          transform:translateZ(0)
      }
  }`},{id:"flipInX",name:"Flip In X",css:`@keyframes flipInX {
      0% {
          transform: perspective(400px) rotateX(90deg);
          animation-timing-function: ease-in;
          opacity:0
      }

      40% {
          transform: perspective(400px) rotateX(-20deg);
          animation-timing-function:ease-in
      }

      60% {
          transform: perspective(400px) rotateX(10deg);
          opacity:1
      }

      80% {
          transform:perspective(400px) rotateX(-5deg)
      }

      to {
          transform:perspective(400px)
      }
  }`},{id:"flipInY",name:"Flip In Y",css:`@keyframes flipInY {
      0% {
          transform: perspective(400px) rotateY(90deg);
          animation-timing-function: ease-in;
          opacity:0
      }

      40% {
          transform: perspective(400px) rotateY(-20deg);
          animation-timing-function:ease-in
      }

      60% {
          transform: perspective(400px) rotateY(10deg);
          opacity:1
      }

      80% {
          transform:perspective(400px) rotateY(-5deg)
      }

      to {
          transform:perspective(400px)
      }
  }`},{id:"rotateIn",name:"Rotate In",css:`@keyframes rotateIn {
      0% {
          transform-origin: center;
          transform: rotate(-200deg);
          opacity:0
      }

      to {
          transform-origin: center;
          transform: translateZ(0);
          opacity:1
      }
  }`},{id:"rotateInDownLeft",name:"Rotate In Down Left",css:`@keyframes rotateInDownLeft {
      0% {
          transform-origin: left bottom;
          transform: rotate(-45deg);
          opacity:0
      }

      to {
          transform-origin: left bottom;
          transform: translateZ(0);
          opacity:1
      }
  }`},{id:"rotateInDownRight",name:"Rotate In Down Right",css:`@keyframes rotateInDownRight {
      0% {
          transform-origin: right bottom;
          transform: rotate(45deg);
          opacity:0
      }

      to {
          transform-origin: right bottom;
          transform: translateZ(0);
          opacity:1
      }
  }`},{id:"rotateInUpLeft",name:"Rotate In Up Left",css:`@keyframes rotateInUpLeft {
      0% {
          opacity: 0
      }

      1% {
          transform-origin: left bottom;
          transform: rotate(45deg);
      }

      to {
          transform-origin: left bottom;
          transform: translateZ(0);
          opacity:1
      }
  }`},{id:"rotateInUpRight",name:"Rotate In Up Right",css:`@keyframes rotateInUpRight {
      0% {
          opacity: 0
      }

      1% {
          transform-origin: right bottom;
          transform: rotate(-90deg);
      }

      to {
          transform-origin: right bottom;
          transform: translateZ(0);
          opacity:1
      }
  }`},{id:"slideInDown",name:"Slide In Down",css:`@keyframes slideInDown {
      0% {
          visibility:visible
      }

      1% {
          transform: translate3d(0, -100%, 0);
      }

      to {
          transform:translateZ(0)
      }
  }`},{id:"slideInLeft",name:"Slide In Left",css:`@keyframes slideInLeft {
      0% {
          visibility:visible
      }

      1% {
          transform: translate3d(-100%, 0, 0);
      }

      to {
          transform: translateZ(0)
      }
  }`},{id:"slideInRight",name:"Slide In Right",css:`@keyframes slideInRight {
      0% {
          visibility:visible
      }

      1% {
          transform: translate3d(100%, 0, 0);
      }

      to {
          transform:translateZ(0)
      }
  }`},{id:"slideInUp",name:"Slide In Up",css:`@keyframes slideInUp {
      0% {
          visibility:visible
      }

      1% {
          transform: translate3d(0, 100%, 0);
      }

      to {
          transform:translateZ(0)
      }
  }`},{id:"zoomIn",name:"Zoom In",css:`@keyframes zoomIn {
      0% {
          opacity: 0;
          transform:scale3d(.3, .3, .3)
      }

      50% {
          opacity:1
      }
  }`},{id:"zoomInDown",name:"Zoom In Down",css:`@keyframes zoomInDown {
      0% {
          opacity: 0;
          transform: scale3d(.1, .1, .1) translate3d(0, -1000px, 0);
          animation-timing-function:cubic-bezier(.55, .055, .675, .19)
      }

      60% {
          opacity: 1;
          transform: scale3d(.475, .475, .475) translate3d(0, 60px, 0);
          animation-timing-function:cubic-bezier(.175, .885, .32, 1)
      }
  }`},{id:"zoomInLeft",name:"Zoom In Left",css:`@keyframes zoomInLeft {
      0% {
          opacity: 0;
          transform: scale3d(.1, .1, .1) translate3d(-1000px, 0, 0);
          animation-timing-function:cubic-bezier(.55, .055, .675, .19)
      }

      60% {
          opacity: 1;
          transform: scale3d(.475, .475, .475) translate3d(10px, 0, 0);
          animation-timing-function:cubic-bezier(.175, .885, .32, 1)
      }
  }`},{id:"zoomInRight",name:"Zoom In Right",css:`@keyframes zoomInRight {
      0% {
          opacity: 0;
          transform: scale3d(.1, .1, .1) translate3d(1000px, 0, 0);
          animation-timing-function:cubic-bezier(.55, .055, .675, .19)
      }

      60% {
          opacity: 1;
          transform: scale3d(.475, .475, .475) translate3d(-10px, 0, 0);
          animation-timing-function:cubic-bezier(.175, .885, .32, 1)
      }
  }`},{id:"zoomInUp",name:"Zoom In Up",css:`@keyframes zoomInUp {
      0% {
          opacity: 0;
          animation-timing-function: cubic-bezier(.55, .055, .675, .19);
      }

      1% {
          transform: scale3d(.1, .1, .1) translate3d(0, 100%, 0);
      }

      60% {
          opacity: 1;
          transform: scale3d(.475, .475, .475) translate3d(0, -60px, 0);
          animation-timing-function:cubic-bezier(.175, .885, .32, 1)
      }
  }`}];return _(function(t,e={}){var i;const a=((i=e.animations)==null?void 0:i.call(e,{items:B}))||B,r=[J(t,{...e,animationList:a}),tt(t,e)];t.Styles.addSector("animation",{name:"Animation",properties:[{...k,name:"Type",property:m.name,type:"select",full:!0,default:"fadeIn",options:a},{...k,name:"Duration",property:m.duration,type:"number",units:["s"],min:.1,max:10,step:.1,default:"1s"},{...k,name:"Delay",property:m.delay,type:"number",units:["s"],min:0,max:10,step:.1,default:"1s"},{...k,name:"Iteration Count",property:m.iterationCount,type:"select",default:"1",options:[{id:"1",name:"One time"},{id:"infinite",name:"Infinite"}]},{...k,name:"Timing",property:m.timingFunction,type:"select",default:"ease",options:[{id:"ease",name:"Ease"},{id:"ease-in",name:"Ease in"},{id:"ease-out",name:"Ease out"},{id:"ease-in-out",name:"Ease in out"},{id:"linear",name:"Linear"}]},{...H,name:"Stagger Delay",property:"--stagger-delay",type:"slider",units:["s"],min:.1,max:5,step:.1,default:"0.3s"},{...K,name:"Animate on Scroll",property:x.onScroll,type:"radio",default:"true",options:[{name:"Yes",id:"true"},{name:"No",id:"false"}]},{...O,name:"Scroll Threshold",property:x.threshold,type:"number",full:!0,min:.1,max:1,step:.01,default:"0.2"},{...O,name:"Repeat on Scroll",property:x.repeat,type:"radio",default:"false",options:[{name:"Yes",id:"true"},{name:"No",id:"false"}]}]}),$({editor:t,licenseKey:e.licenseKey,plan:nt,pluginName:et,cleanup:()=>{r.forEach(o=>o())}})})});
