// @grapesjs/studio-sdk-plugins/dist/canvasGridMode/index.es.js
const Y = "app.grapesjs.com", H = "app-stage.grapesjs.com", rt = "app2.grapesjs.com", st = "app-stage2.grapesjs.com", j = [
  Y,
  rt,
  H,
  st,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], at = "license:check:end", it = () => typeof window < "u", lt = ({
  isDev: t,
  isStage: e,
  isPlatform: r
}) => `${t ? "" : `https://${e ? H : Y}`}/${r ? "platform-api" : "api"}`, ct = () => {
  const t = it() && window.location.hostname;
  return !!t && (j.includes(t) || j.some((e) => t.endsWith(e)));
};
function X(t) {
  return typeof t == "function";
}
async function pt({
  path: t,
  baseApiUrl: e,
  method: r = "GET",
  headers: n = {},
  params: o,
  body: i
}) {
  const a = `${e || lt({ isDev: !1, isStage: !1 })}${t}`, p = {
    method: r,
    headers: {
      "Content-Type": "application/json",
      ...n
    }
  };
  i && (p.body = JSON.stringify(i));
  const d = o ? new URLSearchParams(o).toString() : "", c = d ? `?${d}` : "", l = await fetch(`${a}${c}`, p);
  if (!l.ok)
    throw new Error(`HTTP error! status: ${l.status}`);
  return l.json();
}
var G = /* @__PURE__ */ ((t) => (t.free = "free", t.startup = "startup", t.business = "business", t.enterprise = "enterprise", t))(G || {}), _ = /* @__PURE__ */ ((t) => (t.toastAdd = "studio:toastAdd", t.toastRemove = "studio:toastRemove", t.dialogOpen = "studio:dialogOpen", t.dialogClose = "studio:dialogClose", t.dialogExportCode = "studio:dialogExportCode", t.dialogImportCode = "studio:dialogImportCode", t.sidebarLeftSet = "studio:sidebarLeft:set", t.sidebarLeftGet = "studio:sidebarLeft:get", t.sidebarLeftToggle = "studio:sidebarLeft:toggle", t.sidebarRightSet = "studio:sidebarRight:set", t.sidebarRightGet = "studio:sidebarRight:get", t.sidebarRightToggle = "studio:sidebarRight:toggle", t.sidebarTopSet = "studio:sidebarTop:set", t.sidebarTopGet = "studio:sidebarTop:get", t.sidebarTopToggle = "studio:sidebarTop:toggle", t.sidebarBottomSet = "studio:sidebarBottom:set", t.sidebarBottomGet = "studio:sidebarBottom:get", t.sidebarBottomToggle = "studio:sidebarBottom:toggle", t.symbolAdd = "studio:symbolAdd", t.symbolDetach = "studio:symbolDetach", t.symbolOverride = "studio:symbolOverride", t.symbolPropagateStyles = "studio:propagateStyles", t.getPagesConfig = "studio:getPagesConfig", t.setPagesConfig = "studio:setPagesConfig", t.getPageSettings = "studio:getPageSettings", t.setPageSettings = "studio:setPageSettings", t.clearPage = "studio:clearPage", t.projectFiles = "studio:projectFiles", t.validateCode = "studio:validateCode", t.canvasReload = "studio:canvasReload", t.getBlocksPanel = "studio:getBlocksPanel", t.setBlocksPanel = "studio:setBlocksPanel", t.getStateContextMenu = "studio:getStateContextMenu", t.setStateContextMenu = "studio:setStateContextMenu", t.contextMenuComponent = "studio:contextMenuComponent", t.layoutAdd = "studio:layoutAdd", t.layoutRemove = "studio:layoutRemove", t.layoutToggle = "studio:layoutToggle", t.layoutUpdate = "studio:layoutUpdate", t.layoutGet = "studio:layoutGet", t.layoutConfigGet = "studio:layoutConfigGet", t.layoutConfigSet = "studio:layoutConfigSet", t.layoutComponentAdd = "studio:layoutComponentAdd", t.layoutComponentGet = "studio:layoutComponentGet", t.layoutComponentRemove = "studio:layoutComponentRemove", t.layoutComponentRender = "studio:layoutComponentRender", t.getStateTheme = "studio:getStateTheme", t.setStateTheme = "studio:setStateTheme", t.settings = "studio:settings", t.assetProviderGet = "studio:assetProviderGet", t.assetProviderAdd = "studio:assetProviderAdd", t.assetProviderRemove = "studio:assetProviderRemove", t.fontGet = "studio:fontGet", t.fontAdd = "studio:fontAdd", t.fontRemove = "studio:fontRemove", t.fontManagerOpen = "studio:fontManagerOpen", t.menuFontLoad = "studio:menuFontLoad", t.toggleStateDataSource = "studio:toggleStateDataSource", t.getStateDataSource = "studio:getStateDataSource", t.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", t.dataSourceSetImporter = "studio:dataSourceSetImporter", t.dataSourceSetExporter = "studio:dataSourceSetExporter", t.setDragAbsolute = "studio:setDragAbsolute", t))(_ || {});
const W = {
  [G.free]: 0,
  [G.startup]: 10,
  [G.business]: 20,
  [G.enterprise]: 30
};
function dt(t) {
  const e = t;
  return e.init = (r) => (n) => t(n, r), e;
}
const gt = (t) => /* @__PURE__ */ dt(t);
async function ut({
  editor: t,
  plan: e,
  pluginName: r,
  licenseKey: n,
  onLicenseCheckResponse: o,
  cleanup: i
}) {
  let s = "", a;
  const p = (l) => {
    console.warn("Cleanup plugin:", r, "Reason:", l), i();
  }, d = (l = {}) => {
    var w;
    const { error: g, sdkLicense: u } = l, h = (w = l.plan) == null ? void 0 : w.category;
    if (!(u || l.license) || g)
      p(g || "Invalid license");
    else if (h) {
      const v = W[e], R = W[h];
      v > R && p({ pluginRequiredPlan: e, licensePlan: h });
    }
  };
  t.Commands.has(_.settings) && (a = t.runCommand(_.settings), s = (a == null ? void 0 : a.baseUrl) || "");
  const c = (l) => {
    o == null || o(l), l && d(l);
  };
  if (!a) {
    t.onReady(async () => {
      if (!ct())
        if (n) {
          const l = await mt({ licenseKey: n, pluginName: r, baseApiUrl: s });
          c(l);
        } else
          p("The `licenseKey` option not provided");
    });
    return;
  }
  if (a.licensePlan || a.licenseError) {
    const l = ft(a);
    c(l);
    return;
  }
  t.on(at, (l) => c(l));
}
const ft = (t) => ({
  sdkLicense: t.license,
  license: t.license,
  error: t.licenseError,
  plan: t.licensePlan
});
async function mt(t) {
  const { licenseKey: e, pluginName: r, baseApiUrl: n } = t;
  try {
    return (await pt({
      baseApiUrl: n,
      path: `/sdk/${e || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: r
      }
    })).result || {};
  } catch (o) {
    return console.error("Error during SDK license check:", o), !1;
  }
}
const N = "canvasGridMode", yt = G.startup, F = "component-grid-mode", D = "grid-mode", M = "gjs-drag-grid-spot", K = `${M}__grid-item`, J = `${M}__grid-target`, Q = `${M}__snapped`, ht = `${M}__container`, wt = () => {
  const t = "data-gjs-drag-grid-style", e = document, r = "var(--gs-theme-cl-cmp-bg1, var(--gjs-color-highlight))", n = "var(--gs-theme-cl-cmp-bg2, var(--gjs-color-blue))", o = e.querySelector(`style[${t}]`) || e.createElement("style");
  o.setAttribute(t, "true"), o.innerHTML = `
        .${K} {
            outline: 2px solid ${n};
            background: ${r};
            outline-offset: -2px;
            border-radius: 3px;
            width: 100%;
            height: 100%;
            opacity: 0.1;
        }
        .${Q} {
            width: 100%;
            height: 100%;
            outline: 2px solid ${n};
            outline-offset: -2px;
            position: absolute;
        }
        .${J} {
            position: absolute;
            z-index: -1;
            background: ${r};
            outline: 1px solid ${n};
            outline-offset: -1px;
            opacity: 0.35;
        }
    `, e.head.appendChild(o);
};
function Z(t) {
  const { editor: e, type: r } = t, n = Rt(e, t.component), o = O(e, n.component);
  if (!Tt(e, n, o, r))
    return {
      target: n,
      container: o,
      startScroll: ot(e).scroll,
      shadowContainerEl: document.createElement("div"),
      shadowSnappedEl: document.createElement("div"),
      shadowTargetEl: document.createElement("div"),
      spot: e.Canvas.addSpot({
        type: D,
        component: o.component
      })
    };
}
function V(t) {
  const { editor: e, shadowTargetEl: r, shadowSnappedEl: n, shadowContainerEl: o, spotEl: i, target: s, container: a } = t, { Canvas: p } = e, d = e.config.stylePrefix;
  i == null || i.appendChild(o), o == null || o.appendChild(r), o == null || o.appendChild(n), bt(o, a), vt(r, s, a), $t(n, s), St(o, a.el), s.el.style.opacity = "0";
  const c = p.getSpots().filter((l) => l.type !== D);
  e.em.stopDefault({ preserveSelected: !0 }), p.removeSpots(c), p.startAutoscroll(), p.getBody().classList.add(`${d}is__grabbing`);
}
function tt(t) {
  const { editor: e, shadowSnappedEl: r, shadowContainerEl: n, target: o, cancelled: i } = t, { Canvas: s } = e, a = e.config.stylePrefix, { component: p } = o;
  if (!i) {
    const d = window.getComputedStyle(r);
    p.addStyle({
      "grid-area": d.gridArea,
      width: "",
      height: ""
    });
  }
  o.el.style.opacity = "", n.contains(r) && (n == null || n.removeChild(r)), e.getModel().runDefault({ preserveSelected: 1 }), s.stopAutoscroll(), s.getBody().classList.remove(`${a}is__grabbing`), s.removeSpots({ type: D }), e.select(p);
}
function bt(t, {
  rows: e,
  columns: r
}) {
  const n = [];
  for (let o = 0; o < r.length; o++) {
    n[o] = [];
    for (let i = 0; i < e.length; i++) {
      const s = document.createElement("div");
      s.className = K, s.style.gridArea = `${i + 1} / ${o + 1}`, n[o][i] = s, t.appendChild(s);
    }
  }
  return n;
}
function vt(t, e, r) {
  return t.className = J, t.style.top = `${e.offset.top - r.offset.top}px`, t.style.left = `${e.offset.left - r.offset.left}px`, t.style.width = `${e.offset.width + e.margin.left + e.margin.right}px`, t.style.height = `${e.offset.height + e.margin.top + e.margin.bottom}px`, t;
}
function $t(t, e) {
  const r = nt(e.el, e.computedStyles);
  return t.className = Q, t.style.gridArea = `${r.row.start} / ${r.column.start} / ${r.row.end} / ${r.column.end}`, t;
}
function St(t, e) {
  const r = window.getComputedStyle(e);
  t.className = ht, t.style.position = "relative", t.style.display = r.display, t.style.gap = r.gap, t.style.padding = r.padding, t.style.border = r.border, t.style.gridTemplateColumns = r.gridTemplateColumns, t.style.gridTemplateRows = r.gridTemplateRows;
}
function et(t, e) {
  if (e) {
    if (Array.isArray(e))
      return [...e].pop();
  } else return [...t.getSelectedAll()].pop();
  return e;
}
function Rt(t, e) {
  const r = et(t, e), n = r.getEl(), o = window.getComputedStyle(n);
  return {
    component: r,
    el: n,
    offset: t.Canvas.offset(n),
    margin: {
      top: parseFloat(o.marginTop),
      right: parseFloat(o.marginRight),
      bottom: parseFloat(o.marginBottom),
      left: parseFloat(o.marginLeft)
    },
    computedStyles: o
  };
}
function O(t, e) {
  const r = e.parent(), n = r.getEl(), o = window.getComputedStyle(n);
  return {
    component: r,
    el: n,
    offset: t.Canvas.offset(n),
    computedStyles: o,
    padding: {
      top: parseFloat(o.paddingTop),
      right: parseFloat(o.paddingRight),
      bottom: parseFloat(o.paddingBottom),
      left: parseFloat(o.paddingLeft)
    },
    columns: o.gridTemplateColumns.split(" ").map((s) => parseFloat(s)),
    rows: o.gridTemplateRows.split(" ").map((s) => parseFloat(s)),
    gap: {
      column: parseInt(o.columnGap),
      row: parseInt(o.rowGap)
    }
  };
}
function T(t, e, r, n, o = 1) {
  let i = 0;
  for (let s = e; s <= t.length; s++) {
    if (i + (t[s - 1] + n) * o > r)
      return s;
    i += n + t[s - 1];
  }
  return t.length + 1;
}
function ot(t) {
  const { scrollY: e, scrollX: r } = t.Canvas.getWindow();
  return {
    scroll: {
      top: e,
      left: r
    }
  };
}
function Tt(t, e, r, n) {
  const o = (i) => t.em.logWarning(i);
  return n === "draggable" && e.component.get(n) === !1 ? (o(`[${N}] The element is not ${n}`), !0) : r.computedStyles.display !== "grid" ? (o(`[${N}] The container does not have style display:grid`), !0) : !1;
}
function L(t, e, r = {}) {
  const n = et(t, e), o = n == null ? void 0 : n.parent(), i = o == null ? void 0 : o.getEl();
  if (!o || !i)
    return !1;
  const { enableGrid: s } = r, a = window.getComputedStyle(i).display === "grid";
  return X(s) ? s({ component: n, parent: o, isParentGrid: a }) : a;
}
const U = (t) => t.split(" ").map((e) => parseFloat(e));
function nt(t, e) {
  const r = t.ownerDocument.defaultView, n = e || r.getComputedStyle(t), o = n.gridArea.split("/").some((b) => isNaN(parseInt(b.trim()))), i = {
    isComputed: o,
    raw: n.gridArea
  };
  if (!o)
    return {
      ...i,
      row: {
        start: parseInt(n.gridRowStart),
        end: parseInt(n.gridRowEnd)
      },
      column: {
        start: parseInt(n.gridColumnStart),
        end: parseInt(n.gridColumnEnd)
      }
    };
  const s = t.parentElement, a = r.getComputedStyle(s), p = parseFloat(a.columnGap) || 0, d = parseFloat(a.rowGap) || 0, c = U(a.gridTemplateColumns), l = U(a.gridTemplateRows), g = parseFloat(a.paddingLeft) || 0, u = parseFloat(a.paddingTop) || 0, h = s.getBoundingClientRect(), m = t.getBoundingClientRect(), w = (b, $, x, S, C) => {
    let E = 1, P = S + C;
    for (let A = 0; A < $.length; A++) {
      const I = $[A], z = P + I;
      if (b >= P && b < z)
        return E;
      P = z + x, E++;
    }
    return -1;
  }, v = w(m.left, c, p, h.left, g), R = w(m.right - 1, c, p, h.left, g) + 1, y = w(m.top, l, d, h.top, u), f = w(m.bottom - 1, l, d, h.top, u) + 1;
  return {
    ...i,
    row: {
      start: y,
      end: f
    },
    column: {
      start: v,
      end: R
    }
  };
}
function xt(t) {
  t.Commands.add(F, {
    stop() {
    },
    run(e, r, n = {}) {
      const o = Z({ editor: e, component: n.target, type: "draggable" });
      if (!o) {
        this.stopCommand();
        return;
      }
      const { spot: i, target: s, shadowContainerEl: a, shadowSnappedEl: p, shadowTargetEl: d, startScroll: c } = o;
      let { container: l } = o, g;
      new e.Utils.Dragger({
        doc: s.el.ownerDocument,
        onStart: () => {
          g = setTimeout(() => {
            V({
              editor: e,
              shadowTargetEl: d,
              shadowSnappedEl: p,
              shadowContainerEl: a,
              spotEl: i.attributes.spotEl,
              target: s,
              container: l
            }), l = O(e, s.component);
          }, 10);
        },
        onEnd: (h, m, { cancelled: w }) => {
          clearTimeout(g), tt({
            editor: e,
            shadowSnappedEl: p,
            shadowContainerEl: a,
            target: s,
            cancelled: w
          }), this.stopCommand(), e.em.set("_cmpDrag", 1);
        },
        setPosition: ({ x: h, y: m }) => {
          const { scroll: w } = ot(e), { columns: v, rows: R, gap: y, offset: f, padding: b } = l, { offset: $, margin: x } = s, S = $.top - f.top - b.top - x.top + m - c.top + w.top, C = $.left - f.left - b.left - x.left + h - c.left + w.left, E = Math.min(T(v, 1, C, y.column, 0.5), v.length), P = Math.min(T(R, 1, S, y.row, 0.5), R.length), A = Math.min(
            T(v, E, $.width, y.column) + 1,
            v.length + 1
          ), I = Math.min(T(R, P, $.height, y.row) + 1, R.length + 1);
          p.style.gridArea = `${P} / ${E} / ${I} / ${A}`, d.style.top = `${S + b.top}px`, d.style.left = `${C + b.left}px`;
        }
      }).start(n.event);
    }
  });
}
function Pt(t, e) {
  const r = Z({ editor: t, component: e.component, type: "resizable" });
  if (!r)
    return;
  const { onMove: n, onEnd: o } = e.options || {}, { spot: i, target: s, shadowContainerEl: a, shadowSnappedEl: p, shadowTargetEl: d } = r;
  let { container: c } = r, l = "", g;
  return {
    onStart: (u) => {
      l = u.target.dataset.gjsHandler || "", g = setTimeout(() => {
        V({
          editor: t,
          shadowTargetEl: d,
          shadowSnappedEl: p,
          shadowContainerEl: a,
          spotEl: i.attributes.spotEl,
          target: s,
          container: c
        }), c = O(t, s.component);
      }, 10), t.trigger("component:resize", { ...s, type: "start" });
    },
    onEnd: (u, h) => {
      const { startDim: m, rectDim: w } = h.resizer, v = JSON.stringify(m) === JSON.stringify(w);
      clearTimeout(g), tt({ editor: t, shadowSnappedEl: p, shadowContainerEl: a, target: s, cancelled: v }), o == null || o(u, h);
    },
    onMove: (u, h) => {
      const m = nt(p), w = s.margin.left, v = s.margin.top, R = s.margin.bottom, y = s.margin.right, { w: f, t: b, l: $, h: x } = h.resizer.rectDim, S = {
        t: b - v,
        l: $ - w,
        h: x + v + R,
        w: f + w + y
      };
      l.endsWith("l") && (m.column.start = T(
        c.columns,
        1,
        S.l - c.offset.left - c.padding.left,
        c.gap.column,
        0.5
      )), l.startsWith("t") && (m.row.start = T(
        c.rows,
        1,
        S.t - c.offset.top - c.padding.top,
        c.gap.row,
        0.5
      )), l.endsWith("r") && (m.column.end = Math.max(
        T(c.columns, m.column.start, S.w, c.gap.column, 0.5),
        m.column.start + 1
      )), l.startsWith("b") && (m.row.end = Math.max(
        T(c.rows, m.row.start, S.h, c.gap.row, 0.5),
        m.row.start + 1
      )), p.style.gridArea = `${m.row.start} / ${m.column.start} / ${m.row.end} / ${m.column.end}`, d.style.width = `${S.w}px`, d.style.height = `${S.h}px`, d.style.left = `${S.l - c.offset.left}px`, d.style.top = `${S.t - c.offset.top}px`, n == null || n(u, h);
    }
  };
}
const k = ["px", "%", "em", "rem", "vw", "vh"], Gt = (t, e) => {
  const r = t.Styles.getSectors().find((n) => !!n.getProperty(e));
  return {
    sector: r,
    property: r ? r.getProperty(e) : void 0
  };
}, Et = (t, e) => Gt(t, e), At = (t, { x: e, y: r, mergable: n } = {}) => {
  const o = {
    property: `${t}-x`,
    type: "integer",
    units: k,
    ...e
  }, i = {
    property: `${t}-y`,
    type: "integer",
    units: k,
    ...r
  };
  return {
    property: t,
    type: "composite",
    properties: [o, i],
    ...n && {
      fromStyle(s, { name: a, separator: p, property: d }) {
        const [c, l] = d.getProperties(), [g, u] = (s[a] || "").split(p);
        return {
          [c.getId()]: s[c.getName()] || g || "",
          [l.getId()]: s[l.getName()] || u || g || ""
        };
      },
      toStyle(s, { name: a, property: p }) {
        const [d, c] = p.getProperties(), l = s[d.getId()], g = s[c.getId()];
        return {
          [a]: l === g ? l : `${l} ${g}`
        };
      }
    }
  };
}, B = { display: ["grid"] };
function Dt(t = "") {
  const e = /^repeat\(\s*(\d+)\s*,\s*(.+)\s*\)$/i, r = /^minmax\(\s*(.+?)\s*,\s*(.+?)\s*\)$/i, n = t.match(e);
  if (!n)
    return null;
  const o = parseInt(n[1], 10), i = n[2];
  let s = i, a = i;
  const p = i.match(r);
  return p && (s = p[1], a = p[2]), {
    total: o,
    min: s === a ? "" : s,
    max: a
  };
}
function Mt(t = {}) {
  let e = "", r = "", n = "", o = "";
  const i = (m) => m.split("/").map((w) => w.trim()).filter(Boolean), s = t["grid-area"] || "", a = t["grid-column"] || "", p = t["grid-row"] || "", d = i(s), c = i(a), l = i(p), g = d.length, u = c.length, h = l.length;
  return g === 4 ? (n = d[0], e = d[1], o = d[2], r = d[3]) : g === 3 ? (n = d[0], e = d[1], o = d[2]) : g === 2 ? (n = d[0], e = d[1]) : g === 1 && (n = d[0]), u === 2 ? (e = c[0], r = c[1]) : u === 1 && (e = c[0]), h === 2 ? (n = l[0], o = l[1]) : h === 1 && (n = l[0]), {
    columnStart: e,
    columnEnd: r,
    rowStart: n,
    rowEnd: o
  };
}
function q(t) {
  const { property: e, label: r } = t, n = "1", o = "1fr";
  return {
    property: e,
    label: r,
    type: "composite",
    requires: B,
    generic: !0,
    properties: [
      {
        name: " ",
        property: `${e}-repeat`,
        type: "number",
        full: !0,
        min: 1,
        default: "1"
      },
      {
        label: "Min size",
        property: `${e}-min`,
        type: "number",
        min: 0,
        units: ["px", "%"],
        default: "0"
      },
      {
        label: "Max size",
        property: `${e}-max`,
        type: "number",
        min: 1,
        units: ["fr", "px", "%"],
        default: "1"
      }
    ],
    fromStyle(i, s) {
      const { name: a, property: p } = s, [d, c, l] = p.getProperties(), g = i[a], u = Dt(g), h = (u == null ? void 0 : u.max) ?? o;
      return {
        [d.id]: (u == null ? void 0 : u.total) ?? n,
        [c.id]: (u == null ? void 0 : u.min) ?? "",
        [l.id]: h === o ? "" : h
      };
    },
    toStyle(i, { name: s, property: a }) {
      const [p, d, c] = a.getProperties(), l = i[p.id], g = i[d.id], u = i[c.id] || o, h = g && u ? `minmax(${g}, ${u})` : u;
      return {
        [s]: `repeat(${l || n}, ${h})`
      };
    }
  };
}
function Ct(t, e) {
  if (!e.styleableGrid) return;
  const { property: r, ...n } = Et(t, "display");
  if (!r) return;
  const o = r.getOption("grid"), i = n.sector;
  o || r.addOption({ id: "grid", label: "Grid" }), i.addProperty(
    q({
      label: "Columns",
      property: "grid-template-columns"
    }),
    {}
  ), i.addProperty(
    q({
      label: "Rows",
      property: "grid-template-rows"
    }),
    {}
  ), i.addProperty(
    {
      id: "grid-gap",
      requires: B,
      ...At("gap", {
        x: { id: "grid-row-gap", label: "Row", property: "row-gap", min: 0, default: "0" },
        y: { id: "grid-column-gap", label: "Column", property: "column-gap", min: 0, default: "0" },
        mergable: !0
      })
    },
    {}
  ), i.addProperty(
    {
      type: "composite",
      property: "grid-area",
      label: "Grid area",
      requiresParent: B,
      //   detached: true,
      generic: !0,
      properties: [
        {
          label: "Row start",
          property: "grid-row-start",
          type: "integer",
          default: "auto"
        },
        {
          label: "Row end",
          property: "grid-row-end",
          type: "integer",
          default: "auto"
        },
        {
          label: "Column start",
          property: "grid-column-start",
          type: "integer",
          default: "auto"
        },
        {
          label: "Column end",
          property: "grid-column-end",
          type: "integer",
          default: "auto"
        }
      ],
      fromStyle(s) {
        const a = Mt(s);
        return {
          "grid-column-start": s["grid-column-start"] || a.columnStart,
          "grid-column-end": s["grid-column-end"] || a.columnEnd,
          "grid-row-start": s["grid-row-start"] || a.rowStart,
          "grid-row-end": s["grid-row-end"] || a.rowEnd
        };
      },
      toStyle(s, { name: a }) {
        const p = s["grid-row-start"] || "auto", d = s["grid-row-end"] || "auto", c = s["grid-column-start"] || "auto", l = s["grid-column-end"] || "auto";
        return {
          [a]: `${p} / ${c} / ${d} / ${l}`
        };
      }
    },
    {}
  );
}
const It = function(t, e = {}) {
  const r = {
    itemResizable: !0,
    ...e
  }, { Canvas: n, Commands: o, Components: i } = t, { itemResizable: s } = r, a = n.events, p = o.events, d = i.events, c = `${p.runBeforeCommand}tlb-move`, l = `${p.runBeforeCommand}resize`;
  let g;
  xt(t);
  const u = ({ options: y }) => {
    L(t, y.target, r) && (t.runCommand(F, { ...y }), y.abort = !0);
  }, h = (y) => {
    const { options: f } = y;
    L(t, f.component, r) && (f.options = Object.assign(f.options || {}, Pt(t, f)));
  }, m = () => {
    const y = n.getSpots({
      type: D
    }).pop();
    if (!y) return;
    let f = y.attributes.spotEl;
    f || (f = document.createElement("div"), f.className = M, y.set({ spotEl: f })), Object.entries(y.getStyle()).forEach(([b, $]) => f.style.setProperty(b, $)), g == null || g.appendChild(f);
  }, w = ({ spot: y }) => {
    const f = y.attributes.spotEl;
    y.type !== D || !f || g == null || g.removeChild(f);
  };
  if (s) {
    const y = (f) => {
      const b = X(s) ? s({ component: f }) : s;
      return typeof b == "boolean" ? b : {
        tl: !1,
        tc: !1,
        tr: !1,
        cl: !1,
        cr: !1,
        bl: !1,
        bc: !1,
        br: !1,
        ...b
      };
    };
    t.on(d.resizeInit, (f) => {
      const { component: b } = f;
      if (L(t, b, r)) {
        const $ = y(b);
        f.resizable = $;
      }
    });
  }
  const v = [
    [c, u],
    [l, h],
    [a.spot, m],
    [a.spotRemove, w]
  ];
  v.forEach(([y, f]) => {
    t.on(y, f);
  }), t.onReady(() => {
    g = n.getSpotsEl(), wt(), Ct(t, r);
  });
  const R = () => {
    delete t.Commands.commands[F], v.forEach(([y, f]) => {
      t.off(y, f);
    });
  };
  ut({
    editor: t,
    licenseKey: r.licenseKey,
    plan: yt,
    pluginName: N,
    cleanup: R
  });
}, Lt = gt(It);
export {
  Lt as default
};
