// @grapesjs/studio-sdk-plugins/dist/googleFontsAssetProvider/index.es.js
var d = /* @__PURE__ */ ((e) => (e.free = "free", e.startup = "startup", e.business = "business", e.enterprise = "enterprise", e))(d || {}), y = /* @__PURE__ */ ((e) => (e.toastAdd = "studio:toastAdd", e.toastRemove = "studio:toastRemove", e.dialogOpen = "studio:dialogOpen", e.dialogClose = "studio:dialogClose", e.dialogExportCode = "studio:dialogExportCode", e.dialogImportCode = "studio:dialogImportCode", e.sidebarLeftSet = "studio:sidebarLeft:set", e.sidebarLeftGet = "studio:sidebarLeft:get", e.sidebarLeftToggle = "studio:sidebarLeft:toggle", e.sidebarRightSet = "studio:sidebarRight:set", e.sidebarRightGet = "studio:sidebarRight:get", e.sidebarRightToggle = "studio:sidebarRight:toggle", e.sidebarTopSet = "studio:sidebarTop:set", e.sidebarTopGet = "studio:sidebarTop:get", e.sidebarTopToggle = "studio:sidebarTop:toggle", e.sidebarBottomSet = "studio:sidebarBottom:set", e.sidebarBottomGet = "studio:sidebarBottom:get", e.sidebarBottomToggle = "studio:sidebarBottom:toggle", e.symbolAdd = "studio:symbolAdd", e.symbolDetach = "studio:symbolDetach", e.symbolOverride = "studio:symbolOverride", e.symbolPropagateStyles = "studio:propagateStyles", e.getPagesConfig = "studio:getPagesConfig", e.setPagesConfig = "studio:setPagesConfig", e.getPageSettings = "studio:getPageSettings", e.setPageSettings = "studio:setPageSettings", e.clearPage = "studio:clearPage", e.projectFiles = "studio:projectFiles", e.validateCode = "studio:validateCode", e.canvasReload = "studio:canvasReload", e.getBlocksPanel = "studio:getBlocksPanel", e.setBlocksPanel = "studio:setBlocksPanel", e.getStateContextMenu = "studio:getStateContextMenu", e.setStateContextMenu = "studio:setStateContextMenu", e.contextMenuComponent = "studio:contextMenuComponent", e.layoutAdd = "studio:layoutAdd", e.layoutRemove = "studio:layoutRemove", e.layoutToggle = "studio:layoutToggle", e.layoutUpdate = "studio:layoutUpdate", e.layoutGet = "studio:layoutGet", e.layoutConfigGet = "studio:layoutConfigGet", e.layoutConfigSet = "studio:layoutConfigSet", e.layoutComponentAdd = "studio:layoutComponentAdd", e.layoutComponentGet = "studio:layoutComponentGet", e.layoutComponentRemove = "studio:layoutComponentRemove", e.layoutComponentRender = "studio:layoutComponentRender", e.getStateTheme = "studio:getStateTheme", e.setStateTheme = "studio:setStateTheme", e.settings = "studio:settings", e.assetProviderGet = "studio:assetProviderGet", e.assetProviderAdd = "studio:assetProviderAdd", e.assetProviderRemove = "studio:assetProviderRemove", e.fontGet = "studio:fontGet", e.fontAdd = "studio:fontAdd", e.fontRemove = "studio:fontRemove", e.fontManagerOpen = "studio:fontManagerOpen", e.menuFontLoad = "studio:menuFontLoad", e.toggleStateDataSource = "studio:toggleStateDataSource", e.getStateDataSource = "studio:getStateDataSource", e.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", e.dataSourceSetImporter = "studio:dataSourceSetImporter", e.dataSourceSetExporter = "studio:dataSourceSetExporter", e.setDragAbsolute = "studio:setDragAbsolute", e))(y || {});
const x = "app.grapesjs.com", T = "app-stage.grapesjs.com", I = "app2.grapesjs.com", $ = "app-stage2.grapesjs.com", w = [
  x,
  I,
  T,
  $,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], E = "license:check:end", M = () => typeof window < "u", U = ({
  isDev: e,
  isStage: t,
  isPlatform: s
}) => `${e ? "" : `https://${t ? T : x}`}/${s ? "platform-api" : "api"}`, j = () => {
  const e = M() && window.location.hostname;
  return !!e && (w.includes(e) || w.some((t) => e.endsWith(t)));
};
async function k({
  path: e,
  baseApiUrl: t,
  method: s = "GET",
  headers: o = {},
  params: n,
  body: i
}) {
  const r = `${t || U({ isDev: !1, isStage: !1 })}${e}`, c = {
    method: s,
    headers: {
      "Content-Type": "application/json",
      ...o
    }
  };
  i && (c.body = JSON.stringify(i));
  const p = n ? new URLSearchParams(n).toString() : "", u = p ? `?${p}` : "", a = await fetch(`${r}${u}`, c);
  if (!a.ok)
    throw new Error(`HTTP error! status: ${a.status}`);
  return a.json();
}
const F = {
  [d.free]: 0,
  [d.startup]: 10,
  [d.business]: 20,
  [d.enterprise]: 30
};
function B(e) {
  const t = e;
  return t.init = (s) => (o) => e(o, s), t;
}
const N = (e) => /* @__PURE__ */ B(e);
async function W({
  editor: e,
  plan: t,
  pluginName: s,
  licenseKey: o,
  onLicenseCheckResponse: n,
  cleanup: i
}) {
  let l = "", r;
  const c = (a) => {
    console.warn("Cleanup plugin:", s, "Reason:", a), i();
  }, p = (a = {}) => {
    var A;
    const { error: g, sdkLicense: f } = a, b = (A = a.plan) == null ? void 0 : A.category;
    if (!(f || a.license) || g)
      c(g || "Invalid license");
    else if (b) {
      const G = F[t], O = F[b];
      G > O && c({ pluginRequiredPlan: t, licensePlan: b });
    }
  };
  e.Commands.has(y.settings) && (r = e.runCommand(y.settings), l = (r == null ? void 0 : r.baseUrl) || "");
  const u = (a) => {
    n == null || n(a), a && p(a);
  };
  if (!r) {
    e.onReady(async () => {
      if (!j())
        if (o) {
          const a = await V({ licenseKey: o, pluginName: s, baseApiUrl: l });
          u(a);
        } else
          c("The `licenseKey` option not provided");
    });
    return;
  }
  if (r.licensePlan || r.licenseError) {
    const a = _(r);
    u(a);
    return;
  }
  e.on(E, (a) => u(a));
}
const _ = (e) => ({
  sdkLicense: e.license,
  license: e.license,
  error: e.licenseError,
  plan: e.licensePlan
});
async function V(e) {
  const { licenseKey: t, pluginName: s, baseApiUrl: o } = e;
  try {
    return (await k({
      baseApiUrl: o,
      path: `/sdk/${t || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: s
      }
    })).result || {};
  } catch (n) {
    return console.error("Error during SDK license check:", n), !1;
  }
}
const v = (e) => e.replaceAll('"', "").replaceAll("'", ""), P = (e) => e.includes(",") ? e : /\s/.test(e) ? `'${v(e)}'` : v(e), L = (e, t, s) => ({ family: e, variant: t, ...s }), D = (e, t) => {
  e = P(e);
  const { menuVariant: s, variants: o } = t;
  return {
    family: e,
    familyLabel: v(e),
    menuVariant: s && L(
      P(`menuFont-${v(e)}`),
      "regular",
      s
    ),
    variants: Object.assign(
      {},
      ...Object.entries(o).map(([n, i]) => ({
        [n]: L(e, n, i)
      }))
    )
  };
}, R = (e, t) => {
  var r;
  const s = D(e.family, e), o = s.menuVariant ?? ((r = Object.values(s.variants)) == null ? void 0 : r[0]);
  if (!o)
    return "Arial, Helvetica, sans-serif";
  const { source: n, descriptors: i } = o, l = v(o.family);
  if (!t) {
    const c = new FontFace(l, `url("${n}")`, i), p = Array.from(document.fonts.values()).filter((u) => u.family === l);
    for (const u of p)
      document.fonts.delete(u);
    document.fonts.add(c), c.load().catch(() => {
    });
  }
  return P(l);
}, K = (e) => {
  const { family: t, menu: s, axes: o } = e, n = !!o, i = Object.assign(
    {},
    ...e.variants.map((l) => {
      var u, a;
      let r, c;
      if (n) {
        const g = o.find((f) => f.tag === "wght");
        if (g) {
          const { start: f, end: b } = g;
          r = `${f} ${b}`;
        }
      }
      return r || (r = ((u = l.match(/\d+/)) == null ? void 0 : u[0]) ?? "400"), c = ((a = l.match(/[a-zA-Z]+/)) == null ? void 0 : a[0]) === "italic" ? "italic" : void 0, {
        [l]: {
          source: e.files[l],
          descriptors: {
            style: c,
            weight: r
          }
        }
      };
    })
  );
  return D(t, {
    menuVariant: {
      source: s
    },
    variants: i
  });
}, z = (e) => ({
  id: `${e.family}/${e.version}`,
  type: "font",
  src: e.menu,
  name: e.family,
  customData: { font: K(e), needsLoading: !0 }
}), H = (e) => {
  const { font: t } = e.customData;
  return e.customData.needsLoading ? (e.customData.needsLoading = !1, R(t)) : R(t, !0);
};
let h;
const q = (e) => {
  const t = [];
  for (const s of Object.keys(e)) {
    let o = e[s];
    o && (Array.isArray(o) ? t.push(...o.map((n) => [s, n.toString()])) : (o = o.toString(), t.push([s, o])));
  }
  return new URLSearchParams(t);
}, J = (e) => {
  const { apiKey: t, searchParams: s } = e, o = (n = {}) => {
    const i = "https://www.googleapis.com/webfonts/v1/webfonts", l = q(n);
    return `${i}?${l}`;
  };
  return {
    id: "google-fonts",
    types: ["font"],
    label: ({ editor: n }) => n.I18n.t("googleFontsAssetProvider.providerLabel"),
    search: { reloadOnInput: !1 },
    async onLoad() {
      var p;
      if (h)
        return h;
      const n = {
        key: t,
        sort: "alpha",
        subset: "latin",
        capability: ["WOFF2", "VF"]
      }, i = s == null ? void 0 : s(), l = {
        ...n,
        ...i
      }, c = await (await fetch(o(l))).json();
      return (p = c.items) != null && p.length ? (h = c.items.map(z), h) : [];
    },
    itemLayout: (n) => {
      const i = h.find((r) => r.id === n.assetProps.id), l = H(i);
      return {
        type: "column",
        onClick: () => n.onSelect(i),
        htmlAttrs: {
          title: i.name
        },
        style: {
          borderWidth: "1px",
          borderRadius: 8,
          overflow: "hidden",
          cursor: "pointer"
        },
        children: [
          {
            type: "text",
            content: i.name,
            style: {
              overflow: "hidden",
              textOverflow: "ellipsis",
              whiteSpace: "nowrap",
              fontFamily: l,
              fontSize: "24px",
              paddingLeft: "8px",
              paddingRight: "8px"
            }
          },
          {
            type: "text",
            content: i.name,
            style: {
              overflow: "hidden",
              textOverflow: "ellipsis",
              whiteSpace: "nowrap",
              fontSize: "16px",
              paddingLeft: "8px",
              paddingRight: "8px"
            }
          }
        ]
      };
    }
  };
}, Q = {
  googleFontsAssetProvider: {
    searchBtn: "Search on Google Fonts",
    modalTitle: "Select Font",
    providerLabel: "Google Fonts"
  }
}, Z = "googleFontsAssetProvider", X = d.startup, Y = function(e, t) {
  const { i18n: s = {} } = t, o = J(t);
  e.runCommand(y.assetProviderAdd, { provider: o }), e.I18n.addMessages({
    en: Q,
    ...s
  }), W({
    editor: e,
    licenseKey: t.licenseKey,
    plan: X,
    pluginName: Z,
    cleanup: () => {
      e.runCommand(y.assetProviderRemove, { id: o.id });
    }
  });
}, m = N(Y);
export {
  m as default
};
