// @grapesjs/studio-sdk-plugins/dist/canvasEmptyState/index.es.js
const O = "app.grapesjs.com", x = "app-stage.grapesjs.com", k = "app2.grapesjs.com", B = "app-stage2.grapesjs.com", I = [
  O,
  k,
  x,
  B,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], N = "license:check:end", W = () => typeof window < "u", $ = ({
  isDev: e,
  isStage: o,
  isPlatform: n
}) => `${e ? "" : `https://${o ? x : O}`}/${n ? "platform-api" : "api"}`, j = () => {
  const e = W() && window.location.hostname;
  return !!e && (I.includes(e) || I.some((o) => e.endsWith(o)));
};
function F(e) {
  return typeof e == "function";
}
async function _({
  path: e,
  baseApiUrl: o,
  method: n = "GET",
  headers: i = {},
  params: a,
  body: g
}) {
  const r = `${o || $({ isDev: !1, isStage: !1 })}${e}`, p = {
    method: n,
    headers: {
      "Content-Type": "application/json",
      ...i
    }
  };
  g && (p.body = JSON.stringify(g));
  const y = a ? new URLSearchParams(a).toString() : "", f = y ? `?${y}` : "", s = await fetch(`${r}${f}`, p);
  if (!s.ok)
    throw new Error(`HTTP error! status: ${s.status}`);
  return s.json();
}
var P = /* @__PURE__ */ ((e) => (e.free = "free", e.startup = "startup", e.business = "business", e.enterprise = "enterprise", e))(P || {}), M = /* @__PURE__ */ ((e) => (e.toastAdd = "studio:toastAdd", e.toastRemove = "studio:toastRemove", e.dialogOpen = "studio:dialogOpen", e.dialogClose = "studio:dialogClose", e.dialogExportCode = "studio:dialogExportCode", e.dialogImportCode = "studio:dialogImportCode", e.sidebarLeftSet = "studio:sidebarLeft:set", e.sidebarLeftGet = "studio:sidebarLeft:get", e.sidebarLeftToggle = "studio:sidebarLeft:toggle", e.sidebarRightSet = "studio:sidebarRight:set", e.sidebarRightGet = "studio:sidebarRight:get", e.sidebarRightToggle = "studio:sidebarRight:toggle", e.sidebarTopSet = "studio:sidebarTop:set", e.sidebarTopGet = "studio:sidebarTop:get", e.sidebarTopToggle = "studio:sidebarTop:toggle", e.sidebarBottomSet = "studio:sidebarBottom:set", e.sidebarBottomGet = "studio:sidebarBottom:get", e.sidebarBottomToggle = "studio:sidebarBottom:toggle", e.symbolAdd = "studio:symbolAdd", e.symbolDetach = "studio:symbolDetach", e.symbolOverride = "studio:symbolOverride", e.symbolPropagateStyles = "studio:propagateStyles", e.getPagesConfig = "studio:getPagesConfig", e.setPagesConfig = "studio:setPagesConfig", e.getPageSettings = "studio:getPageSettings", e.setPageSettings = "studio:setPageSettings", e.clearPage = "studio:clearPage", e.projectFiles = "studio:projectFiles", e.validateCode = "studio:validateCode", e.canvasReload = "studio:canvasReload", e.getBlocksPanel = "studio:getBlocksPanel", e.setBlocksPanel = "studio:setBlocksPanel", e.getStateContextMenu = "studio:getStateContextMenu", e.setStateContextMenu = "studio:setStateContextMenu", e.contextMenuComponent = "studio:contextMenuComponent", e.layoutAdd = "studio:layoutAdd", e.layoutRemove = "studio:layoutRemove", e.layoutToggle = "studio:layoutToggle", e.layoutUpdate = "studio:layoutUpdate", e.layoutGet = "studio:layoutGet", e.layoutConfigGet = "studio:layoutConfigGet", e.layoutConfigSet = "studio:layoutConfigSet", e.layoutComponentAdd = "studio:layoutComponentAdd", e.layoutComponentGet = "studio:layoutComponentGet", e.layoutComponentRemove = "studio:layoutComponentRemove", e.layoutComponentRender = "studio:layoutComponentRender", e.getStateTheme = "studio:getStateTheme", e.setStateTheme = "studio:setStateTheme", e.settings = "studio:settings", e.assetProviderGet = "studio:assetProviderGet", e.assetProviderAdd = "studio:assetProviderAdd", e.assetProviderRemove = "studio:assetProviderRemove", e.fontGet = "studio:fontGet", e.fontAdd = "studio:fontAdd", e.fontRemove = "studio:fontRemove", e.fontManagerOpen = "studio:fontManagerOpen", e.menuFontLoad = "studio:menuFontLoad", e.toggleStateDataSource = "studio:toggleStateDataSource", e.getStateDataSource = "studio:getStateDataSource", e.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", e.dataSourceSetImporter = "studio:dataSourceSetImporter", e.dataSourceSetExporter = "studio:dataSourceSetExporter", e.setDragAbsolute = "studio:setDragAbsolute", e))(M || {});
const U = {
  [P.free]: 0,
  [P.startup]: 10,
  [P.business]: 20,
  [P.enterprise]: 30
};
function K(e) {
  const o = e;
  return o.init = (n) => (i) => e(i, n), o;
}
const q = (e) => /* @__PURE__ */ K(e);
async function H({
  editor: e,
  plan: o,
  pluginName: n,
  licenseKey: i,
  onLicenseCheckResponse: a,
  cleanup: g
}) {
  let d = "", r;
  const p = (s) => {
    console.warn("Cleanup plugin:", n, "Reason:", s), g();
  }, y = (s = {}) => {
    var A;
    const { error: T, sdkLicense: R } = s, h = (A = s.plan) == null ? void 0 : A.category;
    if (!(R || s.license) || T)
      p(T || "Invalid license");
    else if (h) {
      const w = U[o], E = U[h];
      w > E && p({ pluginRequiredPlan: o, licensePlan: h });
    }
  };
  e.Commands.has(M.settings) && (r = e.runCommand(M.settings), d = (r == null ? void 0 : r.baseUrl) || "");
  const f = (s) => {
    a == null || a(s), s && y(s);
  };
  if (!r) {
    e.onReady(async () => {
      if (!j())
        if (i) {
          const s = await z({ licenseKey: i, pluginName: n, baseApiUrl: d });
          f(s);
        } else
          p("The `licenseKey` option not provided");
    });
    return;
  }
  if (r.licensePlan || r.licenseError) {
    const s = J(r);
    f(s);
    return;
  }
  e.on(N, (s) => f(s));
}
const J = (e) => ({
  sdkLicense: e.license,
  license: e.license,
  error: e.licenseError,
  plan: e.licensePlan
});
async function z(e) {
  const { licenseKey: o, pluginName: n, baseApiUrl: i } = e;
  try {
    return (await _({
      baseApiUrl: i,
      path: `/sdk/${o || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: n
      }
    })).result || {};
  } catch (a) {
    return console.error("Error during SDK license check:", a), !1;
  }
}
const Q = "canvasEmptyState", X = P.startup, Y = function(e, o = {}) {
  const n = /* @__PURE__ */ new WeakMap(), i = /* @__PURE__ */ new WeakMap(), a = /* @__PURE__ */ new WeakMap(), g = /* @__PURE__ */ new Set(), d = /* @__PURE__ */ new WeakMap(), r = {
    emptyStates: [],
    ...o
  }, p = (t, l) => {
    let b = !1;
    const { isValid: u } = l;
    return Array.isArray(u) ? b = u.some((c) => t.is(c)) : F(u) ? b = u({ component: t, editor: e }) : b = t.is(u), b;
  }, y = (t) => {
    const l = n.get(t);
    n.delete(t), l == null || l();
  }, f = (t) => {
    t.views.forEach((l) => y(l)), a.delete(t);
  }, s = (t) => {
    if (!(!t || g.has(t)))
      try {
        g.add(t);
        const l = t.components().length > 0, b = a.get(t);
        if (l && b)
          f(t);
        else if (!l && !b) {
          const u = d.has(t) ? d.get(t) : r.emptyStates.find((c) => p(t, c));
          if (d.set(t, u), !u) return;
          t.views.forEach((c) => {
            const D = u.render({
              editor: e,
              component: t,
              componentView: c,
              mount: (v) => {
                i.set(c, v), a.set(t, !0);
                const G = c.getChildrenContainer();
                G == null || G.appendChild(v);
              },
              unmount: () => y(c)
            });
            n.set(c, () => {
              D == null || D();
              const v = i.get(c);
              v == null || v.remove();
            });
          });
        }
      } finally {
        g.delete(t);
      }
  }, T = (t) => {
    a.has(t) && f(t);
  }, R = ({ model: t }) => {
    s(t.getComponent());
  }, h = e.Components.events, L = `${h.update}:components`, A = "component:mount", w = e.Canvas.events.frameLoadBody, E = h.removed;
  e.on(L, s), e.on(A, s), e.on(E, T), e.on(w, R), H({
    editor: e,
    licenseKey: r.licenseKey,
    plan: X,
    pluginName: Q,
    cleanup: () => {
      e.off(L, s), e.off(A, s), e.off(E, T), e.off(w, R);
    }
  });
}, Z = q(Y);
export {
  Z as default
};
