// @grapesjs/studio-sdk-plugins/dist/listPagesComponent/index.es.js
const B = "app.grapesjs.com", O = "app-stage.grapesjs.com", $ = "app2.grapesjs.com", N = "app-stage2.grapesjs.com", G = [
  B,
  $,
  O,
  N,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], _ = "license:check:end", U = () => typeof window < "u", F = ({
  isDev: e,
  isStage: s,
  isPlatform: a
}) => `${e ? "" : `https://${s ? O : B}`}/${a ? "platform-api" : "api"}`, j = () => {
  const e = U() && window.location.hostname;
  return !!e && (G.includes(e) || G.some((s) => e.endsWith(s)));
};
async function W({
  path: e,
  baseApiUrl: s,
  method: a = "GET",
  headers: t = {},
  params: n,
  body: r
}) {
  const i = `${s || F({ isDev: !1, isStage: !1 })}${e}`, l = {
    method: a,
    headers: {
      "Content-Type": "application/json",
      ...t
    }
  };
  r && (l.body = JSON.stringify(r));
  const p = n ? new URLSearchParams(n).toString() : "", c = p ? `?${p}` : "", o = await fetch(`${i}${c}`, l);
  if (!o.ok)
    throw new Error(`HTTP error! status: ${o.status}`);
  return o.json();
}
var b = /* @__PURE__ */ ((e) => (e.free = "free", e.startup = "startup", e.business = "business", e.enterprise = "enterprise", e))(b || {}), E = /* @__PURE__ */ ((e) => (e.toastAdd = "studio:toastAdd", e.toastRemove = "studio:toastRemove", e.dialogOpen = "studio:dialogOpen", e.dialogClose = "studio:dialogClose", e.dialogExportCode = "studio:dialogExportCode", e.dialogImportCode = "studio:dialogImportCode", e.sidebarLeftSet = "studio:sidebarLeft:set", e.sidebarLeftGet = "studio:sidebarLeft:get", e.sidebarLeftToggle = "studio:sidebarLeft:toggle", e.sidebarRightSet = "studio:sidebarRight:set", e.sidebarRightGet = "studio:sidebarRight:get", e.sidebarRightToggle = "studio:sidebarRight:toggle", e.sidebarTopSet = "studio:sidebarTop:set", e.sidebarTopGet = "studio:sidebarTop:get", e.sidebarTopToggle = "studio:sidebarTop:toggle", e.sidebarBottomSet = "studio:sidebarBottom:set", e.sidebarBottomGet = "studio:sidebarBottom:get", e.sidebarBottomToggle = "studio:sidebarBottom:toggle", e.symbolAdd = "studio:symbolAdd", e.symbolDetach = "studio:symbolDetach", e.symbolOverride = "studio:symbolOverride", e.symbolPropagateStyles = "studio:propagateStyles", e.getPagesConfig = "studio:getPagesConfig", e.setPagesConfig = "studio:setPagesConfig", e.getPageSettings = "studio:getPageSettings", e.setPageSettings = "studio:setPageSettings", e.clearPage = "studio:clearPage", e.projectFiles = "studio:projectFiles", e.validateCode = "studio:validateCode", e.canvasReload = "studio:canvasReload", e.getBlocksPanel = "studio:getBlocksPanel", e.setBlocksPanel = "studio:setBlocksPanel", e.getStateContextMenu = "studio:getStateContextMenu", e.setStateContextMenu = "studio:setStateContextMenu", e.contextMenuComponent = "studio:contextMenuComponent", e.layoutAdd = "studio:layoutAdd", e.layoutRemove = "studio:layoutRemove", e.layoutToggle = "studio:layoutToggle", e.layoutUpdate = "studio:layoutUpdate", e.layoutGet = "studio:layoutGet", e.layoutConfigGet = "studio:layoutConfigGet", e.layoutConfigSet = "studio:layoutConfigSet", e.layoutComponentAdd = "studio:layoutComponentAdd", e.layoutComponentGet = "studio:layoutComponentGet", e.layoutComponentRemove = "studio:layoutComponentRemove", e.layoutComponentRender = "studio:layoutComponentRender", e.getStateTheme = "studio:getStateTheme", e.setStateTheme = "studio:setStateTheme", e.settings = "studio:settings", e.assetProviderGet = "studio:assetProviderGet", e.assetProviderAdd = "studio:assetProviderAdd", e.assetProviderRemove = "studio:assetProviderRemove", e.fontGet = "studio:fontGet", e.fontAdd = "studio:fontAdd", e.fontRemove = "studio:fontRemove", e.fontManagerOpen = "studio:fontManagerOpen", e.menuFontLoad = "studio:menuFontLoad", e.toggleStateDataSource = "studio:toggleStateDataSource", e.getStateDataSource = "studio:getStateDataSource", e.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", e.dataSourceSetImporter = "studio:dataSourceSetImporter", e.dataSourceSetExporter = "studio:dataSourceSetExporter", e.setDragAbsolute = "studio:setDragAbsolute", e))(E || {});
const V = {
  [b.free]: 0,
  [b.startup]: 10,
  [b.business]: 20,
  [b.enterprise]: 30
};
function K(e) {
  const s = e;
  return s.init = (a) => (t) => e(t, a), s;
}
const m = (e) => /* @__PURE__ */ K(e);
async function q({
  editor: e,
  plan: s,
  pluginName: a,
  licenseKey: t,
  onLicenseCheckResponse: n,
  cleanup: r
}) {
  let g = "", i;
  const l = (o) => {
    console.warn("Cleanup plugin:", a, "Reason:", o), r();
  }, p = (o = {}) => {
    var h;
    const { error: P, sdkLicense: L } = o, u = (h = o.plan) == null ? void 0 : h.category;
    if (!(L || o.license) || P)
      l(P || "Invalid license");
    else if (u) {
      const T = V[s], I = V[u];
      T > I && l({ pluginRequiredPlan: s, licensePlan: u });
    }
  };
  e.Commands.has(E.settings) && (i = e.runCommand(E.settings), g = (i == null ? void 0 : i.baseUrl) || "");
  const c = (o) => {
    n == null || n(o), o && p(o);
  };
  if (!i) {
    e.onReady(async () => {
      if (!j())
        if (t) {
          const o = await Y({ licenseKey: t, pluginName: a, baseApiUrl: g });
          c(o);
        } else
          l("The `licenseKey` option not provided");
    });
    return;
  }
  if (i.licensePlan || i.licenseError) {
    const o = X(i);
    c(o);
    return;
  }
  e.on(_, (o) => c(o));
}
const X = (e) => ({
  sdkLicense: e.license,
  license: e.license,
  error: e.licenseError,
  plan: e.licensePlan
});
async function Y(e) {
  const { licenseKey: s, pluginName: a, baseApiUrl: t } = e;
  try {
    return (await W({
      baseApiUrl: t,
      path: `/sdk/${s || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: a
      }
    })).result || {};
  } catch (n) {
    return console.error("Error during SDK license check:", n), !1;
  }
}
const x = "gjs-plg-", Z = "listPagesComponent", J = b.startup;
function z(e, s, a) {
  let t, n, r, g, i;
  const l = s;
  function p() {
    const o = Date.now() - g;
    o < l && o >= 0 ? t = setTimeout(p, l - o) : (t = null, i = e.apply(r, n), r = n = null);
  }
  const c = function() {
    return r = this, n = arguments, g = Date.now(), t || (t = setTimeout(p, l)), i;
  };
  return c.clear = function() {
    t && (clearTimeout(t), t = null);
  }, c.flush = function() {
    t && (i = e.apply(r, n), r = n = null, clearTimeout(t), t = null);
  }, c;
}
const k = "page://", d = "list-pages", f = "list-pages-item", y = "list-pages-link", Q = (e) => e.components().reduce(
  (s, a) => {
    var g;
    const t = a.findFirstType(y), n = (g = t == null ? void 0 : t.attributes.attributes) == null ? void 0 : g.href, r = n != null && n.startsWith(k) ? n.replace(k, "") : "";
    if (r && t) {
      const i = s[r];
      i ? i.push(t) : s[r] = [t];
    }
    return s;
  },
  {}
), A = "data-type-role", S = "page_root", C = (e) => (s) => s.is(e), ee = (e) => (s, a) => a.is(e), M = (e) => (s) => {
  var a;
  return ((a = s.getAttribute) == null ? void 0 : a.call(s, A)) === e;
}, te = function(e, s = {}) {
  const { Blocks: a, Components: t, Pages: n } = e, { licenseKey: r, block: g = {} } = s, i = "List Pages", l = "List Pages Item", p = "Link", c = `${x}${d}`, o = `${x}${y}`, P = '<svg viewBox="0 0 24 24"><path d="M11 15H17V17H11V15M9 7H7V9H9V7M11 13H17V11H11V13M11 9H17V7H11V9M9 11H7V13H9V11M21 5V19C21 20.1 20.1 21 19 21H5C3.9 21 3 20.1 3 19V5C3 3.9 3.9 3 5 3H19C20.1 3 21 3.9 21 5M19 5H5V19H19V5M9 15H7V17H9V15Z"/></svg>', L = '<svg viewBox="0 0 24 24"><path d="M7,5H21V7H7V5M7,13V11H21V13H7M4,4.5A1.5,1.5 0 0,1 5.5,6A1.5,1.5 0 0,1 4,7.5A1.5,1.5 0 0,1 2.5,6A1.5,1.5 0 0,1 4,4.5M4,10.5A1.5,1.5 0 0,1 5.5,12A1.5,1.5 0 0,1 4,13.5A1.5,1.5 0 0,1 2.5,12A1.5,1.5 0 0,1 4,10.5M7,19V17H21V19H7M4,16.5A1.5,1.5 0 0,1 5.5,18A1.5,1.5 0 0,1 4,19.5A1.5,1.5 0 0,1 2.5,18A1.5,1.5 0 0,1 4,16.5Z" /></svg>';
  t.addType(d, {
    block: g && {
      label: i,
      media: P,
      category: "Extra",
      select: !0,
      ...s.block
    },
    isComponent: M(d),
    model: {
      defaults: {
        tagName: "ul",
        name: i,
        [S]: "",
        classes: c,
        attributes: { [A]: d, role: "navigation" },
        droppable: C(f),
        styles: `
          .${c} {
            list-style: none;
            margin: 0;
            padding: 0;
          }
        `
      },
      init() {
        this.syncPageItems = z(this.syncPageItems.bind(this), 0), this.listenTo(this.em, n.events.all, this.syncPageItems), this.syncPageItems();
      },
      syncPageItems() {
        if (e.Components.getSymbolInfo(this).isInstance) return;
        const u = this.em.Pages.getAll(), H = this.components(), h = Q(this);
        u.forEach((T, I) => {
          const D = T.id, R = T.getName(), v = h[D];
          v ? v == null || v.forEach((w) => {
            w.getInnerHTML() !== R && w.components(R);
          }) : H.add(
            {
              type: f,
              components: [
                {
                  type: y,
                  attributes: { href: `page://${D}` },
                  components: { type: "textnode", content: R }
                }
              ]
            },
            { at: I }
          );
        });
      }
    }
  }), t.addType(f, {
    isComponent: M(f),
    model: {
      defaults: {
        name: l,
        tagName: "li",
        icon: L,
        attributes: { [A]: f },
        components: { type: y },
        draggable: ee(d)
      }
    }
  }), t.addType(y, {
    extend: "link",
    isComponent: M(y),
    model: {
      defaults: {
        removable: !1,
        copyable: !1,
        delegate: {
          move: (u) => u.closestType(f)
        },
        name: p,
        classes: o,
        attributes: { [A]: y },
        components: { type: "textnode", content: "Item" },
        styles: `
          .${o} {
            text-decoration: none;
          }
        `
      }
    }
  }), q({
    editor: e,
    licenseKey: r,
    plan: J,
    pluginName: Z,
    cleanup: () => {
      a.remove(d), t.removeType(d);
    }
  });
}, se = m(te);
export {
  se as default
};
