// @grapesjs/studio-sdk-plugins/dist/tableComponent/index.es.js
function ce(e, t, o) {
  let l, s, n, c, a;
  const r = t;
  function h() {
    const d = Date.now() - c;
    d < r && d >= 0 ? l = setTimeout(h, r - d) : (l = null, a = e.apply(n, s), n = s = null);
  }
  const v = function() {
    return n = this, s = arguments, c = Date.now(), l || (l = setTimeout(h, r)), a;
  };
  return v.clear = function() {
    l && (clearTimeout(l), l = null);
  }, v.flush = function() {
    l && (a = e.apply(n, s), n = s = null, clearTimeout(l), l = null);
  }, v;
}
const F = "app.grapesjs.com", U = "app-stage.grapesjs.com", ie = "app2.grapesjs.com", he = "app-stage2.grapesjs.com", S = [
  F,
  ie,
  U,
  he,
  "localhost",
  "127.0.0.1",
  ".local-credentialless.webcontainer.io",
  // For stackblitz.com demos
  ".local.webcontainer.io",
  // For stackblitz.com demos
  "-sandpack.codesandbox.io"
  // For Sandpack demos
], de = "license:check:end", ve = () => typeof window < "u", ue = ({
  isDev: e,
  isStage: t,
  isPlatform: o
}) => `${e ? "" : `https://${t ? U : F}`}/${o ? "platform-api" : "api"}`, ge = () => {
  const e = ve() && window.location.hostname;
  return !!e && (S.includes(e) || S.some((t) => e.endsWith(t)));
};
function $(e) {
  return typeof e == "function";
}
async function be({
  path: e,
  baseApiUrl: t,
  method: o = "GET",
  headers: l = {},
  params: s,
  body: n
}) {
  const a = `${t || ue({ isDev: !1, isStage: !1 })}${e}`, r = {
    method: o,
    headers: {
      "Content-Type": "application/json",
      ...l
    }
  };
  n && (r.body = JSON.stringify(n));
  const h = s ? new URLSearchParams(s).toString() : "", v = h ? `?${h}` : "", d = await fetch(`${a}${v}`, r);
  if (!d.ok)
    throw new Error(`HTTP error! status: ${d.status}`);
  return d.json();
}
var C = /* @__PURE__ */ ((e) => (e.free = "free", e.startup = "startup", e.business = "business", e.enterprise = "enterprise", e))(C || {}), x = /* @__PURE__ */ ((e) => (e.toastAdd = "studio:toastAdd", e.toastRemove = "studio:toastRemove", e.dialogOpen = "studio:dialogOpen", e.dialogClose = "studio:dialogClose", e.dialogExportCode = "studio:dialogExportCode", e.dialogImportCode = "studio:dialogImportCode", e.sidebarLeftSet = "studio:sidebarLeft:set", e.sidebarLeftGet = "studio:sidebarLeft:get", e.sidebarLeftToggle = "studio:sidebarLeft:toggle", e.sidebarRightSet = "studio:sidebarRight:set", e.sidebarRightGet = "studio:sidebarRight:get", e.sidebarRightToggle = "studio:sidebarRight:toggle", e.sidebarTopSet = "studio:sidebarTop:set", e.sidebarTopGet = "studio:sidebarTop:get", e.sidebarTopToggle = "studio:sidebarTop:toggle", e.sidebarBottomSet = "studio:sidebarBottom:set", e.sidebarBottomGet = "studio:sidebarBottom:get", e.sidebarBottomToggle = "studio:sidebarBottom:toggle", e.symbolAdd = "studio:symbolAdd", e.symbolDetach = "studio:symbolDetach", e.symbolOverride = "studio:symbolOverride", e.symbolPropagateStyles = "studio:propagateStyles", e.getPagesConfig = "studio:getPagesConfig", e.setPagesConfig = "studio:setPagesConfig", e.getPageSettings = "studio:getPageSettings", e.setPageSettings = "studio:setPageSettings", e.clearPage = "studio:clearPage", e.projectFiles = "studio:projectFiles", e.validateCode = "studio:validateCode", e.canvasReload = "studio:canvasReload", e.getBlocksPanel = "studio:getBlocksPanel", e.setBlocksPanel = "studio:setBlocksPanel", e.getStateContextMenu = "studio:getStateContextMenu", e.setStateContextMenu = "studio:setStateContextMenu", e.contextMenuComponent = "studio:contextMenuComponent", e.layoutAdd = "studio:layoutAdd", e.layoutRemove = "studio:layoutRemove", e.layoutToggle = "studio:layoutToggle", e.layoutUpdate = "studio:layoutUpdate", e.layoutGet = "studio:layoutGet", e.layoutConfigGet = "studio:layoutConfigGet", e.layoutConfigSet = "studio:layoutConfigSet", e.layoutComponentAdd = "studio:layoutComponentAdd", e.layoutComponentGet = "studio:layoutComponentGet", e.layoutComponentRemove = "studio:layoutComponentRemove", e.layoutComponentRender = "studio:layoutComponentRender", e.getStateTheme = "studio:getStateTheme", e.setStateTheme = "studio:setStateTheme", e.settings = "studio:settings", e.assetProviderGet = "studio:assetProviderGet", e.assetProviderAdd = "studio:assetProviderAdd", e.assetProviderRemove = "studio:assetProviderRemove", e.fontGet = "studio:fontGet", e.fontAdd = "studio:fontAdd", e.fontRemove = "studio:fontRemove", e.fontManagerOpen = "studio:fontManagerOpen", e.menuFontLoad = "studio:menuFontLoad", e.toggleStateDataSource = "studio:toggleStateDataSource", e.getStateDataSource = "studio:getStateDataSource", e.dataSourceSetGlobalData = "studio:dataSourceSetGlobalData", e.dataSourceSetImporter = "studio:dataSourceSetImporter", e.dataSourceSetExporter = "studio:dataSourceSetExporter", e.setDragAbsolute = "studio:setDragAbsolute", e))(x || {});
const E = {
  [C.free]: 0,
  [C.startup]: 10,
  [C.business]: 20,
  [C.enterprise]: 30
};
function pe(e) {
  const t = e;
  return t.init = (o) => (l) => e(l, o), t;
}
const fe = (e) => /* @__PURE__ */ pe(e);
async function me({
  editor: e,
  plan: t,
  pluginName: o,
  licenseKey: l,
  onLicenseCheckResponse: s,
  cleanup: n
}) {
  let c = "", a;
  const r = (d) => {
    console.warn("Cleanup plugin:", o, "Reason:", d), n();
  }, h = (d = {}) => {
    var y;
    const { error: p, sdkLicense: V } = d, m = (y = d.plan) == null ? void 0 : y.category;
    if (!(V || d.license) || p)
      r(p || "Invalid license");
    else if (m) {
      const z = E[t], A = E[m];
      z > A && r({ pluginRequiredPlan: t, licensePlan: m });
    }
  };
  e.Commands.has(x.settings) && (a = e.runCommand(x.settings), c = (a == null ? void 0 : a.baseUrl) || "");
  const v = (d) => {
    s == null || s(d), d && h(d);
  };
  if (!a) {
    e.onReady(async () => {
      if (!ge())
        if (l) {
          const d = await ye({ licenseKey: l, pluginName: o, baseApiUrl: c });
          v(d);
        } else
          r("The `licenseKey` option not provided");
    });
    return;
  }
  if (a.licensePlan || a.licenseError) {
    const d = we(a);
    v(d);
    return;
  }
  e.on(de, (d) => v(d));
}
const we = (e) => ({
  sdkLicense: e.license,
  license: e.license,
  error: e.licenseError,
  plan: e.licensePlan
});
async function ye(e) {
  const { licenseKey: t, pluginName: o, baseApiUrl: l } = e;
  try {
    return (await be({
      baseApiUrl: l,
      path: `/sdk/${t || "na"}`,
      method: "POST",
      params: {
        d: window.location.hostname,
        pn: o
      }
    })).result || {};
  } catch (s) {
    return console.error("Error during SDK license check:", s), !1;
  }
}
const Ce = "tableComponent", He = C.startup;
var i = /* @__PURE__ */ ((e) => (e.table = "table", e.tableSettings = "tableSettings", e.tableDelete = "tableDelete", e.tableHead = "tableHead", e.tableFoot = "tableFoot", e.tableCellItems = "tableCellItems", e.tableCellProperties = "tableCellProperties", e.tableCellMerge = "tableCellMerge", e.tableCellSplit = "tableCellSplit", e.tableRowItems = "tableRowItems", e.tableRowInsertBefore = "tableRowInsertBefore", e.tableRowInsertAfter = "tableRowInsertAfter", e.tableRowDelete = "tableRowDelete", e.tableRowProperties = "tableRowProperties", e.tableRowCut = "tableRowCut", e.tableRowCopy = "tableRowCopy", e.tableRowPasteBefore = "tableRowPasteBefore", e.tableRowPasteAfter = "tableRowPasteAfter", e.tableColumn = "tableColumn", e.tableColumnItems = "tableColumnItems", e.tableColumnInsertBefore = "tableColumnInsertBefore", e.tableColumnInsertAfter = "tableColumnInsertAfter", e.tableColumnDelete = "tableColumnDelete", e.tableColumnCut = "tableColumnCut", e.tableColumnCopy = "tableColumnCopy", e.tableColumnPasteBefore = "tableColumnPasteBefore", e.tableColumnPasteAfter = "tableColumnPasteAfter", e))(i || {});
const u = {
  table: '<svg viewBox="0 0 24 24"><path d="M5 4h14a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2m0 4v4h6V8H5m8 0v4h6V8h-6m-8 6v4h6v-4H5m8 0v4h6v-4h-6Z"/></svg>',
  tableSettings: '<svg viewBox="0 0 24 24"><path d="M5 4h14a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2m0 4v4h6V8H5m8 0v4h6V8h-6m-8 6v4h6v-4H5m8 0v4h6v-4h-6Z"/></svg>',
  tableDelete: '<svg viewBox="0 0 24 24"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6c0-1.1.9-2 2-2h14zM5 6v12h14V6H5z"/><path d="m14.4 8.6 1 1-2.3 2.4 2.3 2.4-1 1-2.4-2.3-2.4 2.3-1-1 2.3-2.4-2.3-2.4 1-1 2.4 2.3z"/></svg>',
  tableHead: '<svg viewBox="0 0 24 24"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6c0-1.1.9-2 2-2h14Zm-8 11H5v3h6v-3Zm8 0h-6v3h6v-3Zm0-5h-6v3h6v-3ZM5 13h6v-3H5v3Z"/></svg>',
  tableFoot: '<svg viewBox="0 0 24 24" data-test="1" style="transform: rotate(180deg);"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6c0-1.1.9-2 2-2h14Zm-8 11H5v3h6v-3Zm8 0h-6v3h6v-3Zm0-5h-6v3h6v-3ZM5 13h6v-3H5v3Z"/></svg>',
  tableColumn: '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6c0-1.1.9-2 2-2h14Zm0 9h-4v5h4v-5Zm-6 0H9v5h4v-5Zm0-7H9v5h4V6Zm6 0h-4v5h4V6Z"/></svg>',
  tableCellItems: "",
  tableCellProperties: '<svg viewBox="0 0 24 24"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6l2-2h14zm-8 9H5v5h6v-5zm8 0h-6v5h6v-5zm-8-7H5v5h6V6z"/></svg>',
  tableCellMerge: '<svg viewBox="0 0 24 24"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6l2-2h14zM5 16v2h3v-2H5zm14-5h-9v7h9v-7zm0-5h-4v3h4V6zM8 6H5v3h3V6zm5 0h-3v3h3V6zm-8 8h3v-3H5v3z"/></svg>',
  tableCellSplit: '<svg viewBox="0 0 24 24"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6c0-1.1.9-2 2-2h14zM8 15.5H5V18h3v-2.5zm11-5h-9V18h9v-7.5zm-2.5 1 1 1-2 2 2 2-1 1-2-2-2 2-1-1 2-2-2-2 1-1 2 2 2-2zm-8.5-1H5v3h3v-3zM19 6h-4v2.5h4V6zM8 6H5v2.5h3V6zm5 0h-3v2.5h3V6z"/></svg>',
  tableRowItems: "",
  tableRowInsertBefore: '<svg viewBox="0 0 24 24"><path d="M6 4a1 1 0 1 1 0 2H5v6h14V6h-1a1 1 0 0 1 0-2h2c.6 0 1 .4 1 1v13a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5c0-.6.4-1 1-1h2zm5 10H5v4h6v-4zm8 0h-6v4h6v-4zM12 3c.5 0 1 .4 1 .9V6h2a1 1 0 0 1 0 2h-2v2a1 1 0 0 1-2 .1V8H9a1 1 0 0 1 0-2h2V4c0-.6.4-1 1-1z"/></svg>',
  tableRowInsertAfter: '<svg viewBox="0 0 24 24"><path d="M12 13c.5 0 1 .4 1 .9V16h2a1 1 0 0 1 .1 2H13v2a1 1 0 0 1-2 .1V18H9a1 1 0 0 1-.1-2H11v-2c0-.6.4-1 1-1zm6 7a1 1 0 0 1 0-2h1v-6H5v6h1a1 1 0 0 1 0 2H4a1 1 0 0 1-1-1V6c0-1.1.9-2 2-2h14a2 2 0 0 1 2 2v13c0 .5-.4 1-.9 1H18zM11 6H5v4h6V6zm8 0h-6v4h6V6z"/></svg>',
  tableRowDelete: '<svg viewBox="0 0 24 24"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6c0-1.1.9-2 2-2h14zm0 2H5v3h2.5v2H5v2h2.5v2H5v3h14v-3h-2.5v-2H19v-2h-2.5V9H19V6zm-4.7 1.8 1.2 1L13 12l2.6 3.3-1.2 1-2.3-3-2.3 3-1.2-1L11 12 8.5 8.7l1.2-1 2.3 3 2.3-3z"/></svg>',
  tableRowProperties: '<svg viewBox="0 0 24 24"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6c0-1.1.9-2 2-2h14zM5 15v3h6v-3H5zm14 0h-6v3h6v-3zm0-9h-6v3h6V6zM5 9h6V6H5v3z"/></svg>',
  tableRowCut: '<svg viewBox="0 0 24 24"><path fill-rule="evenodd" d="M22 3v5H9l3 3 2-2h4l-4 4 1 1h.5c.4 0 .8 0 1.2.3.5.2.9.4 1.2.8.7.6 1 1.3 1 2.2.2.8 0 1.5-.6 2l-.8.5-1 .2c-.4 0-.8 0-1.2-.3a3.9 3.9 0 0 1-2.1-2.2c-.2-.5-.3-1-.2-1.5l-1-1-1 1c0 .5 0 1-.2 1.5-.1.5-.4 1-.9 1.4-.3.4-.7.6-1.2.8l-1.2.3c-.4 0-.7 0-1-.2-.3 0-.6-.3-.8-.5-.5-.5-.8-1.2-.7-2a3 3 0 0 1 1-2.2A3.7 3.7 0 0 1 8.6 14H9l1-1-4-4-.5-1a3.3 3.3 0 0 1 0-2c0-.4.3-.7.5-1l2 2V3h14zM8.5 15.3h-.3a2.6 2.6 0 0 0-.8.4 2.5 2.5 0 0 0-.9 1.1l-.1.4v.7l.2.5.5.2h.7a2.5 2.5 0 0 0 .8-.3L9 18l.4-.4.2-.4.1-.4v-.7a1 1 0 0 0-.2-.5 1 1 0 0 0-.4-.2h-.5zm7 0H15a1 1 0 0 0-.4.3 1 1 0 0 0-.2.5 1.5 1.5 0 0 0 0 .7v.4a2.8 2.8 0 0 0 .5.7h.1a2.8 2.8 0 0 0 .8.6l.4.1h.7l.5-.2.2-.5v-.7a2.6 2.6 0 0 0-.3-.8 2.4 2.4 0 0 0-.7-.7 2.5 2.5 0 0 0-.8-.3h-.3zM12 11.6c-.4 0-.7.1-1 .4-.3.3-.4.6-.4 1s.1.7.4 1c.3.3.6.4 1 .4s.7-.1 1-.4c.3-.3.4-.6.4-1s-.1-.7-.4-1c-.3-.3-.6-.4-1-.4zm8.5-7.1h-11v2h11v-2z"/></svg>',
  tableRowCopy: '<svg viewBox="0 0 24 24"><path d="M22 11v7H6v-7h16zm-2 2H8v3h12v-3zm-1-6v2H4v5H2V7h17z"/></svg>',
  tableRowPasteBefore: '<svg viewBox="0 0 24 24"><path fill-rule="evenodd" d="M12 1a3 3 0 0 1 2.8 2H18c1 0 2 .8 2 1.9V7h-2V5h-2v1c0 .6-.4 1-1 1H9a1 1 0 0 1-1-1V5H6v13h12v-4h2v4c0 1-.8 2-1.9 2H6c-1 0-2-.8-2-1.9V5c0-1 .8-2 1.9-2h3.3A3 3 0 0 1 12 1zm10 7v5H8V8h14zm-1.5 1.5h-11v2h11v-2zM12 3a1 1 0 1 0 0 2 1 1 0 0 0 0-2z"/></svg>',
  tableRowPasteAfter: '<svg viewBox="0 0 24 24"><path fill-rule="evenodd" d="M12 1a3 3 0 0 1 2.8 2H18c1 0 2 .8 2 1.9V11h-2V5h-2v1c0 .6-.4 1-1 1H9a1 1 0 0 1-1-1V5H6v13h14c0 1-.8 2-1.9 2H6c-1 0-2-.8-2-1.9V5c0-1 .8-2 1.9-2h3.3A3 3 0 0 1 12 1zm10 11v5H8v-5h14zm-1.5 1.5h-11v2h11v-2zM12 3a1 1 0 1 0 0 2 1 1 0 0 0 0-2z"/></svg>',
  tableColumnItems: "",
  tableColumnInsertBefore: '<svg viewBox="0 0 24 24"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H4a1 1 0 0 1-1-1v-2a1 1 0 0 1 2 0v1h8V6H5v1a1 1 0 1 1-2 0V5c0-.6.4-1 1-1h15zm0 9h-4v5h4v-5zM8 8c.5 0 1 .4 1 .9V11h2a1 1 0 0 1 .1 2H9v2a1 1 0 0 1-2 .1V13H5a1 1 0 0 1-.1-2H7V9c0-.6.4-1 1-1zm11-2h-4v5h4V6z"/></svg>',
  tableColumnInsertAfter: '<svg viewBox="0 0 24 24"><path d="M20 4c.6 0 1 .4 1 1v2a1 1 0 0 1-2 0V6h-8v12h8v-1a1 1 0 0 1 2 0v2c0 .5-.4 1-.9 1H5a2 2 0 0 1-2-2V6c0-1.1.9-2 2-2h15zM9 13H5v5h4v-5zm7-5c.5 0 1 .4 1 .9V11h2a1 1 0 0 1 .1 2H17v2a1 1 0 0 1-2 .1V13h-2a1 1 0 0 1-.1-2H15V9c0-.6.4-1 1-1zM9 6H5v5h4V6z"/></svg>',
  tableColumnDelete: '<svg viewBox="0 0 24 24"><path d="M19 4a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V6c0-1.1.9-2 2-2h14zm-4 4h-2V6h-2v2H9V6H5v12h4v-2h2v2h2v-2h2v2h4V6h-4v2zm.3.5 1 1.2-3 2.3 3 2.3-1 1.2L12 13l-3.3 2.6-1-1.2 3-2.3-3-2.3 1-1.2L12 11l3.3-2.5z"/></svg>',
  tableColumnCut: '<svg viewBox="0 0 24 24"><path fill-rule="evenodd" d="M7.2 4.5a3 3 0 0 1 2.2 1A3.7 3.7 0 0 1 10.5 8v.5l1 1 4-4 1-.5a3.3 3.3 0 0 1 2 0c.4 0 .7.3 1 .5L17 8h4v13h-6V10l-1.5 1.5.5.5v4l-2.5-2.5-1 1v.5c0 .4 0 .8-.3 1.2-.2.5-.4.9-.8 1.2-.6.7-1.3 1-2.2 1-.8.2-1.5 0-2-.6l-.5-.8-.2-1c0-.4 0-.8.3-1.2A3.9 3.9 0 0 1 7 12.7c.5-.2 1-.3 1.5-.2l1-1-1-1c-.5 0-1 0-1.5-.2-.5-.1-1-.4-1.4-.9-.4-.3-.6-.7-.8-1.2L4.5 7c0-.4 0-.7.2-1 0-.3.3-.6.5-.8.5-.5 1.2-.8 2-.7zm12.3 5h-3v10h3v-10zM8 13.8h-.3l-.4.2a2.8 2.8 0 0 0-.7.4v.1a2.8 2.8 0 0 0-.6.8l-.1.4v.7l.2.5.5.2h.7a2.6 2.6 0 0 0 .8-.3 2.4 2.4 0 0 0 .7-.7 2.5 2.5 0 0 0 .3-.8 1.5 1.5 0 0 0 0-.8 1 1 0 0 0-.2-.4 1 1 0 0 0-.5-.2H8zm3.5-3.7c-.4 0-.7.1-1 .4-.3.3-.4.6-.4 1s.1.7.4 1c.3.3.6.4 1 .4s.7-.1 1-.4c.3-.3.4-.6.4-1s-.1-.7-.4-1c-.3-.3-.6-.4-1-.4zM7 5.8h-.4a1 1 0 0 0-.5.3 1 1 0 0 0-.2.5v.7a2.5 2.5 0 0 0 .3.8l.2.3h.1l.4.4.4.2.4.1h.7L9 9l.2-.4a1.6 1.6 0 0 0 0-.8 2.6 2.6 0 0 0-.3-.8 2.5 2.5 0 0 0-1.2-1l-.4-.1H7z"/></svg>',
  tableColumnCopy: '<svg viewBox="0 0 24 24"><path d="M17 6v16h-7V6h7zm-2 2h-3v12h3V8zm-2-6v2H8v15H6V2h7z"/></svg>',
  tableColumnPasteBefore: '<svg viewBox="0 0 24 24"><path fill-rule="evenodd" d="M12 1a3 3 0 0 1 2.8 2H18c1 0 2 .8 2 1.9V18c0 1-.8 2-1.9 2H11v-2h7V5h-2v1c0 .6-.4 1-1 1H9a1 1 0 0 1-1-1V5H6v2H4V5c0-1 .8-2 1.9-2h3.3A3 3 0 0 1 12 1zm-2 7v12H4V8h6zM8.5 9.5h-3v9h3v-9zM12 3a1 1 0 1 0 0 2 1 1 0 0 0 0-2z"/></svg>',
  tableColumnPasteAfter: '<svg viewBox="0 0 24 24"><path fill-rule="evenodd" d="M12 1a3 3 0 0 1 2.8 2H18c1 0 2 .8 2 1.9V7h-2V5h-2v1c0 .6-.4 1-1 1H9a1 1 0 0 1-1-1V5H6v13h7v2H6c-1 0-2-.8-2-1.9V5c0-1 .8-2 1.9-2h3.3A3 3 0 0 1 12 1zm8 7v12h-6V8h6zm-1.5 1.5h-3v9h3v-9zM12 3a1 1 0 1 0 0 2 1 1 0 0 0 0-2z"/></svg>'
}, f = "table", W = "thead", _ = "tfoot", M = "tbody", b = "row", g = "cell", D = "clipboardTable", Re = [f, W, _, M, b, g], B = (e) => e.em.get(D) || [], Pe = (e) => e.every((t) => t.is(g)), T = (e, t) => {
  const { editor: o, event: l, opts: s } = e, n = l.currentTarget || l.target, { x: c, y: a, width: r, height: h } = n.getBoundingClientRect(), v = {
    style: { width: 240, height: 300 },
    placer: { type: "popover", closeOnClickAway: !0, x: c, y: a, w: r, h },
    layout: {
      type: "tabs",
      value: "props",
      tabs: [
        {
          id: "styles",
          label: "Styles",
          children: { type: "panelStyles" }
        },
        {
          id: "props",
          label: "Properties",
          children: { type: "panelProperties", style: { padding: "14px 7px" } }
        }
      ]
    },
    ...t
  };
  $(s.openSettings) ? s.openSettings({ ...e, layoutProps: v }) : o.runCommand(x.layoutToggle, v);
}, Ve = (e, {
  rows: t = 1,
  cells: o = 5,
  text: l
} = {}) => ({
  type: e,
  components: Array.from(Array(t).keys()).map((s) => ({
    type: b,
    components: Array.from(Array(o).keys()).map((n) => ({
      type: g,
      components: l ? { type: "text", content: $(l) ? l({ y: s, x: n }) : l } : void 0
    }))
  }))
}), j = "Table settings", K = (e) => {
  const { component: t } = e;
  return t.is(f) ? t : t.closestType(f);
}, ze = (e) => {
  var t;
  (t = K(e)) == null || t.remove(e.opts);
}, Ae = (e) => {
  const t = K(e);
  t && (e.editor.select(t), T(e, {
    id: "tableSettings",
    header: { label: j }
  }));
}, q = (e) => [{
  id: i.table,
  icon: u[i.table],
  label: "Table",
  items: [
    {
      id: i.tableSettings,
      label: j,
      icon: u[i.tableSettings],
      onClick: ({ event: o }) => Ae({
        ...e,
        // @ts-expect-error workaround for the build
        event: o
      })
    },
    {
      id: i.tableDelete,
      label: "Delete table",
      icon: u[i.tableDelete],
      onClick: () => ze(e)
    }
  ]
}], X = ({ editor: e }) => {
  const t = B(e);
  return t.length && Pe(t);
}, G = ({ component: e, editor: t, after: o }) => {
  if (!e.is(g)) return;
  const l = e.closestType(f), s = e.closestType(b), n = e.index() + (o ? 1 : 0);
  l == null || l.findType(b).forEach((a) => a.append({ type: g }, { at: n }));
  const c = s == null ? void 0 : s.components().at(n);
  c && t.select(c);
}, Y = ({ component: e, opts: t }) => {
  if (!e.is(g)) return;
  const o = e.closestType(f), l = e.index();
  o == null || o.findType(b).forEach((s) => {
    var n;
    return (n = s.components().at(l)) == null ? void 0 : n.remove(t);
  });
}, J = ({ component: e, editor: t }) => {
  if (!e.is(g)) return;
  const o = e.closestType(f), l = e.index(), s = (o == null ? void 0 : o.findType(b).map((n) => n.components().at(l))) || [];
  s.length && t.em.set(D, s);
}, N = (e) => {
  const { component: t, editor: o, after: l } = e;
  if (!t.is(g) || !X(e)) return;
  const s = B(o), n = t.closestType(f), c = t.index() + (l ? 1 : 0);
  n == null || n.findType(b).forEach((a, r) => {
    const h = s[r], v = (h == null ? void 0 : h.clone()) || { type: g };
    a.append(v, { at: c });
  });
}, xe = (e) => {
  const { component: t } = e;
  t.is(g) && (J(e), Y({ ...e, opts: { temporary: !0 } }));
}, Be = (e) => {
  const t = X(e);
  return [{
    id: i.tableColumnItems,
    icon: u[i.tableColumn],
    label: "Column",
    items: [
      {
        id: i.tableColumnInsertBefore,
        label: "Insert column before",
        icon: u[i.tableColumnInsertBefore],
        onClick: () => G(e)
      },
      {
        id: i.tableColumnInsertAfter,
        label: "Insert column after",
        icon: u[i.tableColumnInsertAfter],
        onClick: () => G({ ...e, after: !0 })
      },
      {
        id: i.tableColumnDelete,
        label: "Delete column",
        icon: u[i.tableColumnDelete],
        onClick: () => Y(e)
      },
      {
        id: i.tableColumnCut,
        label: "Cut column",
        icon: u[i.tableColumnCut],
        onClick: () => xe(e)
      },
      {
        id: i.tableColumnCopy,
        label: "Copy column",
        icon: u[i.tableColumnCopy],
        onClick: () => J(e)
      },
      {
        id: i.tableColumnPasteBefore,
        label: "Paste column before",
        icon: u[i.tableColumnPasteBefore],
        disabled: !t,
        onClick: () => N(e)
      },
      {
        id: i.tableColumnPasteAfter,
        label: "Paste column after",
        icon: u[i.tableColumnPasteAfter],
        disabled: !t,
        onClick: () => N({ ...e, after: !0 })
      }
    ]
  }];
}, Q = "Row settings", Ie = (e) => e.every((t) => t.is(b)), P = (e) => {
  const { component: t } = e, o = t.is(g), l = t.is(b);
  if (!(!o && !l))
    return l ? t : t.closestType(b);
}, Z = (e) => {
  const { component: t, editor: o, after: l } = e, s = P(e);
  if (!s) return;
  const n = s.index() + (l ? 1 : 0), c = s.components().length, a = Array.from(Array(c).keys()).map(() => ({ type: g })), r = s.parent();
  if (r) {
    const h = r.append({ type: b, components: a }, { at: n })[0];
    if (t.is(g) && h) {
      const v = t.index(), d = h.components().at(v);
      d && o.select(d);
    }
  }
}, ee = (e) => {
  var t;
  (t = P(e)) == null || t.remove(e.opts);
}, te = (e) => {
  const t = P(e);
  t && e.editor.em.set(D, [t]);
}, O = (e) => {
  var a;
  const { editor: t, after: o } = e, l = P(e);
  if (!l || !le(e)) return;
  const n = B(t)[0], c = l.index() + (o ? 1 : 0);
  n && ((a = l.parent()) == null || a.append(n.clone(), { at: c }));
}, Me = (e) => {
  P(e) && (te(e), ee({ ...e, opts: { temporary: !0 } }));
}, le = ({ editor: e }) => {
  const t = B(e);
  return t.length && Ie(t);
}, ke = (e) => {
  const t = P(e);
  t && (e.editor.select(t), T(e, {
    id: "tableRowSettings",
    header: { label: Q }
  }));
}, oe = (e) => {
  const t = le(e);
  return [{
    id: i.tableRowItems,
    icon: u[i.tableRowProperties],
    label: "Row",
    items: [
      {
        id: i.tableRowInsertBefore,
        label: "Insert row before",
        icon: u[i.tableRowInsertBefore],
        onClick: () => Z(e)
      },
      {
        id: i.tableRowInsertAfter,
        label: "Insert row after",
        icon: u[i.tableRowInsertAfter],
        onClick: () => Z({ ...e, after: !0 })
      },
      {
        id: i.tableRowDelete,
        label: "Delete row",
        icon: u[i.tableRowDelete],
        onClick: () => ee(e)
      },
      {
        id: i.tableRowProperties,
        label: Q,
        icon: u[i.tableRowProperties],
        onClick: ({ event: l }) => ke({
          ...e,
          // @ts-expect-error workaround for the build
          event: l
        })
      },
      {
        id: i.tableRowCut,
        label: "Cut row",
        icon: u[i.tableRowCut],
        onClick: () => Me(e)
      },
      {
        id: i.tableRowCopy,
        label: "Copy row",
        icon: u[i.tableRowCopy],
        onClick: () => te(e)
      },
      {
        id: i.tableRowPasteBefore,
        label: "Paste row before",
        icon: u[i.tableRowPasteBefore],
        disabled: !t,
        onClick: () => O(e)
      },
      {
        id: i.tableRowPasteAfter,
        label: "Paste row after",
        icon: u[i.tableRowPasteAfter],
        disabled: !t,
        onClick: () => O({ ...e, after: !0 })
      }
    ]
  }, ...q(e)];
}, L = (e) => {
  const t = e.attributes.attributes || {}, o = parseInt(`${t.rowspan || 1}`, 10), l = parseInt(`${t.colspan || 1}`, 10);
  return { rowspan: o, colspan: l };
};
function se(e) {
  const t = [];
  return e.components().forEach((o, l) => {
    o.components().each((s, n) => {
      const { rowspan: c, colspan: a } = L(s);
      if (t[l])
        for (; t[l][n]; )
          n++;
      for (let r = l; r < l + c; r++) {
        t[r] || (t[r] = []);
        for (let h = n; h < n + a; h++)
          t[r][h] = {
            real: r === l && h === n,
            cell: s,
            rowspan: c,
            colspan: a
          };
      }
    });
  }), t;
}
function H(e, t, o) {
  var l;
  return (l = e[o]) == null ? void 0 : l[t];
}
const k = (e) => e.closestType(b);
function De(e, { x: t, y: o, rows: l, cols: s }) {
  var a;
  let n = k(e[o][t].cell), c;
  for (let r = 1; r <= l; r++)
    if (n = (a = n == null ? void 0 : n.parent()) == null ? void 0 : a.components().at(n.index() + 1), n) {
      for (c = t; c >= 0; c--) {
        const { cell: h } = e[o + r][c], v = k(h);
        if (v === n) {
          for (let d = 1; d <= s; d++) {
            const p = h.index();
            v.append({ type: g }, { at: p + 1 });
          }
          break;
        }
      }
      if (c === -1) {
        const v = n.components().at(0).index();
        for (let d = 1; d <= s; d++)
          n.append({ type: g }, { at: v });
      }
    }
}
function ne(e) {
  for (let t = 0; t < e.length; t++)
    for (let o = 0; o < e[t].length; o++) {
      const s = e[t][o].cell;
      if (s.em.selected.hasComponent(s)) {
        const { rowspan: n, colspan: c } = L(s);
        if (c > 1 || n > 1) {
          s.addAttributes({ colspan: 1, rowspan: 1 });
          for (let a = 0; a < c - 1; a++) {
            const r = s.index();
            k(s).append({ type: g }, { at: r + 1 });
          }
          De(e, { x: o, y: t, rows: n - 1, cols: c });
        }
      }
    }
}
function Te(e) {
  const t = [];
  e.components().forEach((o) => {
    o.components().length || t.push(o);
  }), t.forEach((o) => o.remove());
}
const I = (e, t = {}) => {
  var c;
  const o = se(e), { em: l } = e;
  let s, n;
  for (let a = 0; a < o.length; a++)
    for (let r = 0; r < o[a].length; r++) {
      const h = o[a][r];
      if ((c = t.selected) != null && c.includes(h.cell) || l.selected.hasComponent(h.cell)) {
        const v = { x: r, y: a };
        s || (s = v), n = v;
      }
    }
  return {
    grid: o,
    startPos: s,
    endPos: n
  };
};
function Le(e, t = {}) {
  const o = I(e), { startPos: l, endPos: s } = o;
  let { grid: n } = o;
  const { em: c } = e;
  if (!l || !s)
    return;
  const a = l.x, r = l.y, h = s.x, v = s.y, d = H(n, a, r), p = H(n, h, v);
  if (d && p) {
    ne(n), n = se(e);
    const V = [], m = H(n, a, r).cell;
    m.addAttributes({
      colspan: h - a + 1,
      rowspan: v - r + 1
    });
    for (let w = r; w <= v; w++)
      for (let y = a; y <= h; y++) {
        const z = H(n, y, w);
        if (!z) continue;
        const { cell: A } = z;
        A !== m && V.push(A);
      }
    V.forEach((w) => {
      t.mergeContent && m.append(w.components().models), w.remove();
    }), c.setSelected(m), Te(e);
  }
}
const ae = "Cell settings", Se = "Merge cells", Ee = "Split cell", Ge = (e) => e.closestType(b), R = (e) => {
  var t;
  return (t = Ge(e)) == null ? void 0 : t.parent();
}, Ne = (e) => {
  const t = e.editor.getSelectedAll(), o = R(t[0]), l = t[0];
  if (!o || !(l != null && l.is(g)) || t.length > 1)
    return !1;
  const { rowspan: s, colspan: n } = L(l);
  return !(n < 2 && s < 2);
}, Ze = (e) => {
  const o = e.editor.getSelectedAll()[0], l = R(o), { grid: s } = I(l);
  ne(s);
}, re = (e) => {
  const t = e.editor.getSelectedAll(), o = R(t[0]);
  if (!o || t.length < 2 || t.some((a) => !a.is(g) || R(a) !== o))
    return !1;
  const { grid: l, startPos: s, endPos: n } = I(o), { em: c } = o;
  if (!s || !n) return !1;
  for (let a = s.y; a <= n.y; a++)
    for (let r = s.x; r <= n.x; r++) {
      const h = H(l, r, a);
      if (!h || !c.selected.hasComponent(h.cell))
        return !1;
    }
  return !0;
}, Oe = (e) => {
  const { editor: t } = e;
  if (!re(e)) return;
  const o = t.getSelectedAll(), l = R(o[0]);
  Le(l, { mergeContent: !0 });
}, Fe = (e) => {
  T(e, {
    id: "tableCellSettings",
    header: { label: ae }
  });
}, Ue = (e) => [{
  id: i.tableCellItems,
  icon: u[i.tableCellProperties],
  label: "Cell",
  items: [
    {
      id: i.tableCellProperties,
      label: ae,
      icon: u[i.tableCellProperties],
      onClick: ({ event: o }) => Fe({
        ...e,
        // @ts-expect-error workaround for the build
        event: o
      })
    },
    {
      id: i.tableCellMerge,
      label: Se,
      icon: u[i.tableCellMerge],
      onClick: () => Oe(e),
      disabled: !re(e)
    },
    {
      id: i.tableCellSplit,
      label: Ee,
      icon: u[i.tableCellSplit],
      onClick: () => Ze(e),
      disabled: !Ne(e)
    }
  ]
}, ...Be(e), ...oe(e)], $e = (e) => {
  const t = e[0] && R(e[0]);
  if (!t) return;
  const { grid: o, startPos: l, endPos: s } = I(t, { selected: e });
  if (!l || !s) return;
  const n = [];
  for (let c = l.y; c <= s.y; c++)
    for (let a = l.x; a <= s.x; a++) {
      const r = H(o, a, c);
      r && n.push(r.cell);
    }
  n.length && t.em.setSelected(n, {
    // @ts-expect-error custom option
    fromCell: !0
  });
}, We = function(e, t = {}) {
  const o = {
    block: {},
    ...t
  }, { Components: l } = e, s = {
    copyable: !1,
    emptyState: !0,
    highlightable: !1,
    stylable: [
      "color",
      "font-family",
      "font-size",
      "font-weight",
      "line-height",
      "text-align",
      "vertical-align",
      "background-color",
      "border-width",
      "border-style",
      "border-color"
    ]
  };
  l.addType(f, {
    block: o.block && {
      label: "Table",
      media: u.table,
      category: "Extra",
      content: {
        type: f,
        style: { width: "100%" }
      },
      select: !0,
      ...o.block
    },
    model: {
      defaults: {
        emptyState: !0,
        highlightable: !1,
        icon: u.table,
        stylable: [
          "width",
          "height",
          "margin",
          "font-family",
          "font-size",
          "font-weight",
          "line-height",
          "text-align",
          "background",
          "background-color",
          "border-width",
          "border-style",
          "border-color"
        ],
        traits: [
          {
            type: "number",
            min: 0,
            name: "cellspacing",
            label: "Cell spacing"
          },
          {
            // CSS alt: 'border-spacing' (doesn't work with border collapsed)
            type: "number",
            min: 0,
            name: "cellpadding",
            label: "Cell padding"
          },
          {
            id: "tableCaption",
            type: "checkbox",
            label: "Caption",
            setValue({ component: c, value: a, emitUpdate: r }) {
              const h = c.find("caption")[0];
              a ? !h && c.append({
                type: "text",
                tagName: "caption",
                components: { type: "textnode", content: "Caption" }
              }) : h == null || h.remove(), r(), e.refresh({ tools: !0 });
            },
            getValue(c) {
              return !!c.component.find("caption")[0];
            }
          }
        ],
        components: Ve(M, { rows: 5 }),
        contextMenu: ({ items: c, editor: a, component: r }) => [
          ...q({ editor: a, component: r, opts: o }),
          ...c
        ]
      }
    }
  }), l.addType(W, {
    model: {
      defaults: {
        icon: u.tableHead,
        ...s
      }
    }
  }), l.addType(_, {
    model: {
      defaults: {
        icon: u.tableFoot,
        ...s
      }
    }
  }), l.addType(M, {
    model: {
      defaults: {
        icon: u.tableRowProperties,
        ...s
      }
    }
  }), l.addType(b, {
    model: {
      defaults: {
        icon: u.tableRowProperties,
        emptyState: !0,
        highlightable: !1,
        stylable: ["height", "text-align", "border-width", "border-style", "border-color"],
        traits: [],
        contextMenu: ({ items: c, editor: a, component: r }) => [
          ...oe({ editor: a, component: r, opts: o }),
          ...c
        ]
      }
    }
  }), l.addType(g, {
    model: {
      defaults: {
        icon: u.tableCellProperties,
        emptyState: {
          styleOut: "height: 1px"
        },
        stylable: [
          "width",
          "height",
          "padding",
          "background",
          "background-color",
          "text-align",
          "vertical-align",
          "border-width",
          "border-style",
          "border-color"
        ],
        traits: [
          {
            id: "tableCellType",
            type: "select",
            options: [
              { value: "td", name: "Cell" },
              { value: "th", name: "Header Cell" }
            ],
            label: "Cell type",
            name: "tagName",
            changeProp: !0
          },
          {
            id: "tableCellScope",
            type: "select",
            options: [
              { value: !1, name: "None" },
              { value: "row", name: "Row" },
              { value: "col", name: "Column" },
              { value: "rowgroup", name: "Row group" },
              { value: "colgroup", name: "Column group" }
            ],
            label: "Scope",
            name: "scope"
          }
        ],
        contextMenu: ({ items: c, editor: a, component: r }) => [
          ...Ue({ editor: a, component: r, opts: o }),
          ...c
        ]
      }
    }
  });
  const n = ce((c, a) => {
    var d;
    const r = e.getSelectedAll(), h = !!((d = a == null ? void 0 : a.event) != null && d.shiftKey);
    if (r.length < 2 || !h || a.fromCell) return;
    const v = r.map((p) => p.is(g) ? p : p.closestType(g)).filter(Boolean);
    v.length === r.length && $e(v);
  }, 0);
  e.on("component:select", n), me({
    editor: e,
    plan: He,
    pluginName: Ce,
    licenseKey: o.licenseKey,
    cleanup: () => {
      e.off("component:select", n), e.Blocks.remove(f), Re.forEach((c) => l.removeType(c));
    }
  });
}, _e = fe(We);
export {
  _e as default
};
